shiny应用的记录变量

get_DEPglobal_env

assign(".OmicsModule", data.frame(name = input$OmicsName,omictype = input$OmicsType, ID = ID), envir = get_DEPglobal_env())

assign(".PostModule", data.frame(name = input$OmicsName,omictype = input$OmicsType), envir = get_DEPglobal_env())

"Over-representation analysis" = "ORA",
                                                             "GSEA" = "GSEA",
                                                             "Protein-protein interaction" = "PPI",
                                                             "integrated analysis" = "Instegrated"





```R
awesomeRadio(inputId = "OmicsType",
                                                 label = "Select omics type",
                                                 width = "100%",
                                                 choices = list("Quantificated proteome(based on proteingroup quantification)" = "Proteomepg",
                                                                "Quantificated proteome(based on peptide quantification)" = "Proteomepep",
                                                                "modification-specific proteome(modified peptide quantification)" = "PTM",
                                                                "transcriptome(RNA counts)" = "RNAseq")))
```



genelist tool 的 参数传递

```R
1 
# genelist global UI
genelist_tool_UI(ID)
genelist_tool_Server(id, omics_res) 
#从shiny server 中得到 page 的 ID 和 omics_res 信息并，渲染dragzone 和 dropzone， 提供对应的dragZone choices和dropZoneInput choices
2. 
# exctract list。为每个dropzone choice 的渲染 UI
purrr::map(setdiff(input$dropzone, names(layer_modules)), ~ { layer_modules[[.]] <- exctract_genelist_Server(id = ., ID=., Omics_Serv_res = Omics_res)  })
# choices id 通过 ID， Omics_res 通过 Omics_Serv_res 传递到 exctract_genelist_Server
# creat the choices list for dropzone from Omics_res_list
drop_choose <- reactive({
            x = lapply(names(Omics_res_list()), exctract_genelist_UI)
            names(x) = names(Omics_res_list())
            return(x)
          })
          dropZoneInput(session$ns("dropzone"),
                        choices = drop_choose(),、
                        selectable = F,
                        multivalued = TRUE,
                        server = exctract_genelist_UI)
        })
# exctract_genelist_UI 将接收dropZoneInput 的 choice ID，及所对应的omics id， 得到对应的 session。
# exctract_genelist_Server 根据 omics ID 判断 dropzoneinput 的 omics 类型，进一步进行 renderUI
```



peptide 归属 proteingroup的问题

```R
#方法一： 只保留最短的UniqueGroups。c(a,b,a;b;c，c;d;e)。 只保留a,b,c;d;e。直接去除非UniqueGroups的肽段
pe_norm2 <- filterFeatures(pe_norm, ~ Proteins %in% smallestUniqueGroups(rowData(pe_norm[["peptideNorm"]])$Proteins))

#方法二：MQ方法，最少的proteingroups组合。保留最短UniqueGroups的基础上，对剩余非 unique 肽段进行razor peptide 的分配。使得每一条肽段都有分配
rd <- rowData(pe_norm[["peptideNorm"]])
SU_PGs <- smallestUniqueGroups(rd$Proteins)
SU_PGs_uniquecounts <- rd$Proteins[rd$Proteins %in% SU_PGs] %>% table

```









PTM protein-level correct

```R
 bsCollapsePanel("Protein quantification correction",
                          style = "primary",
                          prettySwitch(
                            inputId = ns("Protein_level_correction"),
                            label = "Normalization before filter peptide", 
                            status = "success",
                            fill = F
                          ),
                          uiOutput(ns("protein_group_choose")),
                          uiOutput(ns("Protein_level_correction_options")),
                          shinyBS::bsTooltip(ns("Protein_level_correction"), "correct modified peptide intensity by protein-level qutification")
                          
          ),


output$protein_group_choose <- renderUI({
          Omics_res_list = Omics_res_list()
          Omics_res_list2 = Omics_res_list[grep("^Proteome",names(Omics_res_list))]
          
          if(length(Omics_res_list2) == 0) {
            Omics_res_list2 = NULL
          }else{
            Omics_res_list2 = Omics_res_list2[
              which((Omics_res_list2 %>% 
                       lapply(., function(x){
                         x1 <<- x
                         temp = try(x(), silent = T)
                         if(class(temp) == "try-error")
                           return(FALSE)
                         return(TRUE)
                         # is.na(x1 <<- x)
                       }) %>%
                       unlist()))
            ]}
          
          conditionalPanel(
            condition = paste0("input['", session$ns("Protein_level_correction"),"']"),
            selectInput(inputId = session$ns("protein_group_choose"),
                        label = "Corresponding proteomics", choices = names(Omics_res_list2))
          )
          
        })

 output$Protein_level_correction_options <- renderUI({
          options <- tagList(
            radioButtons(
              inputId = session$ns("correct_key"),
              label = "The column correction based on", 
              choices = list("Gene name" = "name", "Protein ID" = "ID"),
              inline = TRUE
              # status = "success",
              # fill = TRUE
            ),
            radioButtons(
              inputId = session$ns("unidentified_protein_treatment"),
              label = "How to treat peptide whose proteins unidentified in corresponding proteomics", 
              choices = c("retain", "remove"),
              inline = TRUE,
              # status = "success",
              # fill = TRUE
            ),
            radioButtons(
              inputId = session$ns("correct_level"),
              label = "Corresponding relationship between", 
              choices = c("condition to condition", "repetition to repetition"),
              inline = TRUE,
              # status = "success",
              # fill = TRUE
            ),
            shinyBS::bsTooltip(session$ns("correct_key"), 
                               "Match enriched-peptides quantity and corresponding proteomics based on Gene name or protein ID (both were set in the Columns options)", 
                               "top", options = list(container = "body")),
            shinyBS::bsTooltip(session$ns("unidentified_protein_treatment"), 
                               "Some enriched peptides' proteins may lack effective quantity in corresponding proteomics. /n
                               Choose remove to eliminate these peptides or choose retain to transmit the original peptide quantity(without correction)", 
                               "top", options = list(container = "body"))
            
          )
          conditionalPanel(
            condition = paste0("input['", session$ns("Protein_level_correction"),"']"),
            options)
          
        })
```

