---
title: "DEP2 Analyze workflow"
author: 
- name: Zhenhuan Feng
package: DEP2
output: BiocStyle::html_document
abstract: |
  This vigenette introduce proteiomics analysis workflow in DEP2 package. 
  This vigenette used the example dataset from a multiple omics study of silicosis mouse model.
vignette: |
  %\VignetteIndexEntry{DEP2: Proteomics Analyze}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = F,
  eval = T,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

```{r include=FALSE}
suppressPackageStartupMessages({
  # devtools::load_all()
  library("BiocStyle")
  library(tidyr)
  library(DEP2)
  library(SummarizedExperiment)
  library(tibble)
  library(dplyr)
})
writeLines(capture.output(sessionInfo()), "sessionInfo.txt")
```

# Introduction
`DEP2` provides a differentially expressed/enriched analyze toolkits for mass spectrometry based quantitative proteomics data via `limma`, upgraded from a previous package `DEP`. This package provides an integrated workflow for proteomics analyze, including data processing, missing value imputation, hypothesis test, visualization and downstream function exploration. It accepts various of proteomics results generated by upstream search and quantitative software.

Now, `DEP2` provide three types of differentail proteomics analyze:

1. The pipeline for proteins-level expression/enrichment analysis, started from protein-level quantity.

    This pipeline is basically follows the methods in `DEP`, requires a protein-level quantity result (e.g. proteingroups.txt). 
    
2. The pipeline for proteins-level expression/enrichment analysis, started from peptide-level quantity.

    This pipeline clustered the peptide to protein summarization strategies in `QFeatures`, requires a protein-level quantity (e.g. proteingroups.txt). 
    
3. The pipeline for post-translation modification (PTM) specified proteomics, performed upon modified peptides quantity.

    This ppipline derived from the first one, and concludes the additional modification information in analysis.
    
# Input data
`DEP2` requires wide result table, and formats the identifiers (gene name and protein id in general) by `make_unique()` or `make_unique_ptm()`.

## Load protein-level data
Read in a proteingroup example (the quantitative result from MaxQuant), then format name and id of proteingroups.
```{r read-in,collapse=T}
## a proteinGroups data of a silicosis mouse model
data(Silicosis_pg) 
colnames(Silicosis_pg)

## Formatting name(gene symbol) and id(protein ID). 
## Generate a unique names for each protein. 
unique_pg <- make_unique(Silicosis_pg, names = "Gene.names", 
                         ids = "Protein.IDs", delim = ";") # names and ids are columns in table
## The duplicated names are uniqued by increasing decimal number.
unique_pg %>% filter(Gene.names!="") %>% group_by(Gene.names) %>% summarize(frequency = n(), names = paste0(name,collapse = ";")) %>% 
  arrange(desc(frequency)) %>% filter(frequency > 1) %>% head
```


Build a SummarizedExperiment object. DEP2 still use `r Biocpkg("SummarizedExperiment") ` objects as the analyze container to store expression assay,
features information (except expression value) and experiment design. Experiment design that label sample information is optional in workflow. DEP/DEP2
can except a table contains 'label', 'condition' and 'replicate' columns, or automatically assign it by column names parsing. 

```{r SE,collapse=T}
## Take expression columns(LFQ intensity in this cases).
ecols <- grep("LFQ.intensity.", colnames(unique_pg))

## Construct SE
### 1. import a experiment design table.
expdesign <- read.delim("../inst/example/Silicosis_expdesign.txt")
head(expdesign,6) # the example experiment design
se_pg <- DEP2::make_se(unique_pg, columns = ecols, expdesign = expdesign, 
                       log2transform = T)
class(se_pg)
head(assay(se_pg), 5) ## log2-transformed expression assay 

### 2. Generate a experiment design by parsing expression column names.
se_pg2 <- DEP2::make_se_parse(unique_pg, columns = ecols, mode = "delim", 
                              sep = "_", remove_prefix = T, log2transform = T)
```

## Load peptide quantification result for protein aggregation

The second pipeline is to calculate protein-level abundance by peptide summation. It can bypass the summation method in the upstream software.
DEP2 draw the protein aggregation though the `r Biocpkg("QFeatures") `, and use the `QFeatures` container to process peptide-level data. 
This pipeline requires a peptide result table contains both abundance assay and relative protein information (gene names or protein ID) for protein aggregation.


<!-- ```{r message=FALSE} -->
<!-- library(QFeatures) -->
<!-- ``` -->

```{r}
## a peptides table from a silicosis mouse model.
data(Silicosis_peptide) 
(ecols <- grep("Intensity.", colnames(Silicosis_peptide), value = T)) 
pe_peptides <- make_pe_parse(Silicosis_peptide, columns = ecols,   # columns is the abundance columns
                             remove_prefix = T, log2transform = T) # log2transform 
pe_peptides # a QFeatures object, with a peptideRaw assay
colData(pe_peptides)
```

## Load the modified-site quantification result to perform differetial PTM analysis.

For, PTM-specific proteomics study, DEP2 design a new workflow similar from global proteomics analyze. 
In this pipeline, post-translation modification information is necessary to distinguish modified sites.
This pipeline required the unique identifier for distinguish different modified sites, 
modification information such as position, modified residue(amino acid), 
gene name or protein ID that modification is associated with.

```{r}
## phosphorylated peptides table of the silicosis mouse model.
data(Silicosis_phos) 

## Format the modification information and generated modified-peptides identifier.
## aa and pos is the modified amino acids and modified site in protein.
unique_pho <- make_unique_ptm(Silicosis_phos, gene_name = "Gene.names", 
                              protein_ID = "Protein", aa = "Amino.acid",
                              pos = "Position") 
```

`make_unique_ptm` creates(or overwrites) the PTM information columns, including 
*name*,*ID*,*gene_name*,*protein_ID*,*modified_aa*,*modified_pos*. The *name*,*ID* columns follow such naming ruleï¼š
'(gene name/protein ID)_(modified amino acid)(position of modification)' (e.g. "TBCA_K51" and "O75347_K51").

```{r}
## The formatted name and ID for PTM
unique_pho %>% select("name","ID","gene_name","protein_ID","modified_aa","modified_pos") %>% head(7)

## Take expression columns.
ecols <- grep("Intensity.", colnames(unique_pho))

## Construct a SE object mentioned before.
se_ptm <- DEP2::make_se_parse(unique_pho, columns = ecols, mode = "delim", sep = "_", remove_prefix = T, log2transform = T)
class(se_ptm)
```


# Differentially expression analysis
DEP2 provide an entire workflow for each pipeline.

## Analysis for proteomics data
**Filter**
Undesired features, like reverse sequence or contaminant protein hits, may contain in the result table.
Besides, missing values is inevasible label-free MS-based proteomics especially for DDA data.
Filter out reverse, contaminant, and low quality features with many missing is essential for following statistical test.
`filter_pe`
```{r}

```


## Analysis based on protein re-aggregation

## Analysis for PTM-specific proteomics
















