---
title: "Data import"
author: 
- name: Zhenhuan Feng
package: DEP2
output: 
  # rmarkdown::html_document
  BiocStyle::html_document:
    toc_float: true
abstract: |
  This vigenette introduces how DEP2 receives results from upstream quantitative tools like MaxQuant, DIA-NN, 
  Spectronaut, or Fragpipe.
vignette: |
  %\VignetteIndexEntry{DEP2: Data import}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  message = FALSE,
  comment = "#>"
)
library(BiocStyle)
```


# Introduction

The `DEP2` package utilizes the `make_unique` function to format protein/PTM-peptide-level tables and then imports them into S4 objects via `make_se/pe` function. The parameters in these functions can be adjusted to accommodate different result tables generated from search and quantitative software. Here, we present the workflow and provide detailed information on the inputs from four different quantitative software programs.

# Example data

We obtained a benchmark proteomics dataset by spiking Yeast and E.coli lysate into a Hela background. The raw data was obtained using an Orbitrap Fusion Lumos Tribrid equipment in either DDA or DIA mode. The DDA data was searched and quantified using MaxQuant or FragPipe, while the DIA data was analyzed using DIA-NN or Spectronaut in library-free mode.

The example tables used in this paper can be found in this in this [repository](https://github.com/mildpiggy/OmicsExample).

```{r workplace,eval=FALSE}
# Download OmicsExample firstly
# Change the path to Omics Example
example_path = "the/path/to/OmicsExample"
knitr::opts_knit$set(
  root.dir = example_path
  )

library(DEP2)
library(dplyr)
```

```{r workplace0,include=FALSE}
example_path = "../example_data/OmicsExample/"
knitr::opts_knit$set(
  root.dir = example_path
  )
library(DEP2)
library(dplyr)
```

# Constructing S4 object

The DEP2 analysis workflow is build upon S4 object, so the initial step is to format the table and convert it into an S4 container.

## Converting protein-level data into SE format

For protein-level data, the gene name or protein ID should be used as the feature identifier. DEP2 provides `make_unique` and `make_unique_ptm` functions to format the feature identifiers in proteomics results. The `make_unique` function cleans the identifiers, keeps the first name/ID in each group, and makes repeated identifiers unique by adding a suffix ('ProXXX', 'ProXXX.1', 'ProXXX.2'). Repeated identifiers are common when library search is performed with isoforms or unreviewed sequences. The *names* and *ids* parameters should be gene/protein names or IDs. The unique identifiers are saved in the new columns "name" and "id". Then, the `make_se` function can convert the 'uniqued' table into a *SE* object.

### Conversion from MaxQuant proteinGroups

In this pipeline, DEP2 use the table `proteinGroups.txt` as input.

```{r mq}
# Load proteinGroups table
mq_pgfile <- "./A_spikeIn/MaxQuant_output/proteinGroups.txt.gz"

pg <- read.csv(mq_pgfile,sep = "\t")
```

The *names* and *ids* are in the columns "Gene.names" and "Gene.names", respectively.

```{r, mq_uni}
## Generate unique names and ids
unique_pg <- make_unique(pg, 
                         name = "Gene.names", #gene 
                         ids = "Protein.IDs"   #protein
                         )

## New columns "name" and "id", which is necessary to function make_se
head(unique_pg[,c("name","ID","Gene.names","Protein.IDs")])
```

In this pipeline DEP2 uses SE object as the container to store expression assay, features information (row data) and experiment design (columns data). After `make_unique` identifiers, DEP2 provides the `make_se`/`make_se_parse` functions to convert *SummarizedExperiment* object. The former required an input experiment design table to annotate samples, which must contain three columns: 'label', 'condition', and 'replicate'. The 'parse' functions can automatically assign experiment design table by parsing the column names of the abundance assay.

```{r mq_se}
## Expression columns in the input data
ecols <- grep("LFQ.intensity.", colnames(unique_pg))

## Experiment design table
expDesign_file <- "./A_spikeIn/MaxQuant_output/expdesign.txt.gz"
expdesign <- read.table(expDesign_file, sep = "\t", header = T)
expdesign

## Convert SE with expdesign table
se <- make_se(unique_pg, columns = ecols, expdesign = expdesign)

## Convert SE with expdesign table using parse function
se <- make_se_parse(unique_pg, columns = ecols, mode = "delim")
```

How experiment design is extracted from colnames:

```{r get_exdesign}
## How experiment design is extracted from colnames. Split by delim or by character number 
sample_names = c("Quantity.A_1","Quantity.A_2","Quantity.B_1",
                 "Quantity.B_3","Quantity.B_2","Quantity.A_3")
get_exdesign_parse(sample_names,mode = "delim", sep = "_", remove_prefix = TRUE)

sample_names = c("Quantity.A1","Quantity.A2","Quantity.B1",
                 "Quantity.B3","Quantity.B2","Quantity.A3")
get_exdesign_parse(sample_names,mode = "char", chars = 1, remove_prefix = TRUE)
```

### Conversion from FragPipe combined_protein

Next, we use the table `combined_protein.tsv` from FragPipe as input.


```{r FragP}
# Load combined_protein.tsv table
FragP_pgfile <- "./A_spikeIn/FragPipe_output/combined_protein.tsv.gz"

FragP_pg <- read.csv(FragP_pgfile,sep = "\t")
```

**Unique identifier**. The *names* and *ids* are in the columns "Gene" and "Protein.ID".

```{r FragP_unique2}
## Generate unique names and ids
unique_FragP_pg <- make_unique(FragP_pg, 
                               name = "Gene",  #gene 
                               ids = "Protein.ID"   #protein
                               )

## new columns name and id, which is necessary to make_se
head(unique_FragP_pg[,c("name", "ID", "Gene", "Protein.ID")])
```

**Convert SE**. Here, We used the "MaxLFQ Intensity" values. "Intesity" is also practicable, but may bring some effect on the test result.

```{r FragP_se_frg}
## Expression col in the combined_protein.tsv
ecols <- grep(".MaxLFQ.Intensity", colnames(unique_FragP_pg),value = T)
ecols

## Convert SE with expdesign table
se <- make_se_parse(unique_FragP_pg, 
                    columns = ecols, 
                    mode = "delim",
                    remove_suffix = TRUE    ## remove the col suffix ".MaxLFQ.Intensity"
                    )

## The suffix is removed in SE
colData(se)
colnames(se)
```

### Conversion from DIA-NN result

As in the example above, `make_se` and `make_se_parse` accept wide-format tables which each row represents a features (a protein or a proteingroup), and rows contain the identification information and quantification information of samples. However, long-format tables are also widely used in many cases, such as the *report.tsv* output from DIA-NN or the *MSstats* output. Function `reshape_long2wide` can turn long-format tables into wide tables. Next, we will show the difference between handling wide- or long-tables.

#### Wide table output

The *report.pg_matrix.tsv* from DIA-NN is a wide-format table similar to *proteinGroups*.

```{r diann_load1}
# Load combined_protein.tsv table
Diann_pgfile <- "./A_spikeIn/Diann_output/report.pg_matrix.tsv.gz"

Diann_pg <- read.csv(Diann_pgfile,sep = "\t", fileEncoding="latin1")
```

**Unique identifier**. The *name* and *ids* are in the columns "Genes" and "Protein.Group".

```{r unique2}
## Generate unique names and ids
unique_diann_pg <- make_unique(Diann_pg, 
                               name = "Genes",  #gene 
                               ids = "Protein.Group"   #protein
                               )

## New columns ”name“ and "ID", which is necessary to make_se
head(unique_diann_pg[,c("name", "ID", "Genes", "Protein.Group")])
```

**Convert SE**. The expression columns are directly named by the file names of samples. We recommend to renaming the columns or renaming the MS files before DIAN-NN search.

```{r make_se_diann1}
## Expression col in the DIA-NN report.pg_matrix
ecols <- grep(".raw$", colnames(unique_diann_pg),value = T)
ecols

## Convert SE with expdesign table
se <- make_se_parse(unique_diann_pg, 
                    columns = ecols, 
                    mode = "delim",
                    remove_prefix = TRUE,   ## remove the file prefix 
                    remove_suffix = TRUE    ## remove the col suffix ".raw"
                    )

## The file prefix and suffix is removed in SE
colData(se)
colnames(se)
```

#### Long table output

Unlike the above cases, the **report.tsv** from DIA-NN is in long format. Function `reshape_long2wide` can reshape long tables to wide format, before constructing *SE* object.

```{r diann_load2}
## Load combined_protein.tsv table
Diann_repfile <- "./A_spikeIn/Diann_output/report.tsv.gz"
# Diann_repfile <- "./example_data/OmicsExample/A_spikeIn/Diann_output/report.tsv.gz"
Diann_rep <- read.csv(Diann_repfile,sep = "\t") #
dim(Diann_rep) # 390479 rows, each row is a precursor.

## Filter out proteingroups exceeding the threshold value
Diann_rep = dplyr::filter(Diann_rep, PG.Q.Value < 0.01)


## Table report.tsv stores both Precursor- and PG-level quality
DT::datatable(head(Diann_rep) %>% mutate_if(is.character, utf8::utf8_encode), options = list(scrollX = T,pageLength = 6))
```

**Reshape** long-table.

```{r diann_resha}

Diann_rep_wided = reshape_long2wide(Diann_rep,
                                    sample_col = "File.Name",      # the column labeling sample names
                                    feature_col = "Protein.Group", # PG IDs
                                    expression_col = "PG.MaxLFQ",  # PG quantity. Normalized one is also ok。
                                    shrink_ident_cols  = "Genes",  # Gene names
                                    extend_ident_cols =  "Protein.Q.Value",  # optional, some identification info.
                                    remove_sample_prefix = FALSE,  # remove prefix in sample_col
                                    remove_sample_suffix = FALSE   # remove suffix in sample_col
                                    )

```

The reshaping result has 'Genes' in a column and extends the 'Protein.Q.Value' to 6 columns, each representing a different sample. The variables in *shrink_ident_cols* are combined into a string by concatenating all the values for each feature and separating them with a ';'. The variables in *extend_ident_cols* are expanded into multiple columns, with each column containing the values for each sample (e.g., "Q.Value").

```{r DT_wide}
DT::datatable(head(Diann_rep_wided,3), options = list(scrollX = T,pageLength = 3))
```

**Unique identifier**. The *names* and *ids* are in the columns "Genes" and "Protein.Group".

```{r diann_unique1}
## Generate unique names and ids
unique_diann_pg2 <- make_unique(Diann_rep_wided, 
                                name = "Genes",  #gene 
                                ids = "Protein.Group"   #protein
                                )

## new columns name and id, which is necessary to make_se
head(unique_diann_pg2[,c("name", "ID", "Genes", "Protein.Group")],4)
```

**Convert SE**. The expression columns are also directly named by the file names of samples.

```{r make_se_diann3}
## Expression col in the DIA-NN report.pg_matrix
ecols <- grep(".raw$", colnames(unique_diann_pg2),value = T)
ecols # contains Protein.Q.Value columns

ecols = ecols[1:6]

## Convert SE with expdesign table
se <- make_se_parse(unique_diann_pg2, 
                    columns = ecols, 
                    mode = "delim",
                    remove_prefix = TRUE,   ## remove the file prefix 
                    remove_suffix = TRUE    ## remove the col suffix ".raw"
                    )

## The file prefix and suffix is removed in SE
colData(se)
colnames(se)
```


### Conversion from Spectronaut Report

Spectronaut offers the ability to customize the output table. In this example, we will use the built-in MSstats style table, which is in a long-format.

```{r spe_load1}
Spe_repfile <- "./A_spikeIn/Spectronaut_output/DIA_MSStats_Report.xls.gz"
Spe_rep <- read.csv(Spe_repfile,sep = "\t") 
dim(Spe_rep) # 318132 rows, each row is a precursor.

## Store both Precursor-, Peptides- and PG-level quality
DT::datatable(head(Spe_rep) %>% mutate_if(is.character, utf8::utf8_encode), options = list(scrollX = T,pageLength = 6))
```

**Reshape** long table.

```{r spe_resha}
Spe_rep_wided = reshape_long2wide(Spe_rep,
                                  sample_col = "R.FileName",      # the column labeling sample names
                                  feature_col = "PG.ProteinGroups", # PG IDs
                                  expression_col = "PG.Quantity",   # PG quantity.
                                  shrink_ident_cols  = c("PG.Genes","PG.Qvalue"),  # Gene names
                                  remove_sample_prefix = TRUE,    # remove prefix in sample_col
                                  remove_sample_suffix = FALSE  
                                  )

```

```{r spe_DT_wide}
DT::datatable(head(Spe_rep_wided), options = list(scrollX = T,pageLength = 6))
```

**Unique identifier**. The *names* and *Protein.Group* are located in the "Genes" and "Protein.Group" columns.

```{r spe_unique2}
## Generate unique names and ids
unique_spe_pg <- make_unique(Spe_rep_wided, 
                             name = "PG.Genes",  #gene 
                             ids = "PG.ProteinGroups"   #protein
                             )

## new columns name and id, which is necessary to make_se
head(unique_spe_pg[,c("name", "ID", "PG.Genes", "PG.ProteinGroups")],4)
```

**Convert SE**. The expression columns is directly named by the file names of samples.

```{r make_se_diann2}
## Expression columns in the Spectronaut output
ecols <- 2:7
colnames(unique_spe_pg)[ecols]

## Convert SE with expdesign table
se <- make_se_parse(unique_spe_pg, 
                    columns = ecols, 
                    mode = "delim",
                    remove_prefix = TRUE,   ## remove the file prefix 
                    remove_suffix = FALSE   ## no suffix in this case
                    )

## The file prefix and suffix is removed in SE
colData(se)
colnames(se)
```


## Converting peptide-level data to a QFeatures object

DEP2 provides a pipeline, using the QFeature package, to aggregate and summarize peptide quantities into protein-level.The protein-level quantities used in the preceding section were counted by upstream software and mostly summarized using the maxLFQ algorithm. DEP2 offers other aggregation strategies in this peptide-to-protein analysis pipeline.

The following code snippet demonstrates the conversion of peptide-level data to a QFeatures object using different software outputs.

### Conversion from MaxQuant peptides

The first step is to load in peptide-level data into *QFeatures* class via the `make_pe` or `make_pe_pars` function. We start by reading the peptides.txt file from the MaxQuant txt directory. 

```{r MQ_pep}
mq_pepfile <- "./A_spikeIn/MaxQuant_output/peptides.txt.gz"

mq_pep <- read.csv(mq_pepfile,sep = "\t")
```

To convert the data into a `QFeatures` object, we extract the 'intensity' columns.

```{r MQ_pep2}
ecols <- grep("Intensity.", colnames(mq_pep)) ## the peptides intensity cols
mq_pe <- make_pe(mq_pep, columns = ecols, expdesign = expdesign)
mq_pe
```


### Conversion from FragPipe combined_peptides

Next, we process the combined_peptide.tsv file generated by FragPipe.

```{r Fragp_pep1}
FragP_pepfile <- "./A_spikeIn/FragPipe_output/combined_peptide.tsv.gz"

FragP_pep <- read.csv(FragP_pepfile,sep = "\t")
```

We use the 'Intensity' columns as expression columns.

```{r Frag_pe1}
ecols <- grep("[0-9].Intensity", colnames(FragP_pep),value = T)  ## the peptides intensity cols

FragP_pe <- make_pe_parse(FragP_pep, columns = ecols, 
                          mode = "delim", 
                          remove_suffix = T ## remove suffix
                          )
colData(FragP_pe)
```


### Conversion from DIA-NN report

We **reshape** the peptide quantities into a wide table using the `reshape_long2wide` function. The 'Stripped.Sequence' column represents the peptide IDs. A stripped peptide may have multiple precursors due to variable modifications or different charge states. For these peptides, the function retains the maximum expression values.

```{r diann_pep1}

## filter out proteingroups exceed the threshold value
Diann_rep = dplyr::filter(Diann_rep, PG.Q.Value < 0.01)

Diann_pep_wided = reshape_long2wide(Diann_rep,
                                    sample_col = "File.Name",      # the column labeling sample names
                                    feature_col = "Stripped.Sequence", # PG IDs
                                    expression_col = "Precursor.Quantity",  # PG quantity. Normalized one is also ok。
                                    shrink_ident_cols  = c("Protein.Group","Genes"),  # Gene names
                                    extend_ident_cols =  c("Global.Q.Value"),  # optional, some identification info.
                                    remove_sample_prefix = FALSE,  # remove prefix in sample_col
                                    remove_sample_suffix = FALSE   # remove suffix in sample_col
                                    )
```

```{r diann_wide2}
DT::datatable(head(Diann_pep_wided), options = list(scrollX = T,pageLength = 6))
```

```{r diann_pe}
ecols = 2:7

pe <- make_pe_parse(Diann_pep_wided, columns = ecols, 
                    mode = "delim", 
                    remove_suffix = T ## remove suffix
                    )

```

### Conversion from Spectronaut result

The Spectronaut result also store the peptide-level quantities.

```{r spe_load2}
Spe_repfile <- "./A_spikeIn/Spectronaut_output/DIA_MSStats_Report.xls.gz"
Spe_rep <- read.csv(Spe_repfile,sep = "\t") #
dim(Spe_rep) # 318132 rows, each row is a precursor.

## Store both Precursor-, Peptides- and PG-level quality
DT::datatable(head(Spe_rep) %>% mutate_if(is.character, utf8::utf8_encode), options = list(scrollX = T,pageLength = 6))
```

**Reshape** the long-table.


```{r spe_resha2}
Spe_rep_wided2 = reshape_long2wide(Spe_rep,
                                   sample_col = "R.FileName",      # the column labeling sample names
                                   feature_col = "PEP.StrippedSequence", # peptide sequence
                                   expression_col = "PEP.Quantity",      # peptide quantity.
                                   shrink_ident_cols  = c("PG.Genes","PG.ProteinAccessions"),  # Gene names and protien IDs
                                   remove_sample_prefix = TRUE,    # remove prefix in sample_col
                                   remove_sample_suffix = FALSE  
                                   )
```

Convert *QFeatures* object

```{r Spe_pe}
ecols = 2:7

pe <- make_pe_parse(Spe_rep_wided2, columns = ecols, 
                    mode = "delim", 
                    remove_suffix = T ## remove suffix
                    )
colData(pe)
```


# Session information

```{r sessioninfo, echo=FALSE}
sessionInfo()
```
