<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DEP2">
<title>DEP2 analysis workflow • DEP2</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DEP2 analysis workflow">
<meta property="og:description" content="DEP2">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DEP2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.7.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/01_Quick_start.html">Quick start</a>
    <a class="dropdown-item" href="../articles/02_Data_import.html">Data import</a>
    <a class="dropdown-item" href="../articles/03_DEP2_analysis.html">DEP2 analysis workflow</a>
    <a class="dropdown-item" href="../articles/04_app_demo.html">DEP2 Application Guide</a>
    <a class="dropdown-item" href="../articles/05_Benchmark_on_spike_in_data.html">A benchmark on spike-in data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="03_DEP2_analysis_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="03_DEP2_analysis_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="03_DEP2_analysis_files/datatables-binding-0.28/datatables.js"></script><link href="03_DEP2_analysis_files/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="03_DEP2_analysis_files/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="03_DEP2_analysis_files/dt-core-1.13.4/js/jquery.dataTables.min.js"></script><link href="03_DEP2_analysis_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="03_DEP2_analysis_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>DEP2 analysis workflow</h1>
                        <h4 data-toc-skip class="author">Zhenhuan Feng</h4>
                        
      
      
      <div class="d-none name"><code>03_DEP2_analysis.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette introduces proteomics analysis workflow in DEP2 package. Here, we utilized the example dataset from a multiple omics study of silicosis mouse model.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>DEP2</code> is a downstream analysis toolkit for mass spectrometry-based quantitative proteomics data. The package contains an integrated workflow for proteomics analysis, including data processing, missing value imputation, hypothesis test, visualization and downstream function exploration. It accepts various of proteomics results generated by upstream search and quantitative software and provides functions to format table. In the vignette, we utilize the built-in data in DEP2. The native tables out from upstream software are available in this <a href="https://github.com/mildpiggy/OmicsExample?_blank" class="external-link">repository</a>.</p>
<p>Now, <code>DEP2</code> provide three types of differential proteomics analysis:</p>
<ol style="list-style-type: decimal">
<li>
<p>Protein-Level Expression/Enrichment Analysis Pipeline</p>
<p>This pipeline is basically follows the method in <code>DEP</code>, and requires a protein-level quantitative result (e.g., proteingroups.txt).</p>
</li>
<li>
<p>Protein Expression/Enrichment Analysis Pipeline with Peptide-protein aggregation.</p>
<p>This pipeline utilizes the peptide to protein summarization strategies in <code>QFeatures</code>, and requires a peptide-level quantitative result (e.g. peptide.txt).</p>
</li>
<li>
<p>Pipeline for Post-Translational Modification (PTM)-Specific Proteomics.</p>
<p>This pipeline is derived from the first one and requires PTM pepeide data as input. It includes the additional modification information in analysis.</p>
</li>
</ol>
<p>In addition to these proteomics analysis pipelines, DEP2 packages an optional RNA-seq analysis pipeline from <em><a href="https://bioconductor.org/packages/3.17/DESeq2" class="external-link">DESeq2</a></em>. Downstream exploration functions are also integrated into workflow, including functional enrichment provided by <em><a href="https://bioconductor.org/packages/3.17/clusterProfiler" class="external-link">clusterProfiler</a></em> <span class="citation">(Wu et al. 2021)</span> and <em><a href="https://bioconductor.org/packages/3.17/ReactomePA" class="external-link">ReactomePA</a></em>`, protein-protein interaction inference from STRING database, and time-course expression pattern clustering though cmeans cluster.</p>
</div>
<div class="section level2">
<h2 id="load-required-packages">Load required packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a>
</h2>
</div>
<div class="section level2">
<h2 id="differentially-expression-analysis">Differentially expression analysis<a class="anchor" aria-label="anchor" href="#differentially-expression-analysis"></a>
</h2>
<p>The steps of four analysis pipelines different slightly, such as the input table read in. For more details, please refer to the vignette titled <code>Data import</code>.</p>
<div class="section level3">
<h3 id="protein-level-expressionenrichment-analysis-pipeline">Protein-Level Expression/Enrichment Analysis Pipeline<a class="anchor" aria-label="anchor" href="#protein-level-expressionenrichment-analysis-pipeline"></a>
</h3>
<ol style="list-style-type: decimal">
<li><strong>Loading data</strong></li>
</ol>
<p><img src="figs/DEP2workflow1.png" title="Fig. 1 Pipeline for proteingroups"></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ProteinGroups table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Silicosis_pg"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Format name(gene symbol) and id(protein ID). </span></span>
<span><span class="co"># Generate a unique names for each protein. names and ids are columns in table</span></span>
<span><span class="va">unique_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_unique.html">make_unique</a></span><span class="op">(</span><span class="va">Silicosis_pg</span>, names <span class="op">=</span> <span class="st">"Gene.names"</span>, </span>
<span>                         ids <span class="op">=</span> <span class="st">"Protein.IDs"</span>, delim <span class="op">=</span> <span class="st">";"</span><span class="op">)</span> </span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li><strong>Converting SE object</strong></li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Take expression columns(LFQ intensity in this cases).</span></span>
<span><span class="va">ecols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"LFQ.intensity."</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">unique_pg</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Construct SE. The experiement design is exctracted from column.</span></span>
<span><span class="va">se_pg</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/make_se_parse.html">make_se_parse</a></span><span class="op">(</span><span class="va">unique_pg</span>, columns <span class="op">=</span> <span class="va">ecols</span>, mode <span class="op">=</span> <span class="st">"delim"</span>, </span>
<span>                             sep <span class="op">=</span> <span class="st">"_"</span>, remove_prefix <span class="op">=</span> <span class="cn">T</span>, log2transform <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li><strong>Filtering</strong></li>
</ol>
<p>The result table may contain undesired features such as reverse sequences or contaminant protein hits. Additionally, missing values (MVs) are inevitable in label-free MS-based proteomics, especially for DDA data. Low quantity features with too many missing values can introduce noise and bias into the analysis, potentially leading to false positive.Filtering out reverse, contaminant, and low-quality features with many missing values is necessary before performing statistical tests to improve accuracy and reliability of the statistical tests. The <code>filter_se</code> function can be used to filter features based on either an MVs threshold or other specified criteria.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Filter on MVs.</span></span>
<span><span class="va">filter_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">se_pg</span>,</span>
<span>                       thr <span class="op">=</span> <span class="fl">1</span>,  <span class="co">## the threshold of missing number in at least one condition</span></span>
<span>                       fraction <span class="op">=</span> <span class="fl">0.3</span> <span class="co">## the threshold of missing occupancy in each protein</span></span>
<span>                       </span>
<span>                      <span class="op">)</span></span>
<span><span class="co">#&gt; filter base on missing number is &lt;= 1 in at least one condition.</span></span>
<span><span class="co">#&gt; filter base on missing number fraction &lt; 0.3 in each row</span></span>
<span></span>
<span><span class="co">## Further filter can be performed though filter_formula.</span></span>
<span><span class="co">### Fox example, remove decoys or contaminants.</span></span>
<span><span class="va">filter_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">filter_pg</span>,</span>
<span>                       filter_formula <span class="op">=</span> <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">'+'</span> <span class="op">&amp;</span> <span class="va">Potential.contaminant</span> <span class="op">!=</span><span class="st">"+"</span> <span class="co">## filter upon Reverse and contaminant</span></span>
<span>                      <span class="op">)</span></span>
<span><span class="co">#&gt; filter base on giving formula</span></span>
<span><span class="co">### Or set a unique.peptide threshold.</span></span>
<span><span class="va">filter_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">filter_pg</span>,</span>
<span>                       filter_formula <span class="op">=</span> <span class="op">~</span> <span class="va">Unique.peptides</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; filter base on giving formula</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/get_df_wide.html">get_df_wide</a></span><span class="op">(</span><span class="va">se_pg</span><span class="op">)</span><span class="op">$</span><span class="va">Reverse</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">table</span></span>
<span><span class="co">#&gt; .</span></span>
<span><span class="co">#&gt;           + </span></span>
<span><span class="co">#&gt; 10378   340</span></span>
<span><span class="fu"><a href="../reference/get_df_wide.html">get_df_wide</a></span><span class="op">(</span><span class="va">filter_pg</span><span class="op">)</span><span class="op">$</span><span class="va">Reverse</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">table</span></span>
<span><span class="co">#&gt; .</span></span>
<span><span class="co">#&gt;      </span></span>
<span><span class="co">#&gt; 8982</span></span>
<span></span>
<span><span class="op">(</span><span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span><span class="va">se_pg</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Identification overlap before filter"</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> </span>
<span>  <span class="op">(</span><span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span><span class="va">filter_pg</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Identification overlap after before filter"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/filter%20pg-1.png" width="700"></p>
<ol start="4" style="list-style-type: decimal">
<li><strong>Normalization</strong></li>
</ol>
<p>In the DEP and DEP2 workflow, assay is log2-transformed during SE construction. And proteingroup assay is further normalized using variance stabilizing transformation(vsn).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">norm_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_vsn.html">normalize_vsn</a></span><span class="op">(</span><span class="va">filter_pg</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_normalization.html">plot_normalization</a></span><span class="op">(</span><span class="va">filter_pg</span>, <span class="va">norm_pg</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<ol start="5" style="list-style-type: decimal">
<li><strong>imputation</strong></li>
</ol>
<p>Even after filtering, a considerable proportion of MVs may still remain in the assay.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_missval.html">plot_missval</a></span><span class="op">(</span><span class="va">filter_pg</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/missing%20heatmap-1.png" width="700"></p>
<p>MVs can be classified as missing at random (MAR), or missing not at random (MNAR) based on the frequency and intensity of the NA signals. The <code><a href="../reference/impute.html">DEP2::impute</a></code> function provides kinds of imputation methods for SE objects. In addition to the methods Inheriting from DEP, <code>impute</code> also includes imputation functions from the <em><a href="https://bioconductor.org/packages/3.17/MSnbase" class="external-link">MSnbase</a></em> and <em><a href="https://bioconductor.org/packages/3.17/MsCoreUtils" class="external-link">MsCoreUtils</a></em> packages. These methods include:</p>
<p>– Left-censored imputation methods, replace: <em>“MinDet,” “MinProb,” “min”(LOD), “QRILC,” “zero,” “man”(impute by a left-shifted distribution)</em></p>
<p>– Local similarity methods: <em>“knn,” “nbavg”</em></p>
<p>– Global-structure methods: <em>“MLE,” “BPCA”</em></p>
<p>For more details on imputation, please refer to the vignettes about imputation in the DEP and MSnbase packages. Additionally, DEP2 includes two machine learning based imputation in <code>impute</code>: <em>“RF”</em> and <em>“GSimp”</em>. “RF” uses the <code><a href="https://rdrr.io/pkg/missForest/man/missForest.html" class="external-link">missForest::missForest</a></code> function, which train a random forest based on observed parts of dataset. And “GSimp” is a Gibbs sampler based left-censored imputation method proposed by <a href="http:s//doi.org/10.1371/journal.%20pcbi.1005973">Runmin Wei, Jingye Wang, etc</a>. These two methods may take a long time on iteration but are reported that they could draw a better estimation on missing values.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">35</span><span class="op">)</span></span>
<span><span class="va">sample_rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">norm_pg</span><span class="op">)</span>, <span class="fl">300</span><span class="op">)</span></span>
<span><span class="va">norm_pg_sample</span> <span class="op">=</span> <span class="va">norm_pg</span><span class="op">[</span><span class="va">sample_rows</span>,<span class="op">]</span> <span class="co"># random sample 150 features to reduce runing time</span></span>
<span><span class="fu"><a href="../reference/plot_detect.html">plot_detect</a></span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/impute-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Impute missing data using random draws from a Gaussian distribution centered around a minimal value (for MNAR)</span></span>
<span><span class="va">imp_pg_MinProb</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"MinProb"</span>, q <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.31744</span></span>
<span></span>
<span><span class="co">## Impute missing data using k-nearest neighbour approach</span></span>
<span><span class="va">imp_pg_knn</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Impute missing data using missForest</span></span>
<span><span class="va">imp_pg_RF</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"RF"</span>, ntree <span class="op">=</span> <span class="fl">50</span>, mtry <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## Impute missing data using Gibbs</span></span>
<span><span class="va">imp_pg_GSimp</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"GSimp"</span>, hi_q <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                       iters_each<span class="op">=</span><span class="fl">40</span>, iters_all<span class="op">=</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; Iteration 1 start...end!</span></span>
<span><span class="co">#&gt; Iteration 2 start...end!</span></span>
<span><span class="co">#&gt; Iteration 3 start...end!</span></span>
<span><span class="co">#&gt; Iteration 4 start...end!</span></span>
<span><span class="co">#&gt; Iteration 5 start...end!</span></span>
<span><span class="co">#&gt; Iteration 6 start...end!</span></span>
<span><span class="co">#&gt; Iteration 7 start...end!</span></span>
<span><span class="co">#&gt; Iteration 8 start...end!</span></span>
<span></span>
<span><span class="co">## Impute missing data using QRILC</span></span>
<span><span class="va">imp_pg_QRILC</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"QRILC"</span><span class="op">)</span> </span></code></pre></div>
<p>The distribution of mean intensity of proteingroups with or without MVs</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_detect.html">plot_detect</a></span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/unnamed-chunk-2-1.png" width="700"> The difference among imputation methods.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">NAs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co">## the imputed values by different methods.</span></span>
<span><span class="va">imps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"GSimp"</span> <span class="op">=</span> <span class="va">imp_pg_GSimp</span>, <span class="st">"QRILC"</span> <span class="op">=</span> <span class="va">imp_pg_QRILC</span>, <span class="st">"MinProb"</span> <span class="op">=</span> <span class="va">imp_pg_MinProb</span>, <span class="st">"RF"</span> <span class="op">=</span> <span class="va">imp_pg_RF</span>, <span class="st">"knn"</span> <span class="op">=</span> <span class="va">imp_pg_knn</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">gather</span><span class="op">(</span><span class="st">"label"</span>, <span class="st">"value"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"label"</span>,<span class="st">"condition"</span><span class="op">)</span><span class="op">]</span>,copy <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu">magrittr</span><span class="fu">::</span><span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">extract</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">NAs</span><span class="op">)</span>,<span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html" class="external-link">rbindlist</a></span><span class="op">(</span>idcol <span class="op">=</span> <span class="st">"method"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## the original normalized values without imputation</span></span>
<span><span class="va">nonimps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">gather</span><span class="op">(</span><span class="st">"label"</span>, <span class="st">"value"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"label"</span>,<span class="st">"condition"</span><span class="op">)</span><span class="op">]</span>,copy <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">magrittr</span><span class="fu">::</span><span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">extract</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">NAs</span><span class="op">)</span>,<span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"non_impute"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">method</span>,<span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/ggridges/" class="external-link">ggridges</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">imps</span>, <span class="va">nonimps</span><span class="op">)</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span>,y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">method</span>,level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">method</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html" class="external-link">geom_density_ridges</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#027ad450"</span>, scale <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>                      jittered_points <span class="op">=</span> <span class="cn">TRUE</span>,position <span class="op">=</span> <span class="fu"><a href="https://wilkelab.org/ggridges/reference/position_points_jitter.html" class="external-link">position_points_jitter</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>                      point_shape <span class="op">=</span> <span class="st">'|'</span>, point_size <span class="op">=</span> <span class="fl">2</span>, point_alpha <span class="op">=</span> <span class="fl">1</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Impute method"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Log2 value"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">9</span>,<span class="fl">39</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_DEP1.html">theme_DEP1</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/compare%20imp-1.png" width="700"></p>
<p>Left-censored approaches such as MinProb and GSimp replace MVs with conservatively low values, while the local similarity method “knn” and the global data learning method “RF” impute relatively larger values. These two kinds of methods cater to MNAR and MAR, respectively. However, “knn” will impute a mean value when the row contains more than 50% missing values.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"norm_pg_sample"</span>, <span class="st">"imp_pg_knn"</span>,<span class="st">"imp_pg_MinProb"</span>,<span class="st">"imp_pg_RF"</span>,<span class="st">"imp_pg_GSimp"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Impute missing data using q-th quantile for following analysis</span></span>
<span><span class="va">imp_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg</span>, fun <span class="op">=</span> <span class="st">"MinDet"</span> ,q <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">#&gt; Imputing along margin 2 (samples/columns).</span></span></code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li><strong>Hypothesis testing</strong></li>
</ol>
<p>The <code>test_diff</code> function performs a moderated t-test using limma. In DEP, p-values are corrected, using the fdrtool to classify significantly regulated/enriched candidates and stable proteins. DEP2 additionally provides two alternative frequently-used FDR control methods: “Benjamini-Hochberg fdr” and “Storey’s qvalue.”</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Test every sample versus PBS control</span></span>
<span><span class="va">diff_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_pg</span>, type <span class="op">=</span> <span class="st">"control"</span>, control <span class="op">=</span> <span class="st">"PBS"</span>, fdr.type <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Tested contrasts: W10_vs_PBS, W2_vs_PBS, W4_vs_PBS, W6_vs_PBS, W9_vs_PBS</span></span>
<span><span class="co">#&gt; BH</span></span>
<span><span class="co">## Test on manual contrasts</span></span>
<span><span class="va">diff_pg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_pg</span>, type <span class="op">=</span> <span class="st">"manual"</span>, test  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W4_vs_PBS"</span><span class="op">)</span>, fdr.type <span class="op">=</span> <span class="st">"Storey's qvalue"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Tested contrasts: W4_vs_PBS</span></span>
<span><span class="co">#&gt; Storey's qvalue</span></span></code></pre></div>
<p>Function <code>add_rejections</code> can classify significant hits according L2FC (<em>lfc</em>) and adjusted p value (<em>alpha</em>) threshold.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Add significant rejections for features, based on </span></span>
<span><span class="va">dep_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_pg</span>, alpha <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get the significant subset</span></span>
<span><span class="va">dep_pg_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dep_pg_sig</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 249</span></span></code></pre></div>
<p>Use the <code>plot_volcano</code> function to quickly visualize the results.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### volcano plot on contrast "W4_vs_PBS"</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pg</span>, contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, adjusted <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/rejections%20plot%201-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## plot the cutoff line</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pg</span>, contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, adjusted <span class="op">=</span> <span class="cn">F</span>,</span>
<span>             add_threshold_line <span class="op">=</span> <span class="st">"intersect"</span>, pCutoff  <span class="op">=</span> <span class="fl">0.05</span>, fcCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; add intersect threshold line. fcCutoff = 1pCutoff = 0.05</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/rejections%20plot%201-2.png" width="700"></p>
<p>In addition to the intersect method, <code>add_rejections</code> can draw a curve cutoff approach described by <a href="https://www.sciencedirect.com/science/article/pii/S1535947620316674" class="external-link">Eva C.Keilhauer</a>. Instead of intersect straight lines, the approach uses curve lines with <span class="math inline">\(y &gt; c/(x - x_0)\)</span>, where <span class="math inline">\(x\)</span> is the log2 FC, <span class="math inline">\(y\)</span> is the adjusted p values of features. The parameters <em>c</em> and <span class="math inline">\(x_0\)</span> represent the curvature and minimum L2FC, respectively. The value of <span class="math inline">\(x_0\)</span> is determined using the standard deviation <span class="math inline">\(\sigma\)</span> by <span class="math inline">\(x_0 = x_0.fold*\sigma\)</span>. <span class="math inline">\(\sigma\)</span> is the standard deviation of the Gaussian curve distribution of log2 fold changes in each contrast.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## thresholdmethod = "curve"</span></span>
<span><span class="va">dep_pg_curve</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_pg</span>, thresholdmethod <span class="op">=</span> <span class="st">"curve"</span>, curvature  <span class="op">=</span> <span class="fl">2</span>, x0_fold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## the cutoff line on volcano</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pg_curve</span>, contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, add_threshold_line <span class="op">=</span> <span class="st">"curve"</span>, curvature <span class="op">=</span> <span class="fl">2</span>, x0_fold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pg_curve</span>, contrast <span class="op">=</span> <span class="st">"W6_vs_PBS"</span>, add_threshold_line <span class="op">=</span> <span class="st">"curve"</span>, curvature <span class="op">=</span> <span class="fl">2</span>, x0_fold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; add curve threshold line. Sigma = 0.416241386089189, x0 = 0.832482772178379, curvature = 2</span></span>
<span><span class="co">#&gt; add curve threshold line. Sigma = 0.478600344109222, x0 = 0.957200688218444, curvature = 2</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/rejections2-1.png" width="700"></p>
<p>You can check the <span class="math inline">\(\sigma\)</span> and the fitted Gaussian curve via the <code>plot_diff_hist</code> function.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Check the fitted Gaussian curve</span></span>
<span><span class="fu"><a href="../reference/plot_diff_hist.html">plot_diff_hist</a></span><span class="op">(</span><span class="va">dep_pg_curve</span>, contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W4_vs_PBS"</span>, <span class="st">"W6_vs_PBS"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Get the fit result </span></span>
<span><span class="fu"><a href="../reference/plot_diff_hist.html">plot_diff_hist</a></span><span class="op">(</span><span class="va">dep_pg_curve</span>, plot <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="co">## a table of gaussian args \sigma and \mu</span></span>
<span><span class="co">#&gt;          contrast         mu    sigma2</span></span>
<span><span class="co">#&gt; 1 W10_vs_PBS_diff 0.01374325 0.2407074</span></span>
<span><span class="co">#&gt; 2  W2_vs_PBS_diff 0.02327110 0.2266282</span></span>
<span><span class="co">#&gt; 3  W4_vs_PBS_diff 0.02511490 0.1732569</span></span>
<span><span class="co">#&gt; 4  W6_vs_PBS_diff 0.02201209 0.2290583</span></span>
<span><span class="co">#&gt; 5  W9_vs_PBS_diff 0.02354730 0.2595546</span></span></code></pre></div>
<p>The <code>plot_heatmap</code> function can plot a heatmap for significant candidates.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/heatmap-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Reorder columns by condition</span></span>
<span><span class="va">dep_pg</span> <span class="op">=</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/Order_cols.html">Order_cols</a></span><span class="op">(</span><span class="va">dep_pg</span>,order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBS"</span>,<span class="st">"W2"</span>,<span class="st">"W4"</span>,<span class="st">"W6"</span>,<span class="st">"W9"</span>,<span class="st">"W10"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pg</span>, cluster_columns <span class="op">=</span> <span class="cn">F</span>, kmeans <span class="op">=</span> <span class="cn">T</span>, k <span class="op">=</span> <span class="fl">5</span>, seed <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># cluster features</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/heatmap-2.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Only plot the clusters that are up-regulated in treatment groups.</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pg</span>, cluster_columns <span class="op">=</span> <span class="cn">F</span>, kmeans <span class="op">=</span> <span class="cn">T</span>, k <span class="op">=</span> <span class="fl">5</span>, seed <span class="op">=</span> <span class="fl">1</span>, col_limit <span class="op">=</span> <span class="fl">4</span>,</span>
<span>             split_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span></span>
<span>             <span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/heatmap-3.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## plot on select contrast</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pg</span>, manual_contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/heatmap-4.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="protein-expressionenrichment-analysis-pipeline-with-peptide-protein-aggregation-">Protein Expression/Enrichment Analysis Pipeline with Peptide-protein Aggregation.<a class="anchor" aria-label="anchor" href="#protein-expressionenrichment-analysis-pipeline-with-peptide-protein-aggregation-"></a>
</h3>
<p>The second pipeline involves aggregating protein abundance from peptide quantity data. The analysis workflow differs from the previous one in terms of data processing steps.</p>
<p><img src="figs/DEP2workflow2.png" title="Fig. 2 the pipeline for protein re-aggregation"></p>
<ol style="list-style-type: decimal">
<li><strong>Loading Data</strong></li>
</ol>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Silicosis_peptide"</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li><strong>Converting QFeatures Object</strong></li>
</ol>
<p>The <code>make_pe</code>/<code>make_pe_parse</code> functions can be used to convert a QFeatures object. These functions have a similar usage pattern as the <code>make_se</code> series functions.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Here we only select two groups to save computational cost in these vignette</span></span>
<span><span class="op">(</span><span class="va">ecols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Intensity.PBS|Intensity.W4"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Silicosis_peptide</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] "Intensity.PBS_1" "Intensity.PBS_2" "Intensity.PBS_3" "Intensity.PBS_4"</span></span>
<span><span class="co">#&gt; [5] "Intensity.W4_2"  "Intensity.W4_3"  "Intensity.W4_4"  "Intensity.W4_5"</span></span>
<span></span>
<span><span class="va">pe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_pe_parse.html">make_pe_parse</a></span><span class="op">(</span><span class="va">Silicosis_peptide</span>, columns <span class="op">=</span> <span class="va">ecols</span>, mode <span class="op">=</span> <span class="st">"delim"</span><span class="op">)</span></span>
<span><span class="va">pe</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">#&gt;  [1] peptideRaw: SummarizedExperiment with 143528 rows and 8 columns</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li><strong>Filtering</strong></li>
</ol>
<p>The QFeatures objects can be filtered using <code>filter_pe</code>. Its usage is similar to function <code>filter_se</code> and can filter out unexpected features based on missing ratio or identification information in certain rowdata columns.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filt_pe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_pe.html">filter_pe</a></span><span class="op">(</span><span class="va">pe</span>, </span>
<span>                     thr <span class="op">=</span> <span class="fl">1</span>,fraction <span class="op">=</span> <span class="fl">0.4</span>,  <span class="co">## filter based on missing values</span></span>
<span>                     filter_formula <span class="op">=</span> <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">'+'</span> <span class="op">&amp;</span> <span class="va">Potential.contaminant</span> <span class="op">!=</span><span class="st">"+"</span> <span class="op">)</span></span>
<span><span class="co">#&gt; filter base on missing number is &lt;= 1 in at least one condition.</span></span>
<span><span class="co">#&gt; filter base on missing number fraction &lt; 0.4 in each row</span></span>
<span><span class="co">#&gt; filter base on giving formula</span></span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li><strong>Imputation</strong></li>
</ol>
<p>Imputation could be performed before normalization in order to reduce the potential effect caused by MVs. For instance, “quantiles.robust” normalization cannot work on assay with MVs. However, it is also acceptable to normalize data before replacing MVs, such as using “vsn” followed by “knn” imputation.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Use QFeatures::addAssay() and DEP2::impute</span></span>
<span><span class="va">pe</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">filt_pe</span>,</span>
<span>                              <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">filt_pe</span><span class="op">[[</span><span class="st">"peptideRaw"</span><span class="op">]</span><span class="op">]</span>, fun <span class="op">=</span> <span class="st">"QRILC"</span><span class="op">)</span>, </span>
<span>                              name <span class="op">=</span> <span class="st">"peptideImp"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_imputation.html">plot_imputation</a></span><span class="op">(</span><span class="va">pe</span><span class="op">[[</span><span class="st">"peptideRaw"</span><span class="op">]</span><span class="op">]</span>, <span class="va">pe</span><span class="op">[[</span><span class="st">"peptideImp"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/imp%20pe-1.png" width="700"></p>
<ol start="5" style="list-style-type: decimal">
<li><strong>Normalization</strong></li>
</ol>
<p>The <code>normalize_pe</code> function perform data normalization . Four normalization methods are available: <em>“diff.median,” “quantiles,” “quantiles.robust” ,“vsn”</em>. Notice that <em>quantiles.robust</em> cannot handle data table with MVs (ie NA), so it is necessary to impute in advance.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pe</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">:::</span><span class="fu"><a href="../reference/normalize_pe.html">normalize_pe</a></span><span class="op">(</span><span class="va">pe</span>,method <span class="op">=</span> <span class="st">"quantiles.robust"</span>,</span>
<span>                               i <span class="op">=</span> <span class="st">"peptideImp"</span>,    <span class="co">## nomalize on which assay</span></span>
<span>                               name <span class="op">=</span> <span class="st">"peptideNorm"</span> <span class="co">## output assay name</span></span>
<span>                               <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_normalization.html">plot_normalization</a></span><span class="op">(</span><span class="va">pe</span><span class="op">[[</span><span class="st">"peptideRaw"</span><span class="op">]</span><span class="op">]</span>,<span class="va">pe</span><span class="op">[[</span><span class="st">"peptideImp"</span><span class="op">]</span><span class="op">]</span>,<span class="va">pe</span><span class="op">[[</span><span class="st">"peptideNorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/imp%20before%20norm2-1.png" width="700"></p>
<ol start="6" style="list-style-type: decimal">
<li><strong>Aggregation</strong></li>
</ol>
<p>DEP2 offers the <code>aggregate_pe</code> function based on <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">QFeatures::aggregateFeatures</a></code>. The summarisation approach <em>aggrefun</em> can be “RobustSummary” (also called MSqRobSum) <span class="citation">(Sticker et al. 2020)</span>, “medianPolish,” “totalMean.” Details can be seen in <code>?aggregateFeatures</code>. And proteingroups can be summed by “Unique + Razor” peptides or only “Unique” peptides. <code>aggregate_pe</code> require several minutes depending on the peptide assay scale.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">begin_time</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pe</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/aggregate_pe.html">aggregate_pe</a></span><span class="op">(</span><span class="va">pe</span>,</span>
<span>                         aggrefun <span class="op">=</span> <span class="st">"RobustSummary"</span>,</span>
<span>                         aggregate_Peptide_Type <span class="op">=</span> <span class="st">"Unique + Razor"</span>,</span>
<span>                         fcol <span class="op">=</span> <span class="st">"Proteins"</span>,                  <span class="co"># the protein ID column in input table</span></span>
<span>                         i <span class="op">=</span> <span class="st">"peptideNorm"</span>, <span class="co"># the assay to aggregate</span></span>
<span>                         reserve <span class="op">=</span> <span class="st">"Gene.names"</span>              <span class="co"># reserve the Gene.names in row data.</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "aggregate by Unique + Razor peptides"</span></span>
<span><span class="co">#&gt; [1] "peptides distribution finished"</span></span>
<span><span class="co">#&gt; [1] "begin assay aggregate"</span></span>
<span><span class="co">#&gt; [1] "begin reducedataframe"</span></span>
<span><span class="co">#&gt; [1] "reducedataframe finiched"</span></span>
<span><span class="co">#&gt; [1] "assay aggregate finished"</span></span>
<span><span class="co">#&gt; [1] "aggregation finished"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">begin_time</span><span class="op">)</span> <span class="co"># aggregate cost several minutes depent on the assay scale</span></span>
<span><span class="co">#&gt; Time difference of 1.524174 mins</span></span>
<span></span>
<span><span class="va">pe</span> <span class="co"># protein result is in the protein assay</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 4 assays:</span></span>
<span><span class="co">#&gt;  [1] peptideRaw: SummarizedExperiment with 98219 rows and 8 columns </span></span>
<span><span class="co">#&gt;  [2] peptideImp: SummarizedExperiment with 98219 rows and 8 columns </span></span>
<span><span class="co">#&gt;  [3] peptideNorm: SummarizedExperiment with 98219 rows and 8 columns </span></span>
<span><span class="co">#&gt;  [4] protein: SummarizedExperiment with 9768 rows and 8 columns</span></span>
<span><span class="va">se_pep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pe2se.html">pe2se</a></span><span class="op">(</span><span class="va">pe</span>, names <span class="op">=</span> <span class="st">"Gene.names"</span>, ids <span class="op">=</span> <span class="st">"smallestProteingroups"</span><span class="op">)</span></span></code></pre></div>
<ol start="7" style="list-style-type: decimal">
<li><strong>Differential Test</strong></li>
</ol>
<p>The following differential test is the same as the proteingroups pipeline. Use <code>test_diff</code> and <code>add_rejections</code> to classify significant proteins.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Differential test</span></span>
<span><span class="va">diff_pep</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">se_pep</span>,type <span class="op">=</span> <span class="st">"control"</span>, control <span class="op">=</span> <span class="st">"PBS"</span>, fdr.type <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Tested contrasts: W4_vs_PBS</span></span>
<span><span class="co">#&gt; BH</span></span>
<span><span class="co">## Add rejections</span></span>
<span><span class="va">dep_pep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_pep</span>,alpha <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_pep</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">nrow</span></span>
<span><span class="co">#&gt; [1] 76</span></span></code></pre></div>
<ol start="8" style="list-style-type: decimal">
<li>Data visualization functions also work for results from re-aggregation pipeline.</li>
</ol>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Volcano</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pep</span>, contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, add_threshold_line <span class="op">=</span> <span class="st">"intersect"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; add intersect threshold line. fcCutoff = 1pCutoff = 0.05</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/DEP%20visualization-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Heatmap</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pep</span>, kmeans <span class="op">=</span> <span class="cn">T</span>,k <span class="op">=</span> <span class="fl">5</span>,col_limit <span class="op">=</span> <span class="fl">6</span><span class="op">)</span> </span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/DEP%20visualization-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="pipeline-for-post-translational-modification-ptm-specific-proteomics-">Pipeline for Post-Translational Modification (PTM)-Specific Proteomics.<a class="anchor" aria-label="anchor" href="#pipeline-for-post-translational-modification-ptm-specific-proteomics-"></a>
</h3>
<p>DEP2 has designed a new workflow for post-modification proteomics data analysis starting from modified peptides’ quantitative results. The post-translation modification information is necessary to distinguish modified sites, including position, modified residue (amino acid), gene name, or protein ID associated with the modification.</p>
<p><code>make_unique_ptm</code> creates (or overwrites) the PTM information columns, including <em>name</em>, <em>ID</em>, <em>gene_name</em>, <em>protein_ID</em>, <em>modified_aa</em>, <em>modified_pos</em>. The <em>name</em> and <em>ID</em> columns are the identifiers of each modified-peptide, following such naming rule: ’(gene name/protein ID)_(modified amino acid)(position of modification)’ (e.g., “TBCA_K51” and “O75347_K51”).</p>
<ol style="list-style-type: decimal">
<li><strong>Loading Data</strong></li>
</ol>
<p>The example we used here is the phosphorylated peptide result, table <strong>Phospho (STY)Sites.txt</strong>, from MaxQuant.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## phosphorylated peptides table of the silicosis mouse model.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Silicosis_phos</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## Format the modification information and generated modified-peptides identifier.</span></span>
<span><span class="co">## aa and pos is the modified amino acids and modified site in protein.</span></span>
<span><span class="va">unique_pho</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_unique_ptm.html">make_unique_ptm</a></span><span class="op">(</span><span class="va">Silicosis_phos</span>, gene_name <span class="op">=</span> <span class="st">"Gene.names"</span>, </span>
<span>                              protein_ID <span class="op">=</span> <span class="st">"Protein"</span>, aa <span class="op">=</span> <span class="st">"Amino.acid"</span>,</span>
<span>                              pos <span class="op">=</span> <span class="st">"Position"</span><span class="op">)</span> </span></code></pre></div>
<p>The <em>name</em> and <em>ID</em> are generated upon PTM information columns.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">unique_pho</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span>,<span class="st">"ID"</span>,<span class="st">"Gene.names"</span>,<span class="st">"Protein"</span>,<span class="st">"Amino.acid"</span>,<span class="st">"Position"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>              options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">7</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-17466358a46d1399716e" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-17466358a46d1399716e">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7"],["P0DPB4_S60","P0DPB4_S61","P0DPB4_S62","P0DPB4_S63","P0DPB4_S67","Ccdc173_S151","Tbc1d25_S560"],["P0DPB4_S60","P0DPB4_S61","P0DPB4_S62","P0DPB4_S63","P0DPB4_S67","A0JLY1_S151","A1A5B6_S560"],["","","","","","Ccdc173","Tbc1d25"],["P0DPB4","P0DPB4","P0DPB4","P0DPB4","P0DPB4","A0JLY1","A1A5B6"],["S","S","S","S","S","S","S"],[60,61,62,63,67,151,560]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>ID<\/th>\n      <th>Gene.names<\/th>\n      <th>Protein<\/th>\n      <th>Amino.acid<\/th>\n      <th>Position<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":7,"columnDefs":[{"className":"dt-right","targets":6},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[7,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><ol start="2" style="list-style-type: decimal">
<li><strong>Coverting SE</strong></li>
</ol>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Take 'Intensity' columns.</span></span>
<span><span class="va">ecols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Intensity."</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">unique_pho</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Construct a SE object mentioned before.</span></span>
<span><span class="va">se_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_se_parse.html">make_se_parse</a></span><span class="op">(</span><span class="va">unique_pho</span>, columns <span class="op">=</span> <span class="va">ecols</span>, </span>
<span>                        mode <span class="op">=</span> <span class="st">"delim"</span>, sep <span class="op">=</span> <span class="st">"_"</span>, remove_prefix <span class="op">=</span> <span class="cn">T</span>, log2transform <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li><strong>Filtering</strong></li>
</ol>
<p>The following steps are basically similar to the protomics analysis pipeline. <code>filter_se</code> is a universal filter function for SE objects. An additional threshold should be set on ‘Localization.prob’ in the MaxQuant modification result.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Filter base on both missing occupancy and the localization probability for this site.</span></span>
<span><span class="va">filt_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">se_ptm</span>, </span>
<span>                      thr <span class="op">=</span> <span class="fl">1</span>, fraction <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                      filter_formula <span class="op">=</span> <span class="op">~</span> <span class="va">Reverse</span><span class="op">!=</span><span class="st">"+"</span> <span class="op">&amp;</span> </span>
<span>                        <span class="va">Potential.contaminant</span><span class="op">!=</span><span class="st">"+"</span> <span class="op">&amp;</span> </span>
<span>                        <span class="va">Localization.prob</span><span class="op">&gt;</span><span class="fl">0.7</span></span>
<span>                      <span class="op">)</span></span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li><strong>Normalization</strong></li>
</ol>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## VSN normalization</span></span>
<span><span class="va">norm_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_vsn.html">normalize_vsn</a></span><span class="op">(</span><span class="va">filt_ptm</span><span class="op">)</span></span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li><strong>Imputation</strong></li>
</ol>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## KNN impute relatively larger values</span></span>
<span><span class="va">imp_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">filt_ptm</span>, fun<span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Cluster size 11729 broken into 6954 4775 </span></span>
<span><span class="co">#&gt; Cluster size 6954 broken into 2918 4036 </span></span>
<span><span class="co">#&gt; Cluster size 2918 broken into 2328 590 </span></span>
<span><span class="co">#&gt; Cluster size 2328 broken into 1841 487 </span></span>
<span><span class="co">#&gt; Cluster size 1841 broken into 1661 180 </span></span>
<span><span class="co">#&gt; Cluster size 1661 broken into 1404 257 </span></span>
<span><span class="co">#&gt; Done cluster 1404 </span></span>
<span><span class="co">#&gt; Done cluster 257 </span></span>
<span><span class="co">#&gt; Done cluster 1661 </span></span>
<span><span class="co">#&gt; Done cluster 180 </span></span>
<span><span class="co">#&gt; Done cluster 1841 </span></span>
<span><span class="co">#&gt; Done cluster 487 </span></span>
<span><span class="co">#&gt; Done cluster 2328 </span></span>
<span><span class="co">#&gt; Done cluster 590 </span></span>
<span><span class="co">#&gt; Done cluster 2918 </span></span>
<span><span class="co">#&gt; Cluster size 4036 broken into 56 3980 </span></span>
<span><span class="co">#&gt; Done cluster 56 </span></span>
<span><span class="co">#&gt; Cluster size 3980 broken into 2552 1428 </span></span>
<span><span class="co">#&gt; Cluster size 2552 broken into 2523 29 </span></span>
<span><span class="co">#&gt; Cluster size 2523 broken into 468 2055 </span></span>
<span><span class="co">#&gt; Done cluster 468 </span></span>
<span><span class="co">#&gt; Cluster size 2055 broken into 1044 1011 </span></span>
<span><span class="co">#&gt; Done cluster 1044 </span></span>
<span><span class="co">#&gt; Done cluster 1011 </span></span>
<span><span class="co">#&gt; Done cluster 2055 </span></span>
<span><span class="co">#&gt; Done cluster 2523 </span></span>
<span><span class="co">#&gt; Done cluster 29 </span></span>
<span><span class="co">#&gt; Done cluster 2552 </span></span>
<span><span class="co">#&gt; Done cluster 1428 </span></span>
<span><span class="co">#&gt; Done cluster 3980 </span></span>
<span><span class="co">#&gt; Done cluster 4036 </span></span>
<span><span class="co">#&gt; Done cluster 6954 </span></span>
<span><span class="co">#&gt; Cluster size 4775 broken into 2517 2258 </span></span>
<span><span class="co">#&gt; Cluster size 2517 broken into 1379 1138 </span></span>
<span><span class="co">#&gt; Done cluster 1379 </span></span>
<span><span class="co">#&gt; Done cluster 1138 </span></span>
<span><span class="co">#&gt; Done cluster 2517 </span></span>
<span><span class="co">#&gt; Cluster size 2258 broken into 861 1397 </span></span>
<span><span class="co">#&gt; Done cluster 861 </span></span>
<span><span class="co">#&gt; Done cluster 1397 </span></span>
<span><span class="co">#&gt; Done cluster 2258 </span></span>
<span><span class="co">#&gt; Done cluster 4775</span></span></code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li><strong>Differantail testing</strong></li>
</ol>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diff_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_ptm</span>, type <span class="op">=</span> <span class="st">"manual"</span>, test <span class="op">=</span> <span class="st">"PBS_vs_W6"</span> , fdr.type <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Tested contrasts: PBS_vs_W6</span></span>
<span><span class="co">#&gt; BH</span></span>
<span></span>
<span><span class="va">dep_ptm</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_ptm</span>, alpha <span class="op">=</span> <span class="fl">0.05</span>, lfc <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_ptm</span>,adjusted <span class="op">=</span> <span class="cn">T</span>, add_threshold_line <span class="op">=</span> <span class="st">"intersect"</span><span class="op">)</span></span>
<span><span class="co">#&gt; add intersect threshold line. fcCutoff = 1pCutoff = 0.05</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="additional-analysis-pipeline-for-rna-seq-counts">Additional Analysis Pipeline for RNA-seq Counts<a class="anchor" aria-label="anchor" href="#additional-analysis-pipeline-for-rna-seq-counts"></a>
</h3>
<p>Multi-omics studies have already become a widespread method in research. In particular, proteomics-transcriptomics combined studies provide insights into regulatory mechanisms from multiple perspectives. Therefore, DEP2 packages a pipeline for RNA-seq data from DESeq2.</p>
<ol style="list-style-type: decimal">
<li><strong>Loading Data</strong></li>
</ol>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Silicosis_counts</span><span class="op">)</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Silicosis_counts</span><span class="op">)</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-8228581aa3e4de6c04d3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8228581aa3e4de6c04d3">{"x":{"filter":"none","vertical":false,"data":[["ENSMUSG00000000001","ENSMUSG00000000028","ENSMUSG00000000031","ENSMUSG00000000049","ENSMUSG00000000056","ENSMUSG00000000058"],[133,26,0,88,31,211],[235,28,0,71,55,435],[174,25,38,66,74,215],[418,26,4,58,117,940],[388,16,0,60,96,571],[201,19,1,69,80,252],[153,22,0,68,53,460],[208,16,0,72,78,446],[200,17,0,63,26,405],[350,24,0,57,68,609],[147,22,0,61,45,389],[143,35,1,102,49,153],[318,22,0,40,73,446],[385,27,0,60,72,503],[171,27,1,68,43,155],[319,32,0,75,46,403],[185,5,0,39,49,287],[208,16,0,45,67,324]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>PBS_1<\/th>\n      <th>PBS_2<\/th>\n      <th>PBS_3<\/th>\n      <th>W2_1<\/th>\n      <th>W2_2<\/th>\n      <th>W2_3<\/th>\n      <th>W2_4<\/th>\n      <th>W4_1<\/th>\n      <th>W4_2<\/th>\n      <th>W4_3<\/th>\n      <th>W4_4<\/th>\n      <th>W6_1<\/th>\n      <th>W6_2<\/th>\n      <th>W6_3<\/th>\n      <th>W6_4<\/th>\n      <th>W10_1<\/th>\n      <th>W10_2<\/th>\n      <th>W10_3<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":6,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[6,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><ol start="2" style="list-style-type: decimal">
<li><strong>Constructing DESeqDataSet</strong></li>
</ol>
<p>The <code>make_dds</code> function can be used to convert a DESeqDataSet object. Its usage is similar to function <code>make_se</code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_dds_parse.html">make_dds_parse</a></span><span class="op">(</span><span class="va">Silicosis_counts</span>,mode <span class="op">=</span> <span class="st">"delim"</span><span class="op">)</span></span>
<span><span class="va">dds</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 24936 18 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(24936): ENSMUSG00000000001 ENSMUSG00000000028 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000118633 ENSMUSG00000118636</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(18): PBS_1 PBS_2 ... W10_2 W10_3</span></span>
<span><span class="co">#&gt; colData names(4): label ID condition replicate</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li><strong>Filtering</strong></li>
</ol>
<p>The <em>DESeqDataSet</em> object inherits from <em>SummarizedExperiment</em>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">inherits</a></span><span class="op">(</span><span class="va">dds</span>,<span class="st">"SummarizedExperiment"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p><code>Filter_se</code> can also be used to filter <em>DESeqDataSet</em> objects. The <em>rowsum_threshold</em> parameter sets a threshold on the row sum of counts.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">dds</span>, fraction <span class="op">=</span> <span class="fl">0.3</span>, thr <span class="op">=</span> <span class="fl">1</span>, rowsum_threshold <span class="op">=</span> <span class="fl">35</span><span class="op">)</span></span>
<span><span class="co">#&gt; filter base on missing number is &lt;= 1 in at least one condition.</span></span>
<span><span class="co">#&gt; filter base on missing number fraction &lt; 0.3 in each row</span></span>
<span><span class="co">#&gt; filter base on counts sum &gt;=35</span></span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li><strong>Transforming ID</strong></li>
</ol>
<p>In this case, the gene identifier is Ensembl ID. To conveniently check the genes, DEP2 can transform the IDs to gene symbols using the <a href="%22http://www.bioconductor.org/packages/release/BiocViews.html#___OrgDb%22" class="external-link">OrgDb</a> databases. The <code>ID_transform</code> function performs the ID conversion and replaces the rownames of the object. This function requires the species annotation database. You can execute the <code>check_organismDB_depends</code> function to verify and install the necessary package if it is not already installed.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Check if the annotation package has been already installed. Or install it if not yet.</span></span>
<span><span class="fu"><a href="../reference/check_organismDB_depends.html">check_organismDB_depends</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"Mouse"</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ENSMUSG00000000001" "ENSMUSG00000000028" "ENSMUSG00000000049"</span></span>
<span><span class="co">#&gt; [4] "ENSMUSG00000000056"</span></span>
<span></span>
<span><span class="co">## Transform ID</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/ID_transform.html">ID_transform</a></span><span class="op">(</span><span class="va">dds</span>, species <span class="op">=</span> <span class="st">"Mouse"</span><span class="op">)</span></span>
<span><span class="co">#&gt; org.Mm.eg.db</span></span>
<span><span class="co">#&gt; Loading required package: org.Mm.eg.db</span></span>
<span><span class="co">#&gt; Loading required package: AnnotationDbi</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'AnnotationDbi'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     select</span></span>
<span><span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Gnai3" "Cdc45" "Apoh"  "Narf"</span></span></code></pre></div>
<p>If needed, <code>ID_transform</code> can be used on proteinomics data, for example:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/ID_transform.html">ID_transform</a></span><span class="op">(</span><span class="va">dep_pg</span>, </span>
<span>                   species <span class="op">=</span> <span class="st">"Mouse"</span>,</span>
<span>                   from_columns <span class="op">=</span> <span class="st">"ID"</span>,</span>
<span>                   fromtype <span class="op">=</span> <span class="st">"UNIPROT"</span><span class="op">)</span></span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li><strong>Differential Test</strong></li>
</ol>
<p>The differential test steps are similar to the foregoing pipelines, but the <code>test_diff</code> is replaced by <code>test_diff_deg</code> to call <em>DESeq2</em>.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff_deg.html">test_diff_deg</a></span><span class="op">(</span><span class="va">dds</span>, type <span class="op">=</span> <span class="st">"control"</span>, control <span class="op">=</span> <span class="st">"PBS"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Performing DESeq analysis...</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span><span class="co">#&gt; Formatting analysis result...</span></span>
<span><span class="co">#&gt; extracted results: W2_vs_PBS, W4_vs_PBS, W6_vs_PBS, W10_vs_PBS</span></span>
<span><span class="co">#&gt; Calulate a rlog &amp; log2 transform assay, stored in rlg or ntd slot...</span></span>
<span></span>
<span><span class="va">deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff</span>, alpha <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="export-result-table">Export result table<a class="anchor" aria-label="anchor" href="#export-result-table"></a>
</h2>
<p>The function <code>get_results</code> can be used to obtain the result table or significant subset from differential test results.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get result table</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_results.html">get_results</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span><span class="op">)</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-a15b670239e029179d2a" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-a15b670239e029179d2a">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["Acp5","Adamts8","Add2","Aif1","Alox12","Amot"],["Q05117","P57110","Q9QYB8","O70200","P39655","Q8VHG2"],[1.78379863582781e-05,0.1517820521554423,0.0004596542827077434,4.938487998517713e-06,0.003948759345619383,0.02398830811650167],[0.0006786145418469344,0.02086813409472511,0.005256802471013192,8.969555896156061e-08,0.004408082070174645,0.2018580084258191],[0.005296976886465571,0.01143005955272048,0.0001057429676689482,1.69833441208854e-07,0.005314785186328786,0.1635292142797664],[0.0002481157144911807,0.0005789581618138493,0.001027817216055859,2.990391531353203e-06,0.0001779822712837013,0.0004619206759408735],[0.0003753998030738659,0.005888276344182553,0.004540516744869069,0.0001504059546568457,0.005984008020722463,0.0004546908018991406],[0.00157,0.309,0.00928,0.000726,0.0302,0.0926],[0.00616,0.0643,0.0246,2.52e-05,0.0217,0.336],[0.0325,0.0536,0.00296,6.97e-05,0.0325,0.314],[0.00318,0.00523,0.00772,0.000236,0.00255,0.00454],[0.00637,0.0329,0.0279,0.00389,0.0332,0.00701],[true,false,true,true,false,false],[false,false,false,true,false,false],[false,false,true,true,false,false],[false,true,true,true,true,true],[false,false,false,true,false,true],[true,true,true,true,true,true],[2.26,1.81,-2.93,3.27,-1.8,-1.56],[1.29,2.52,-1.76,3.64,-1.45,-0.68],[0.993,2.8,-2.78,3.47,-1.41,-0.746],[1.44,4.19,-2.18,2.78,-2.12,-2.23],[1.69,3.81,-2.2,2.41,-1.7,-2.74],[-1.14,-2.46,1.86,-2.55,1.35,1.16],[0.152,0.0511,0.0965,1.09,-0.104,0.482],[-0.148,0.339,-0.922,0.926,-0.0649,0.416],[0.3,1.73,-0.324,0.236,-0.773,-1.07],[0.549,1.35,-0.345,-0.141,-0.352,-1.58],[1.12,-0.654,-1.07,0.725,-0.457,-0.397]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>ID<\/th>\n      <th>W10_vs_PBS_p.val<\/th>\n      <th>W2_vs_PBS_p.val<\/th>\n      <th>W4_vs_PBS_p.val<\/th>\n      <th>W6_vs_PBS_p.val<\/th>\n      <th>W9_vs_PBS_p.val<\/th>\n      <th>W10_vs_PBS_p.adj<\/th>\n      <th>W2_vs_PBS_p.adj<\/th>\n      <th>W4_vs_PBS_p.adj<\/th>\n      <th>W6_vs_PBS_p.adj<\/th>\n      <th>W9_vs_PBS_p.adj<\/th>\n      <th>W10_vs_PBS_significant<\/th>\n      <th>W2_vs_PBS_significant<\/th>\n      <th>W4_vs_PBS_significant<\/th>\n      <th>W6_vs_PBS_significant<\/th>\n      <th>W9_vs_PBS_significant<\/th>\n      <th>significant<\/th>\n      <th>W10_vs_PBS_ratio<\/th>\n      <th>W2_vs_PBS_ratio<\/th>\n      <th>W4_vs_PBS_ratio<\/th>\n      <th>W6_vs_PBS_ratio<\/th>\n      <th>W9_vs_PBS_ratio<\/th>\n      <th>PBS_centered<\/th>\n      <th>W2_centered<\/th>\n      <th>W4_centered<\/th>\n      <th>W6_centered<\/th>\n      <th>W9_centered<\/th>\n      <th>W10_centered<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7,8,9,10,11,12,19,20,21,22,23,24,25,26,27,28,29]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the significant subset</span></span>
<span><span class="va">sig_deg</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_results.html">get_results</a></span><span class="op">(</span><span class="va">deg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sig_deg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "data.frame"</span></span></code></pre></div>
<p>The function <code>get_df_wide</code> can exctract the full table from a DEP result.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get full data set with row information.</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_df_wide.html">get_df_wide</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span><span class="op">)</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-8e9539872343d5916c5b" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8e9539872343d5916c5b">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["1810009A15Rik","4931429I11Rik","4932418E24Rik","9230110C19Rik","A0A0B4J1G0","A2AG58"],[24.95510952747318,28.69966575907819,27.78597613451763,28.15230611255565,24.95510952747318,28.34921186155433],[27.66446313277172,24.90632973314363,27.95046763390967,27.59884207400334,24.90632973314363,27.51433458244034],[28.46692687314581,29.75345569880781,28.11693013585196,27.18847503022147,25.11995634956691,24.94520456810298],[24.87779382073681,29.96019811153612,28.63124489682529,27.67764610476691,24.87779382073681,28.71010472444583],[24.59815719850685,30.34897639809095,27.9113100466282,27.81968310362362,23.79512516174751,28.63404866067978],[27.74649123545927,29.50435460172689,27.05926876597585,27.37618775275533,27.14457007219492,28.43662920528327],[24.59855824923066,29.03745659840769,27.28385112220987,27.44227555462211,26.21347411972538,28.81399331813932],[24.60246293663941,29.55508561103058,27.0881235846591,27.5904867230838,26.60636297173695,28.30984947351396],[24.60910108962483,28.67996014045045,28.3227277883364,27.83833713496194,26.01545157680024,27.82602353667694],[24.76401383039086,29.31208789612677,28.56487248772729,28.19627931447179,24.73976684423218,28.02198092544246],[24.72379447459864,29.16383857004443,28.0466557799332,28.05545031983613,26.51179248573432,27.61190456908198],[24.76621809154749,29.10072037627216,28.61236805230781,27.53592252131995,26.05774406351986,24.76621809154749],[24.74224678468332,28.38582576803209,28.12602863526298,28.11525871788245,25.89617303018159,27.98555297927077],[27.66022869496513,24.73034010339481,27.52800847674121,27.49192515482403,26.76816717590122,27.7840122152021],[24.70596642966679,24.70596642966679,27.60678331263885,27.90206382574908,26.72837483142786,27.89928083738521],[24.70681025443308,24.70681025443308,27.8273154268193,28.09878039767533,26.84367255330415,27.62197250374319],[27.4926503039888,27.19527660033994,28.25112134844306,28.01384848940292,26.52100396672763,28.38436755398033],[24.72592236451652,24.72592236451652,27.56291512794439,27.10069186358927,27.00508160456531,24.72592236451652],[28.45209876668179,24.79271041184634,27.92594222194998,27.80579021119705,27.32595535083978,28.01033953664414],[24.95840135969208,28.22987599143941,28.32068787046042,28.01779494292224,26.73761872706053,27.44317988795794],["Q9D937","E9Q793","Q8C5V8","Q8CC70","A0A0B4J1G0","A2AG58"],["Q9D937","E9Q793","Q8C5V8","Q8CC70","A0A0B4J1G0","A2AG58"],["2","7","15","14","3","10"],["2","7","15","14","3","10"],["2","7","15","14","3","10"],["","","","","","Uncharacterized protein CXorf23 homolog"],["1810009A15Rik","4931429I11Rik","4932418E24Rik","9230110C19Rik","",""],["&gt;Q9D937 Uncharacterized protein C11orf98 homolog OS=Mus musculus OX=10090 PE=1 SV=1","&gt;E9Q793 Jhy protein OS=Mus musculus OX=10090 GN=Jhy PE=1 SV=1","&gt;Q8C5V8 Coiled-coil domain-containing protein 187 OS=Mus musculus OX=10090 GN=Ccdc187 PE=2 SV=1","&gt;Q8CC70 Cilia- and flagella-associated protein 300 OS=Mus musculus OX=10090 GN=Cfap300 PE=2 SV=1","&gt;A0A0B4J1G0 Low affinity immunoglobulin gamma Fc region receptor IV OS=Mus musculus OX=10090 GN=Fcgr4 PE=1 SV=1","&gt;A2AG58 BCLAF1 and THRAP3 family member 3 OS=Mus musculus OX=10090 GN=Bclaf3 PE=1 SV=1"],[1,1,1,1,1,1],[2,7,15,14,3,10],[2,7,15,14,3,10],[2,7,15,14,3,10],[12094000,770730,4431600,14583000,1371700,986190],[35112000,31396,5686400,14419000,0,2576800],[88801000,600530,7911100,13432000,2541900,925390],[42952000,73060000,7356500,10295000,214960,1450600],[114290000,830230,10941000,18398000,15586000,70833000],[31462000,44453000,10549000,21712000,7950500,35955000],[125700000,101360000,11432000,27832000,1476700,1985800],[82430000,64003000,5231400,16129000,14318000,4812900],[148790000,52267000,6075000,16620000,8884100,2590500],[145950000,48484000,4720600,22499000,9379500,3013700],[117910000,1762100,19217000,36050000,11780000,1383500],[138720000,924580,19552000,29538000,3277000,1824300],[71657000,1383500,14223000,17628000,11987000,1132000],[66123000,2082500,16121000,18354000,7910500,870670],[7558300,1271700,13346000,19766000,9806000,1906800],[58457000,380510,5071000,18147000,11549000,3721400],[17117000,391260,4442200,19754000,56719000,1907400],[44816000,741790,6194200,27357000,13334000,1398100],[51917000,907870,13517000,30529000,9592600,73530000],[78577000,222640,8740800,23216000,17502000,37570000],["","","","","",""],["","","","","",""],["","","","","",""],["Q9D937","E9Q793","Q8C5V8","Q8CC70","A0A0B4J1G0","A2AG58"],[true,true,false,false,true,true],[14,6,0,0,3,3],[-2.450112202734607,-4.582590788915726,-0.5946238975248396,-0.3250150963716357,0.6750967761487709,-1.833379082417945],[2.878465652044715,0.9453525409186154,0.5989445893829374,0.839965589717223,3.458882586291288,2.527470638748258],[0.2141767246550543,-1.818619123998555,0.002160345929048901,0.2574752466727936,2.066989681220029,0.3470457781651568],[0.9277723892135309,0.3479860382030775,0.9960442499403876,0.5423889122044883,0.03990137214030112,0.8479255145248927],[0.8669331621764824,0.1823226782212962,0.9939771571759072,0.3629814908997082,0.006183779783934442,0.7404007804964077],[0.1702328429251978,-1.393351104895615,0.007665815276543024,0.936050225760967,3.144741027081098,0.3370528778610138],[-3.280038733390086,-0.9752174375133853,-1.27278861473953,-0.5727604203848151,-0.1613900743871755,-0.6113932133677036],[1.070726866244428,3.538329390858585,-0.2982440260762502,0.3784423266535746,2.11156152162931,2.949225673904129],[-1.104655933572829,1.2815559766726,-0.7855163204078899,-0.09715904686562027,0.975085723621067,1.168916230268213],[0.4414383658531839,0.3861695635811948,0.01847640585596512,0.776465442063774,0.1827662795778464,0.313966994659415],[0.2979674738531838,0.2464203185489729,0.003499038784346099,0.6710003074258533,0.08776118501661673,0.1831998462491643],[-1.075334893625276,1.202546898382799,-3.413782856952039,-0.4326060061115676,1.816916653125031,1.390401707396107],[-3.950674266808679,-1.522533994103966,-0.2217709675316122,-0.22342138125858,-0.2700844131667184,-2.103491597084576],[0.4000913328258342,2.991012834268004,0.7527736211316671,0.7277813657798098,2.002867182849767,1.457127290187257],[-1.775291466991423,0.7342394200820195,0.2655013268000275,0.2521799922606149,0.8663913848415241,-0.3231821534486592],[0.2322828230993858,0.6525726889124638,0.4301141722939688,0.4422652120749901,0.2627019796930473,0.8119742187344641],[0.102952340097824,0.5005818110228095,0.2652897477742804,0.2778559999709607,0.1257354498664451,0.7056636329816552],[-1.728169652460587,0.6889728996329185,1.153844744380757,1.122845301518842,1.614382096988188,-0.384418494943697],[-3.212643097412048,-4.954450100945723,-0.8363930317422017,-0.2279116798733063,0.4578237419653375,-1.337318743871473],[1.138122502222465,-0.4409032725737534,0.1381515569210777,0.7232910671650834,2.730775337981822,2.22330014340036],[-1.037260297594791,-2.697676686759738,-0.349120737410562,0.2476896936458886,1.59429953997358,0.4429906997644437],[0.4670842478126998,0.06716249331746038,0.26680020730579,0.4259460971112384,0.03485608048853896,0.721665905547428],[0.3274581951098386,0.02206596724335622,0.1484599683939366,0.2861929076003478,0.008913874068378312,0.605378687059535],[-1.009728149622327,-2.53136249336948,-1.517247137220219,1.102851999684049,2.970722792956487,0.5269282856751417],[-3.046075931668875,-5.133284508130365,-0.810920705536314,-0.6795374969351783,0.4063525228450837,-3.004993835470557],[2.282501923110447,0.3946588217039766,0.3826477813714629,0.4854431891536805,3.1901383329876,1.355855885695646],[-0.3817870042792144,-2.369312843213194,-0.2141364620824255,-0.09704715389074892,1.798245427916342,-0.8245689748874554],[0.8601094958437993,0.204297661979755,0.6252750048385585,0.8336518622839095,0.05986936843642167,0.6049972892014515],[0.7654036072455452,0.08802404273676817,0.4582536833607932,0.728772480811986,0.01452408748863981,0.4348113597674632],[-0.3034535486291534,-1.815269961901653,-0.7598461617760552,-0.352814538417893,2.735870539320585,-0.8008261833643773],[false,false,false,false,false,false],[false,false,false,false,false,false],[false,false,false,false,false,false],[false,false,false,false,false,false],[false,false,false,false,false,false],[false,false,false,false,false,false]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>PBS_1<\/th>\n      <th>PBS_2<\/th>\n      <th>PBS_3<\/th>\n      <th>PBS_4<\/th>\n      <th>W2_1<\/th>\n      <th>W2_3<\/th>\n      <th>W2_4<\/th>\n      <th>W2_5<\/th>\n      <th>W4_2<\/th>\n      <th>W4_3<\/th>\n      <th>W4_4<\/th>\n      <th>W4_5<\/th>\n      <th>W6_2<\/th>\n      <th>W6_3<\/th>\n      <th>W6_4<\/th>\n      <th>W6_6<\/th>\n      <th>W9_2<\/th>\n      <th>W9_4<\/th>\n      <th>W10_2<\/th>\n      <th>W10_4<\/th>\n      <th>Protein.IDs<\/th>\n      <th>Majority.protein.IDs<\/th>\n      <th>Peptide.counts..all.<\/th>\n      <th>Peptide.counts..razor.unique.<\/th>\n      <th>Peptide.counts..unique.<\/th>\n      <th>Protein.names<\/th>\n      <th>Gene.names<\/th>\n      <th>Fasta.headers<\/th>\n      <th>Number.of.proteins<\/th>\n      <th>Peptides<\/th>\n      <th>Razor...unique.peptides<\/th>\n      <th>Unique.peptides<\/th>\n      <th>iBAQ.PBS_1<\/th>\n      <th>iBAQ.PBS_2<\/th>\n      <th>iBAQ.PBS_3<\/th>\n      <th>iBAQ.PBS_4<\/th>\n      <th>iBAQ.W10_2<\/th>\n      <th>iBAQ.W10_4<\/th>\n      <th>iBAQ.W2_1<\/th>\n      <th>iBAQ.W2_3<\/th>\n      <th>iBAQ.W2_4<\/th>\n      <th>iBAQ.W2_5<\/th>\n      <th>iBAQ.W4_2<\/th>\n      <th>iBAQ.W4_3<\/th>\n      <th>iBAQ.W4_4<\/th>\n      <th>iBAQ.W4_5<\/th>\n      <th>iBAQ.W6_2<\/th>\n      <th>iBAQ.W6_3<\/th>\n      <th>iBAQ.W6_4<\/th>\n      <th>iBAQ.W6_6<\/th>\n      <th>iBAQ.W9_2<\/th>\n      <th>iBAQ.W9_4<\/th>\n      <th>Only.identified.by.site<\/th>\n      <th>Reverse<\/th>\n      <th>Potential.contaminant<\/th>\n      <th>ID<\/th>\n      <th>imputed<\/th>\n      <th>num_NAs<\/th>\n      <th>W10_vs_PBS_CI.L<\/th>\n      <th>W10_vs_PBS_CI.R<\/th>\n      <th>W10_vs_PBS_diff<\/th>\n      <th>W10_vs_PBS_p.adj<\/th>\n      <th>W10_vs_PBS_p.val<\/th>\n      <th>W10_vs_PBS_t.stastic<\/th>\n      <th>W2_vs_PBS_CI.L<\/th>\n      <th>W2_vs_PBS_CI.R<\/th>\n      <th>W2_vs_PBS_diff<\/th>\n      <th>W2_vs_PBS_p.adj<\/th>\n      <th>W2_vs_PBS_p.val<\/th>\n      <th>W2_vs_PBS_t.stastic<\/th>\n      <th>W4_vs_PBS_CI.L<\/th>\n      <th>W4_vs_PBS_CI.R<\/th>\n      <th>W4_vs_PBS_diff<\/th>\n      <th>W4_vs_PBS_p.adj<\/th>\n      <th>W4_vs_PBS_p.val<\/th>\n      <th>W4_vs_PBS_t.stastic<\/th>\n      <th>W6_vs_PBS_CI.L<\/th>\n      <th>W6_vs_PBS_CI.R<\/th>\n      <th>W6_vs_PBS_diff<\/th>\n      <th>W6_vs_PBS_p.adj<\/th>\n      <th>W6_vs_PBS_p.val<\/th>\n      <th>W6_vs_PBS_t.stastic<\/th>\n      <th>W9_vs_PBS_CI.L<\/th>\n      <th>W9_vs_PBS_CI.R<\/th>\n      <th>W9_vs_PBS_diff<\/th>\n      <th>W9_vs_PBS_p.adj<\/th>\n      <th>W9_vs_PBS_p.val<\/th>\n      <th>W9_vs_PBS_t.stastic<\/th>\n      <th>W10_vs_PBS_significant<\/th>\n      <th>W2_vs_PBS_significant<\/th>\n      <th>W4_vs_PBS_significant<\/th>\n      <th>W6_vs_PBS_significant<\/th>\n      <th>W9_vs_PBS_significant<\/th>\n      <th>significant<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>Function <code>get_signicant</code> can get results of significant candidates.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get full data set of significant features</span></span>
<span><span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_pg</span>,return_type <span class="op">=</span> <span class="st">"table"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">ends_with</a></span><span class="op">(</span><span class="st">"significant"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">summary</span></span>
<span><span class="co">#&gt;  W10_vs_PBS_significant W2_vs_PBS_significant W4_vs_PBS_significant</span></span>
<span><span class="co">#&gt;  Mode :logical          Mode :logical         Mode :logical        </span></span>
<span><span class="co">#&gt;  FALSE:152              FALSE:140             FALSE:178            </span></span>
<span><span class="co">#&gt;  TRUE :97               TRUE :109             TRUE :71             </span></span>
<span><span class="co">#&gt;  W6_vs_PBS_significant W9_vs_PBS_significant significant   </span></span>
<span><span class="co">#&gt;  Mode :logical         Mode :logical         Mode:logical  </span></span>
<span><span class="co">#&gt;  FALSE:114             FALSE:141             TRUE:249      </span></span>
<span><span class="co">#&gt;  TRUE :135             TRUE :108</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="result-comparation">Result comparation<a class="anchor" aria-label="anchor" href="#result-comparation"></a>
</h2>
<p>After identifying significant candidates through add_rejections, <code>plot_multi_venn</code> and <code>plot_multi_heatmap</code> can directly compare the significant result from multiple omics experiments via name identifiers.</p>
<p>Use <code>plot_multi_venn</code> to compare the differences between two or more pipelines, such as the analysis results from MQ proteingroups or the peptides re-aggregation pipeline.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Test on the contrast W4_vs_PBS.</span></span>
<span><span class="va">diff_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_pg</span>, type <span class="op">=</span> <span class="st">"manual"</span>, test  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W4_vs_PBS"</span><span class="op">)</span>, fdr.type <span class="op">=</span> <span class="st">"Storey's qvalue"</span><span class="op">)</span></span>
<span><span class="va">dep_pg</span> <span class="op">&lt;-</span> <span class="va">diff_pg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.05</span>,lfc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">diff_pep</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">se_pep</span>,type <span class="op">=</span> <span class="st">"control"</span>, control <span class="op">=</span> <span class="st">"PBS"</span>, fdr.type <span class="op">=</span> <span class="st">"Storey's qvalue"</span><span class="op">)</span></span>
<span><span class="va">dep_pep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_pep</span>,alpha <span class="op">=</span> <span class="fl">0.05</span>, lfc <span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Filter out pg quantified which only have one peptides.</span></span>
<span><span class="va">dep_pep</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">dep_pep</span>,filter_formula <span class="op">=</span> <span class="op">~</span><span class="va">.n</span>  <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">dep_pg</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">dep_pg</span>,filter_formula <span class="op">=</span> <span class="op">~</span><span class="va">Razor...unique.peptides</span>  <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot Venn to explore the difference between two ways.</span></span>
<span><span class="fu"><a href="../reference/plot_multi_venn.html">plot_multi_venn</a></span><span class="op">(</span>omics_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>PG<span class="op">=</span> <span class="va">dep_pg</span>, pep<span class="op">=</span><span class="va">dep_pep</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/comparation-1.png" width="700"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## If set the proteins remained in both result as background. </span></span>
<span><span class="co">## dep_pep only contain two conditions of the study. PBS and W4. See the pe_peptides above.</span></span>
<span><span class="va">overlap_protein</span> <span class="op">=</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dep_pep</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_multi_venn.html">plot_multi_venn</a></span><span class="op">(</span>omics_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>PG<span class="op">=</span> <span class="va">dep_pg</span>, pep<span class="op">=</span><span class="va">dep_pep</span><span class="op">)</span>, background <span class="op">=</span> <span class="va">overlap_protein</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/comparation2,-1.png" width="700"></p>
<p>Check the expression of interested proteins/genes accorss omics via <code>plot_multi_heatmap</code>. Here we take some proteins that may be related to IL-1 signaling, the signaling pathway that is reported to be upregulated in silicosis mouse model in the original paper.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Take the proteins may response to interleukin-1 as an example.</span></span>
<span><span class="va">IL1_relative_genes</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Irg1"</span>, <span class="st">"Il1rn"</span>, <span class="st">"Saa3"</span>, <span class="st">"Zbp1"</span>, <span class="st">"Ccl6"</span>, </span>
<span>                       <span class="st">"Serpine1"</span>, <span class="st">"Ccl21a"</span>, <span class="st">"Pycard"</span>, <span class="st">"Irak2"</span>, <span class="st">"Vrk2"</span>, </span>
<span>                       <span class="st">"Fn1"</span>, <span class="st">"Il1r1"</span>, <span class="st">"Irf1"</span>, <span class="st">"Ccl9"</span>, <span class="st">"Mapk11"</span>, <span class="st">"Tank"</span>, </span>
<span>                       <span class="st">"Mapk13"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Order the samples by time.</span></span>
<span><span class="va">dep_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Order_cols.html">Order_cols</a></span><span class="op">(</span><span class="va">dep_pg</span>, order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBS"</span>,<span class="st">"W2"</span>,<span class="st">"W4"</span>,<span class="st">"W6"</span>,<span class="st">"W9"</span>,<span class="st">"W10"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot the heatmap upon two DEP results.</span></span>
<span><span class="fu"><a href="../reference/plot_multi_heatmap.html">plot_multi_heatmap</a></span><span class="op">(</span>omics_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>PG<span class="op">=</span> <span class="va">dep_pg</span>, pep<span class="op">=</span><span class="va">dep_pep</span><span class="op">)</span>,</span>
<span>                   choose_name <span class="op">=</span> <span class="va">IL1_relative_genes</span>,</span>
<span>                   width <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/mulheatmap-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="post-analysis">Post analysis<a class="anchor" aria-label="anchor" href="#post-analysis"></a>
</h2>
<p>To reduce the barriers in format conversion between deferential test and downstream analysis, post analysis suites for differential testing results is available in DEP2. The post-analysis functionalities may not be fully completed upon default installation of DEP2. Users have the option to complete these functionalities by executing the “check_” series of functions.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Check required packages for enrichment analysis</span></span>
<span><span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/check_enrichment_depends.html">check_enrichment_depends</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Check required packages for PPI</span></span>
<span><span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/check_PPI_depends.html">check_PPI_depends</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>As an alternative approach (but one that we do not recommend), users can also choose to proactively specify the installation of all dependent packages when installing DEP2.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Install all depend packages at once by setting 'dependencies = TRUE'</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"mildpiggy/DEP2"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Besides, functional analyses (functional enrichment or PPI) require species annotation database packages in the R environment. Therefore, the relevant packages must be installed before conducting functional analysis for any organism. You can run the function <code>check_organismDB_depends</code> in the console to check or install the missing packages. Here is an example code:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Check organism annotation package of a species</span></span>
<span><span class="co">## Parameter organism could one of "Chicken", "Human", "Mouse", "Rat", etc.</span></span>
<span><span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/check_organismDB_depends.html">check_organismDB_depends</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"Human"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="biological-function-enrichment">Biological function enrichment<a class="anchor" aria-label="anchor" href="#biological-function-enrichment"></a>
</h3>
<p>Biological functional enrichment functions from <em><a href="https://bioconductor.org/packages/3.17/enrichR" class="external-link">enrichR</a></em>, <em><a href="https://bioconductor.org/packages/3.17/clusterProfiler" class="external-link">clusterProfiler</a></em> and <em><a href="https://bioconductor.org/packages/3.17/ReactomePA" class="external-link">ReactomePA</a></em> are integrated into <code>test_ORA</code> and <code>test_GSEA</code> functions in DEP2. These functions enable the enrichment analysis of differential testing results obtained from <code>test_diff</code> and <code>add_rejection</code> functions. The objective gene/protein list is extracted and transformed using organism annotation data. Subsequently, the genes are mapped to biological knowledge annotated gene sets. As we mentioned above, DEP2 used the orginsm annotation package to transform protein/gene IDs, therefore the</p>
<p><code>test_ORA</code> can perform enrichment analysis upon Gene ontology, KEGG or REACTOME data. And <code>test_GSEA</code> is designed for GSEA.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1. Extract gene list</span></span>
<span><span class="va">diff_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_pg</span>, type <span class="op">=</span> <span class="st">"manual"</span>, test  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W4_vs_PBS"</span><span class="op">)</span>, fdr.type <span class="op">=</span> <span class="st">"Storey's qvalue"</span><span class="op">)</span></span>
<span><span class="va">dep_pg</span> <span class="op">&lt;-</span> <span class="va">diff_pg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.01</span>,lfc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 2. Check install</span></span>
<span><span class="fu"><a href="../reference/check_enrichment_depends.html">check_enrichment_depends</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># check required packages</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="../reference/check_organismDB_depends.html">check_organismDB_depends</a></span><span class="op">(</span><span class="st">"Mouse"</span><span class="op">)</span> <span class="co"># organism annotation for mouse</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co">## Over representation analysis</span></span>
<span><span class="co"># On one contrast</span></span>
<span><span class="va">res_ora</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/test_ORA.html">test_ORA</a></span><span class="op">(</span><span class="va">dep_pg</span>, contrasts <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, species <span class="op">=</span> <span class="st">"Mouse"</span>,type <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 86 gene(s) transfrom to ENTREZID from SYMBOL. 2 gene(s) transfrom to ENTREZID from UNIPROT. 4 gene(s) transfrom to ENTREZID from ALIAS. </span></span>
<span><span class="co">#&gt; Star enrich GO terms by 92 ENTREZIDs.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">res_ora</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] "enrichResult"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "DOSE"</span></span>
<span></span>
<span><span class="co"># On each contrasts</span></span>
<span><span class="va">res_ora2</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/test_ORA.html">test_ORA</a></span><span class="op">(</span><span class="va">dep_pg</span>, by_contrast <span class="op">=</span> <span class="cn">T</span>, species <span class="op">=</span> <span class="st">"Mouse"</span>,type <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 86 gene(s) transfrom to ENTREZID from SYMBOL. 2 gene(s) transfrom to ENTREZID from UNIPROT. 4 gene(s) transfrom to ENTREZID from ALIAS. </span></span>
<span><span class="co">#&gt; Star enrich GO terms by 92 ENTREZIDs.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">res_ora2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "enrichResult"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "DOSE"</span></span>
<span></span>
<span><span class="co">## Gene set enrichment analysis</span></span>
<span><span class="va">res_gsea</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/test_GSEA.html">test_GSEA</a></span><span class="op">(</span><span class="va">dep_pg</span>, by_contrast <span class="op">=</span> <span class="cn">F</span>, species <span class="op">=</span> <span class="st">"Mouse"</span>,type <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 8301 gene(s) transfrom to ENTREZID from SYMBOL. 110 gene(s) transfrom to ENTREZID from UNIPROT. 510 gene(s) transfrom to ENTREZID from ALIAS. </span></span>
<span><span class="co">#&gt; 61 gene(s) do not match in database.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">res_gsea</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "gseaResult"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "DOSE"</span></span></code></pre></div>
<p>The output of <code>test_ORA</code> and <code>test_GSEA</code> is the same object as from source functions, which is seamlessly compatible for plot functions and manipulate function provided for <em>enrichR</em> result. The visualization methods from <em><a href="https://bioconductor.org/packages/3.17/enrichplot" class="external-link">enrichplot</a></em> package can be easily draw on output of these two output.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Dotplot function from DOSE package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">enrichplot</a></span><span class="op">)</span></span>
<span><span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">res_ora</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/dotplot-1.png" width="700"></p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## GSEA plot on one category</span></span>
<span><span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/gseaplot.html" class="external-link">gseaplot</a></span><span class="op">(</span><span class="va">res_gsea</span>,geneSetID <span class="op">=</span> <span class="st">"GO:0071347"</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/gesaplot-1.png" width="700"></p>
<p>The <strong>cnet</strong> plotted on the interested biological processes which enriched in GSEA.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Cnetplot</span></span>
<span><span class="va">res_gsea2</span> <span class="op">&lt;-</span> <span class="va">res_gsea</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:0071347"</span>,<span class="st">"GO:0035329"</span>,<span class="st">"GO:0032660"</span>,<span class="st">"GO:0090263"</span>,<span class="st">"GO:0035456"</span>,<span class="st">"GO:0038061"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/cnetplot.html" class="external-link">cnetplot</a></span><span class="op">(</span><span class="va">res_gsea2</span>,foldChange <span class="op">=</span> <span class="va">res_gsea</span><span class="op">@</span><span class="va">geneList</span>,node_label <span class="op">=</span> <span class="st">"category"</span>,showCategory <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">size</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">size</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/cnet-1.png" width="700"></p>
<p>More usages and details are available in Yu’s introduction to the <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">biomedical knowledge mining book</a>.</p>
</div>
<div class="section level3">
<h3 id="expression-pattern-cluster">Expression pattern cluster<a class="anchor" aria-label="anchor" href="#expression-pattern-cluster"></a>
</h3>
<p>In the time-course omics studies or the experiments with complex group design, identifying key candidates across the entire experiment from numerous contrasts can be challenging. DEP2 provides the <code>get_tc_cluster</code> function to perform time-course cluster for omics data with multiple groups. The function clusters features using the “cmeans” algorithm, and return a list that contains a heatmap plot and cluster result. In the omics experiement with multiple experimental groups, <code>get_tc_cluster</code> can classify genes/proteins according to their expression patterns.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Perform cluster on signicant protein in all contrast</span></span>
<span><span class="va">tc_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_tc_cluster.html">get_tc_cluster</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span>,</span>
<span>               group_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBS"</span>,<span class="st">"W2"</span>,<span class="st">"W4"</span>,<span class="st">"W6"</span>,<span class="st">"W9"</span>,<span class="st">"W10"</span><span class="op">)</span> <span class="co"># Set group order</span></span>
<span>              <span class="op">)</span></span>
<span><span class="co">#&gt; aabb[1] "start timeclust"</span></span>
<span><span class="co">#&gt; beging 'cm' cluster by dist = 'euclidean'</span></span>
<span><span class="co">#&gt; [1] "finish heatmap"</span></span>
<span><span class="co">## The time course heatmap</span></span>
<span><span class="va">tc_pg</span><span class="op">$</span><span class="va">ht</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/tcheatmap-1.png" width="700"></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Expression cluster on phospholated sites</span></span>
<span><span class="va">tc_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_tc_cluster.html">get_tc_cluster</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_ptm</span><span class="op">)</span>,</span>
<span>                         group_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBS"</span>,<span class="st">"W2"</span>,<span class="st">"W4"</span>,<span class="st">"W6"</span>,<span class="st">"W9"</span>,<span class="st">"W10"</span><span class="op">)</span>, <span class="co"># Set group order</span></span>
<span>                         row_font_size <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                         k <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                         heatmap_height <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; aabb[1] "start timeclust"</span></span>
<span><span class="co">#&gt; beging 'cm' cluster by dist = 'euclidean'</span></span>
<span><span class="co">#&gt; `use_raster` is automatically set to TRUE for a matrix with more than</span></span>
<span><span class="co">#&gt; 2000 rows. You can control `use_raster` argument by explicitly setting</span></span>
<span><span class="co">#&gt; TRUE/FALSE to it.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Set `ht_opt$message = FALSE` to turn off this message.</span></span>
<span><span class="co">#&gt; [1] "finish heatmap"</span></span>
<span><span class="va">tc_ptm</span><span class="op">$</span><span class="va">ht</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/tcheatmap2-1.png" width="700"></p>
<p>The output of <code>get_tc_cluster</code> contains a heatmap combined with expression line chart, as well as a cluster result table stored in the ‘res’ slot. The cluster table is a data.frame that records the closest cluster, membership values, and expression values.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The clustering result table</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tc_pg</span><span class="op">$</span><span class="va">res</span><span class="op">)</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-c403e1f7b162d7d6b0f4" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-c403e1f7b162d7d6b0f4">{"x":{"filter":"none","vertical":false,"data":[["Add2","Aif1","Ank1","Apob","Apobec1","Arpc3"],["Add2","Aif1","Ank1","Apob","Apobec1","Arpc3"],["5","3","5","5","2","6"],["0.850455964612496","0.611992854766843","0.961110863494774","0.347502423141861","0.721798530939003","0.421372841122647"],["1.4031651357608","-2.95383674846652","1.36561856239679","0.770023191760622","-2.91925275953339","-2.77728077765259"],["0.955760049771879","-3.00261654279607","1.60504582100852","0.810079354823003","0.491157269021848","-2.18424394899136"],["3.34674481426824","-2.96374170783671","2.51084415357867","1.97673886603011","-2.92915771890359","-3.03568763902474"],["1.72185719929221","-1.26382193700541","1.52969869041471","1.02185198065155","-2.99656846626976","-0.24208351404598"],["-1.17976238487416","0.660651522946292","-1.35390658995662","-0.42229789397453","0.49088888591632","-0.310026982206907"],["0.738692770143135","2.20399710532227","0.969056724811942","0.809063948714495","0.99036220283616","0.0859763791939514"],["0.454436133223705","0.939742783460883","0.0606094953455454","0.346875690433741","0.752896223811419","0.543920263841152"],["0.372676510007","0.565227804927591","0.318148140626498","-0.00945952340494216","0.778161667235409","0.19289203778893"],["-1.16881849375617","1.50287186837277","-0.978807836434243","-1.79624891613905","0.414429204987435","0.33614831875132"],["-1.01390575299015","0.181303703494141","-1.09898432359999","-1.561093004867","0.0572895705329231","0.437131617765196"],["-1.60913166393511","1.0735227411656","-0.258315345156461","-0.62407679395848","1.02823299315419","1.14464041625407"],["0.103275922578302","0.944632902827326","-0.350548727881865","-1.14106316271157","0.434183238146801","0.287487933472178"],["0.284319854715836","0.129000828585479","-0.328133110348389","-0.302432571850865","0.433874661970414","0.673625826011808"],["0.201384419502499","0.218937617140245","-0.358624221733791","0.0523036330229019","0.563265541721464","1.12260556842368"],["-0.709619846943678","0.545253592049175","-0.268285741513004","0.0763236995886132","0.721167115440821","0.925178593859627"],["-1.07110932894793","0.0512807120004162","-0.522983609461221","-0.195394817452616","0.761114051144901","0.87875747272075"],["0.361086838227603","-0.0534141775450685","-0.154670536404176","1.12588259814946","-0.139427477690688","0.120048466599862"],["-1.05199721886448","-0.228191835240938","-1.49875588067605","0.623258587908722","0.959393134688483","0.605097494531854"],["-1.31953673349061","0.934238476935956","-0.904369130554194","-0.489895702146789","0.120520896471994","0.512180334685659"],["-0.819518223688927","0.514961289662541","-0.282636534462654","-1.07043916457733","-0.0125302346831617","0.68363213802153"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>Timecourse_cluster<\/th>\n      <th>mebership<\/th>\n      <th>PBS_1<\/th>\n      <th>PBS_2<\/th>\n      <th>PBS_3<\/th>\n      <th>PBS_4<\/th>\n      <th>W2_1<\/th>\n      <th>W2_3<\/th>\n      <th>W2_4<\/th>\n      <th>W2_5<\/th>\n      <th>W4_2<\/th>\n      <th>W4_3<\/th>\n      <th>W4_4<\/th>\n      <th>W4_5<\/th>\n      <th>W6_2<\/th>\n      <th>W6_3<\/th>\n      <th>W6_4<\/th>\n      <th>W6_6<\/th>\n      <th>W9_2<\/th>\n      <th>W9_4<\/th>\n      <th>W10_2<\/th>\n      <th>W10_4<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":6,"columnDefs":[{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[6,10,25,50,100]}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="protein-protein-interaction">Protein-protein interaction<a class="anchor" aria-label="anchor" href="#protein-protein-interaction"></a>
</h3>
<p>DEP2 provides a function <code>test_PPI</code> to construct protein-protein interaction network based on STRING data. <code>test_PPI</code> performs PPI analysis locally. Notably, <code>test_PPI</code> use a local STRING data. The function <code>load_PPIdata</code> checks for the existence of a local database. If the local data is absent, <code>load_PPIdata</code> and <code>test_PPI</code> will try to download the PPI data of the species from website and save in <code>system.file("PPIdata", "species_name",package = "DEP2")</code>.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Check required packages for network construction.</span></span>
<span><span class="fu"><a href="../reference/check_PPI_depends.html">check_PPI_depends</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co"># 2. Load STRING data. </span></span>
<span><span class="co">## If local STRING is missing, load_PPIdata will download to system.file("PPIdata", "Mouse",package = "DEP2")</span></span>
<span><span class="fu"><a href="../reference/load_PPIdata.html">load_PPIdata</a></span><span class="op">(</span>speciesname <span class="op">=</span> <span class="st">"Mouse"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Input species is Mouse   The species ID is 10090</span></span>
<span><span class="co">#&gt; fristDownload</span></span>
<span><span class="co">#&gt; fristDownload</span></span>
<span><span class="co">#&gt; fristDownload</span></span>
<span><span class="co">#&gt; load succeed</span></span>
<span><span class="co">#&gt; [1] "Loading finished!"</span></span>
<span></span>
<span><span class="co"># 3.Load STRING data and perform PPI analysis. </span></span>
<span><span class="co">## It also automatically download data if loacl files is missing. </span></span>
<span><span class="va">PPI_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_PPI.html">test_PPI</a></span><span class="op">(</span></span>
<span>  <span class="va">dep_pg</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Mouse"</span>,</span>
<span>  STRING.version <span class="op">=</span> <span class="st">"11.5"</span>,</span>
<span>  score_cutoff <span class="op">=</span> <span class="fl">400</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 86 gene(s) transfrom to ENTREZID from SYMBOL. 2 gene(s) transfrom to ENTREZID from UNIPROT. 4 gene(s) transfrom to ENTREZID from ALIAS.</span></span>
<span></span>
<span><span class="co"># igraph network</span></span>
<span><span class="va">PPI_ig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PPInetwork.html">PPInetwork</a></span><span class="op">(</span><span class="va">PPI_res</span>,returntype <span class="op">=</span> <span class="st">"igraph"</span><span class="op">)</span></span>
<span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/plot.igraph.html" class="external-link">plot.igraph</a></span><span class="op">(</span><span class="va">PPI_ig</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_DEP2_analysis_files/figure-html/PPI1-1.png" width="700"></p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## visNetwork network</span></span>
<span><span class="fu"><a href="../reference/PPInetwork.html">PPInetwork</a></span><span class="op">(</span><span class="va">PPI_res</span>,returntype <span class="op">=</span> <span class="st">"visNetwork"</span><span class="op">)</span> </span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 18.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=zh_CN.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=zh_CN.UTF-8        LC_COLLATE=zh_CN.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=zh_CN.UTF-8    LC_MESSAGES=zh_CN.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=zh_CN.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=zh_CN.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] enrichplot_1.20.0           org.Mm.eg.db_3.17.0        </span></span>
<span><span class="co">#&gt;  [3] AnnotationDbi_1.62.2        doRNG_1.8.6                </span></span>
<span><span class="co">#&gt;  [5] rngtools_1.5.2              foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [7] randomForest_4.7-1.1        ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [9] patchwork_1.1.2             missForest_1.5             </span></span>
<span><span class="co">#&gt; [11] ggplot2_3.4.2               dplyr_1.1.2                </span></span>
<span><span class="co">#&gt; [13] tibble_3.2.1                DEP2_0.3.7.3               </span></span>
<span><span class="co">#&gt; [15] R6_2.5.1                    limma_3.56.2               </span></span>
<span><span class="co">#&gt; [17] MSnbase_2.26.0              ProtGenerics_1.32.0        </span></span>
<span><span class="co">#&gt; [19] mzR_2.34.1                  Rcpp_1.0.11                </span></span>
<span><span class="co">#&gt; [21] MsCoreUtils_1.12.0          SummarizedExperiment_1.30.2</span></span>
<span><span class="co">#&gt; [23] Biobase_2.60.0              GenomicRanges_1.52.0       </span></span>
<span><span class="co">#&gt; [25] GenomeInfoDb_1.36.1         IRanges_2.34.1             </span></span>
<span><span class="co">#&gt; [27] S4Vectors_0.38.1            BiocGenerics_0.46.0        </span></span>
<span><span class="co">#&gt; [29] MatrixGenerics_1.12.2       matrixStats_1.0.0          </span></span>
<span><span class="co">#&gt; [31] tidyr_1.3.0                 BiocStyle_2.28.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.2                    bitops_1.0-7               </span></span>
<span><span class="co">#&gt;   [3] sf_1.0-14                   HDO.db_0.99.1              </span></span>
<span><span class="co">#&gt;   [5] httr_1.4.6                  RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;   [7] doParallel_1.0.17           tools_4.3.1                </span></span>
<span><span class="co">#&gt;   [9] utf8_1.2.3                  DT_0.28                    </span></span>
<span><span class="co">#&gt;  [11] lazyeval_0.2.2              GetoptLong_1.0.5           </span></span>
<span><span class="co">#&gt;  [13] withr_2.5.0                 graphite_1.46.0            </span></span>
<span><span class="co">#&gt;  [15] gridExtra_2.3               preprocessCore_1.62.1      </span></span>
<span><span class="co">#&gt;  [17] fdrtool_1.2.17              cli_3.6.1                  </span></span>
<span><span class="co">#&gt;  [19] textshaping_0.3.6           scatterpie_0.2.1           </span></span>
<span><span class="co">#&gt;  [21] sandwich_3.0-2              labeling_0.4.2             </span></span>
<span><span class="co">#&gt;  [23] sass_0.4.6                  mvtnorm_1.2-2              </span></span>
<span><span class="co">#&gt;  [25] proxy_0.4-27                askpass_1.1                </span></span>
<span><span class="co">#&gt;  [27] pkgdown_2.0.7               yulab.utils_0.0.6          </span></span>
<span><span class="co">#&gt;  [29] Rsamtools_2.16.0            systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt;  [31] gson_0.1.0                  R.utils_2.12.2             </span></span>
<span><span class="co">#&gt;  [33] DOSE_3.26.1                 itertools_0.1-3            </span></span>
<span><span class="co">#&gt;  [35] rstudioapi_0.15.0           impute_1.74.1              </span></span>
<span><span class="co">#&gt;  [37] RSQLite_2.3.1               visNetwork_2.1.2           </span></span>
<span><span class="co">#&gt;  [39] gridGraphics_0.5-1          generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [41] ggVennDiagram_1.2.2         shape_1.4.6                </span></span>
<span><span class="co">#&gt;  [43] crosstalk_1.2.0             GO.db_3.17.0               </span></span>
<span><span class="co">#&gt;  [45] Matrix_1.5-4.1              MALDIquant_1.22.1          </span></span>
<span><span class="co">#&gt;  [47] fansi_1.0.4                 imputeLCMD_2.1             </span></span>
<span><span class="co">#&gt;  [49] abind_1.4-5                 R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt;  [51] lifecycle_1.0.3             yaml_2.3.7                 </span></span>
<span><span class="co">#&gt;  [53] edgeR_3.42.4                qvalue_2.32.0              </span></span>
<span><span class="co">#&gt;  [55] Rtsne_0.16                  blob_1.2.4                 </span></span>
<span><span class="co">#&gt;  [57] crayon_1.5.2                lattice_0.21-8             </span></span>
<span><span class="co">#&gt;  [59] msigdbr_7.5.1               cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [61] KEGGREST_1.40.0             magick_2.7.4               </span></span>
<span><span class="co">#&gt;  [63] pillar_1.9.0                knitr_1.43                 </span></span>
<span><span class="co">#&gt;  [65] ComplexHeatmap_2.16.0       fgsea_1.26.0               </span></span>
<span><span class="co">#&gt;  [67] rjson_0.2.21                codetools_0.2-19           </span></span>
<span><span class="co">#&gt;  [69] fastmatch_1.1-3             glue_1.6.2                 </span></span>
<span><span class="co">#&gt;  [71] ggfun_0.1.1                 downloader_0.4             </span></span>
<span><span class="co">#&gt;  [73] pcaMethods_1.92.0           data.table_1.14.8          </span></span>
<span><span class="co">#&gt;  [75] MultiAssayExperiment_1.26.0 treeio_1.24.1              </span></span>
<span><span class="co">#&gt;  [77] vctrs_0.6.3                 png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [79] gtable_0.3.3                assertthat_0.2.1           </span></span>
<span><span class="co">#&gt;  [81] cachem_1.0.8                xfun_0.39                  </span></span>
<span><span class="co">#&gt;  [83] S4Arrays_1.0.4              tidygraph_1.2.3            </span></span>
<span><span class="co">#&gt;  [85] RVenn_1.1.0                 survival_3.5-5             </span></span>
<span><span class="co">#&gt;  [87] ncdf4_1.21                  iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [89] gmm_1.8                     units_0.8-2                </span></span>
<span><span class="co">#&gt;  [91] ellipsis_0.3.2              nlme_3.1-162               </span></span>
<span><span class="co">#&gt;  [93] ggtree_3.8.0                bit64_4.0.5                </span></span>
<span><span class="co">#&gt;  [95] rprojroot_2.0.3             bslib_0.5.0                </span></span>
<span><span class="co">#&gt;  [97] affyio_1.70.0               tmvtnorm_1.5               </span></span>
<span><span class="co">#&gt;  [99] KernSmooth_2.23-21          colorspace_2.1-0           </span></span>
<span><span class="co">#&gt; [101] DBI_1.1.3                   DESeq2_1.40.2              </span></span>
<span><span class="co">#&gt; [103] tidyselect_1.2.0            bit_4.0.5                  </span></span>
<span><span class="co">#&gt; [105] compiler_4.3.1              graph_1.78.0               </span></span>
<span><span class="co">#&gt; [107] glmnet_4.1-7                desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [109] DelayedArray_0.26.6         shadowtext_0.1.2           </span></span>
<span><span class="co">#&gt; [111] bookdown_0.34               scales_1.2.1               </span></span>
<span><span class="co">#&gt; [113] classInt_0.4-9              affy_1.78.1                </span></span>
<span><span class="co">#&gt; [115] rappdirs_0.3.3              stringr_1.5.0              </span></span>
<span><span class="co">#&gt; [117] digest_0.6.33               rmarkdown_2.23             </span></span>
<span><span class="co">#&gt; [119] XVector_0.40.0              TCseq_1.23.0               </span></span>
<span><span class="co">#&gt; [121] htmltools_0.5.5             pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [123] umap_0.2.10.0               dbplyr_2.3.3               </span></span>
<span><span class="co">#&gt; [125] highr_0.10                  fastmap_1.1.1              </span></span>
<span><span class="co">#&gt; [127] rlang_1.1.1                 GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt; [129] htmlwidgets_1.6.2           farver_2.1.1               </span></span>
<span><span class="co">#&gt; [131] jquerylib_0.1.4             zoo_1.8-12                 </span></span>
<span><span class="co">#&gt; [133] jsonlite_1.8.7              BiocParallel_1.34.2        </span></span>
<span><span class="co">#&gt; [135] mzID_1.38.0                 R.oo_1.25.0                </span></span>
<span><span class="co">#&gt; [137] GOSemSim_2.26.1             RCurl_1.98-1.12            </span></span>
<span><span class="co">#&gt; [139] magrittr_2.0.3              ggplotify_0.1.1            </span></span>
<span><span class="co">#&gt; [141] GenomeInfoDbData_1.2.10     munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [143] ggnewscale_0.4.9            babelgene_22.9             </span></span>
<span><span class="co">#&gt; [145] ape_5.7-1                   viridis_0.6.3              </span></span>
<span><span class="co">#&gt; [147] reticulate_1.30             vsn_3.68.0                 </span></span>
<span><span class="co">#&gt; [149] stringi_1.7.12              ggraph_2.1.0               </span></span>
<span><span class="co">#&gt; [151] zlibbioc_1.46.0             MASS_7.3-60                </span></span>
<span><span class="co">#&gt; [153] org.Hs.eg.db_3.17.0         plyr_1.8.8                 </span></span>
<span><span class="co">#&gt; [155] parallel_4.3.1              ggrepel_0.9.3              </span></span>
<span><span class="co">#&gt; [157] Biostrings_2.68.1           graphlayouts_1.0.0         </span></span>
<span><span class="co">#&gt; [159] splines_4.3.1               circlize_0.4.15            </span></span>
<span><span class="co">#&gt; [161] locfit_1.5-9.8              igraph_1.5.0               </span></span>
<span><span class="co">#&gt; [163] QFeatures_1.10.0            clusterProfiler.dplyr_0.0.2</span></span>
<span><span class="co">#&gt; [165] reshape2_1.4.4              XML_3.99-0.14              </span></span>
<span><span class="co">#&gt; [167] evaluate_0.21               BiocManager_1.30.21        </span></span>
<span><span class="co">#&gt; [169] tweenr_2.0.2                openssl_2.0.6              </span></span>
<span><span class="co">#&gt; [171] purrr_1.0.1                 polyclip_1.10-4            </span></span>
<span><span class="co">#&gt; [173] clue_0.3-64                 norm_1.0-11.1              </span></span>
<span><span class="co">#&gt; [175] ReactomePA_1.44.0           BiocBaseUtils_1.2.0        </span></span>
<span><span class="co">#&gt; [177] ggforce_0.4.1               reactome.db_1.84.0         </span></span>
<span><span class="co">#&gt; [179] AnnotationFilter_1.24.0     tidytree_0.4.2             </span></span>
<span><span class="co">#&gt; [181] e1071_1.7-13                RSpectra_0.16-1            </span></span>
<span><span class="co">#&gt; [183] viridisLite_0.4.2           class_7.3-22               </span></span>
<span><span class="co">#&gt; [185] ragg_1.2.5                  aplot_0.1.10               </span></span>
<span><span class="co">#&gt; [187] clusterProfiler_4.8.1       memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [189] GenomicAlignments_1.36.0    cluster_2.1.4</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Sticker:2020" class="csl-entry">
Sticker, A., L. Goeminne, L. Martens, and L. Clement. 2020. <span>“Robust Summarization and Inference in Proteome-Wide Label-Free Quantification.”</span> Journal Article. <em>Mol Cell Proteomics</em> 19 (7): 1209–19. <a href="https://doi.org/10.1074/mcp.RA119.001624" class="external-link">https://doi.org/10.1074/mcp.RA119.001624</a>.
</div>
<div id="ref-Wu:2021" class="csl-entry">
Wu, T., E. Hu, S. Xu, M. Chen, P. Guo, Z. Dai, T. Feng, et al. 2021. <span>“clusterProfiler 4.0: A Universal Enrichment Tool for Interpreting Omics Data.”</span> Journal Article. <em>Innovation (Camb)</em> 2 (3): 100141. <a href="https://doi.org/10.1016/j.xinn.2021.100141" class="external-link">https://doi.org/10.1016/j.xinn.2021.100141</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zhenhuan Feng, Xiaofei Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
