<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DEP2">
<title>DEP2 Application Guide • DEP2</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DEP2 Application Guide">
<meta property="og:description" content="DEP2">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DEP2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.06.12</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/app_demo.html">DEP2 Application Guide</a>
    <a class="dropdown-item" href="../articles/Data_import.html">Data import</a>
    <a class="dropdown-item" href="../articles/DEP2_analysis2.html">DEP2 analysis workflow</a>
    <a class="dropdown-item" href="../articles/Quick_start.html">Quick start</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>DEP2 Application Guide</h1>
                        <h4 data-toc-skip class="author">Zhenhuan
Feng</h4>
                        
      
      
      <div class="d-none name"><code>app_demo.rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>DEP2 contains a shiny application for the researchers without
programming experience. The application provides a easy-used GUI to
perform most analysis in DEP2. We will briefly summary the app
operation. We put the example tables on this <a href="https://github.com/mildpiggy/OmicsExample?_blank" class="external-link">repository</a>.</p>
</div>
<div class="section level2">
<h2 id="run-app">Run App<a class="anchor" aria-label="anchor" href="#run-app"></a>
</h2>
<p>The application can be invoked by executing a single commond.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/run_app.html">run_app</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="add-module">Add Module<a class="anchor" aria-label="anchor" href="#add-module"></a>
</h2>
<p>In DEP2, the shiny application is designed in modules. Different
analysis functionalists are divided into different modules. And mocules
can be import in the add moudle button.</p>
<p><img src="figs/APP_1.png" title="Import analysis module"></p>
<p><img src="figs/APP2.png" title="Import analysis module"></p>
<p>Now, the app contains following modules:</p>
<ol style="list-style-type: decimal">
<li>
<p>The omics modules</p>
<ul>
<li>
<p><strong>Quantificated proteome(based on proteingroup
quantification)</strong></p>
<p>Differential analysis of protein-level quantitative data, like
intensity, LFQ.intensity or quantity tables.</p>
</li>
<li>
<p><strong>Quantificated proteome(based on peptide
quantification)</strong></p>
<p>Analysis differentially expression protein upon the re-aggregated
quantities from peptide-level quantitative data.</p>
</li>
<li>
<p><strong>Modification-specific proteome(modified peptide
quantification)</strong></p>
<p>Analysis differentially enriched modification-specific peptides,
receiving peptide quantity from modification-specific
enrichment.</p>
</li>
<li>
<p><strong>Transcriptome(RNA counts)</strong></p>
<p>Differential analysis of parallel transcriptome based on package
DEseq2.</p>
</li>
</ul>
</li>
<li>
<p>The post analysis modules</p>
<ul>
<li>
<p><strong>Over-representation analysis</strong></p>
<p>The biological functional enrichment by
<code>test_ora</code>.</p>
</li>
<li>
<p><strong>GSEA</strong></p>
<p>Another biological functional enrichment by
<code>test_gsea</code>.</p>
</li>
<li>
<p><strong>Protein-protein interaction</strong></p>
<p>Construct PPI network using a local STRING database by test_ppi. If
local data is absent ()</p>
</li>
<li>
<p><strong>Timecourse cluster</strong></p>
<p>Soft cluster on multi-groups assay.</p>
</li>
<li>
<p><strong>Integrated analysis</strong></p>
<p>Compare results between omics experiments (PTM data isn’t supported
yet).</p>
</li>
</ul>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="omics-analysis-module">Omics Analysis Module<a class="anchor" aria-label="anchor" href="#omics-analysis-module"></a>
</h2>
<div class="section level3">
<h3 id="upload-table">Upload table<a class="anchor" aria-label="anchor" href="#upload-table"></a>
</h3>
<p>Three type to upload table:</p>
<ol style="list-style-type: lower-alpha">
<li><p>Only upload expression table. The experiment design is generated
from expression column names. The colnames must under such rule,
(condition)_(repeataion), for example, <em>control_1</em>,
<em>treat_1</em>.</p></li>
<li><p>Upload both expression table and experiment design
table.</p></li>
<li><p>Upload a saved logfile from previous analysis, and recovery the
analysis options</p></li>
</ol>
<p>The application accepts tables in <em>.txt</em> (tab separated),
<em>.csv</em> (comma separated) or <em>.gz</em> (compressed table)
format. Log file is in a <em>.logrds</em> format and is saved from a
previous analysis module.</p>
<p><img src="figs/APP_omics_upload.png" title="Upload table"></p>
</div>
<div class="section level3">
<h3 id="set-columns-options">Set columns options<a class="anchor" aria-label="anchor" href="#set-columns-options"></a>
</h3>
<p>Set a correct read-in option including: identifiers, expression
values and filter columns.</p>
<p><img src="figs/APP_Column_option.png" title="Column options" width="1000"></p>
</div>
<div class="section level3">
<h3 id="set-analysis-options">Set analysis options<a class="anchor" aria-label="anchor" href="#set-analysis-options"></a>
</h3>
<p>Set the test mode.</p>
<p><img src="figs/APP_test_option.png" title="test options" width="400"></p>
<p>Other options can be adjusted in option pannels.</p>
<p><img src="figs/APP_test_option2.png" title="test options2"></p>
<p>Besides, the peptide-protein aggregation pipeline contains two main
calculate step, aggregation and differential test. They should perform
step by step in the module.</p>
<p><img src="figs/APP_test_PTM_option.png" title="test options PTM"></p>
</div>
<div class="section level3">
<h3 id="result-export">Result export<a class="anchor" aria-label="anchor" href="#result-export"></a>
</h3>
<p><img src="figs/APP_downtable_option.png" title="test options PTM"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zhenhuan Feng, Xiaofei Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
