<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DEP2">
<title>DEP2 analysis workflow • DEP2</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DEP2 analysis workflow">
<meta property="og:description" content="DEP2">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DEP2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.06.12</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/app_demo.html">DEP2 Application Guide</a>
    <a class="dropdown-item" href="../articles/Data_import.html">Data import</a>
    <a class="dropdown-item" href="../articles/DEP2_analysis2.html">DEP2 analysis workflow</a>
    <a class="dropdown-item" href="../articles/Quick_start.html">Quick start</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="DEP2_analysis2_files/htmlwidgets-1.6.1/htmlwidgets.js"></script><link href="DEP2_analysis2_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="DEP2_analysis2_files/datatables-binding-0.27/datatables.js"></script><link href="DEP2_analysis2_files/dt-core-1.12.1/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="DEP2_analysis2_files/dt-core-1.12.1/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="DEP2_analysis2_files/dt-core-1.12.1/js/jquery.dataTables.min.js"></script><link href="DEP2_analysis2_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="DEP2_analysis2_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>DEP2 analysis workflow</h1>
                        <h4 data-toc-skip class="author">Zhenhuan
Feng</h4>
                        
      
      
      <div class="d-none name"><code>DEP2_analysis2.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vigenette introduce proteiomics analysis workflow in DEP2
      package. Here we used the example dataset from a multiple omics
      study of silicosis mouse model.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>DEP2</code> is a downstream analysis toolkit for mass
spectrometry based quantitative proteomics data. The package contains an
integrated workflow for proteomics analysis, including data processing,
missing value imputation, hypothesis test, visualization and downstream
function exploration. It accepts various of proteomics results generated
by upstream search and quantitative software and provides function to
format table. In the vignette, we use the built-in data in DEP2. The
native tables out from upstream software are available in this <a href="https://github.com/mildpiggy/OmicsExample?_blank" class="external-link">repository</a>.</p>
<p>Now, <code>DEP2</code> provide three types of differential proteomics
analysis:</p>
<ol style="list-style-type: decimal">
<li>
<p>The pipeline for proteins-level expression/enrichment analysis,
started from protein-level quantity.</p>
<p>This pipeline is basically follows the methods in <code>DEP</code>,
requires a protein-level quantity result
(e.g. proteingroups.txt).</p>
</li>
<li>
<p>The pipeline for proteins-level expression/enrichment analysis,
started from peptide-level quantity.</p>
<p>This pipeline clustered the peptide to protein summarization
strategies in <code>QFeatures</code>, requires a protein-level quantity
(e.g. peptide.txt).</p>
</li>
<li>
<p>The pipeline for post-translation modification (PTM) specified
proteomics, performed upon modified peptides quantity.</p>
<p>This ppipline derived from the first one, and concludes the
additional modification information in analysis.</p>
</li>
</ol>
<p>Besides, DEP2 packages a classical RNA-seq analysis from DESeq2.
Downstream exploration functions are also coupled into workflow,
including: functional enrichment provided by <em><a href="https://bioconductor.org/packages/3.16/clusterProfiler" class="external-link">clusterProfiler</a></em>
<span class="citation">(Wu et al. 2021)</span> and <em><a href="https://bioconductor.org/packages/3.16/ReactomePA" class="external-link">ReactomePA</a></em>`;
protein-protein interaction inference from STRING database; and
time-course expression pattern clustering though cmeans cluster. As an
optional function, DEP2 also packaged a RNA-seq data analysis
pipeline.</p>
</div>
<div class="section level2">
<h2 id="differentially-expression-analysis">Differentially expression analysis<a class="anchor" aria-label="anchor" href="#differentially-expression-analysis"></a>
</h2>
<p>The steps of four analysis pipelines are slightly different.</p>
<div class="section level3">
<h3 id="analysis-proteomics-data-using-protein-level-results">Analysis proteomics data using protein-level results<a class="anchor" aria-label="anchor" href="#analysis-proteomics-data-using-protein-level-results"></a>
</h3>
<p><strong>Load data</strong></p>
<div class="figure">
<img src="figs/DEP2workflow1.png" title="Fig. 1 Pipeline for proteingroups" alt=""><p class="caption">fig_pipeline1</p>
</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## ProteinGroups table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Silicosis_pg"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Formatting name(gene symbol) and id(protein ID). </span></span>
<span><span class="co">## Generate a unique names for each protein. names and ids are columns in table</span></span>
<span><span class="va">unique_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_unique.html">make_unique</a></span><span class="op">(</span><span class="va">Silicosis_pg</span>, names <span class="op">=</span> <span class="st">"Gene.names"</span>, </span>
<span>                         ids <span class="op">=</span> <span class="st">"Protein.IDs"</span>, delim <span class="op">=</span> <span class="st">";"</span><span class="op">)</span> </span></code></pre></div>
<p><strong>Convert SE object</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Take expression columns(LFQ intensity in this cases).</span></span>
<span><span class="va">ecols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"LFQ.intensity."</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">unique_pg</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Construct SE. The experiement design is exctracted from column.</span></span>
<span><span class="va">se_pg</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/make_se_parse.html">make_se_parse</a></span><span class="op">(</span><span class="va">unique_pg</span>, columns <span class="op">=</span> <span class="va">ecols</span>, mode <span class="op">=</span> <span class="st">"delim"</span>, </span>
<span>                             sep <span class="op">=</span> <span class="st">"_"</span>, remove_prefix <span class="op">=</span> <span class="cn">T</span>, log2transform <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><strong>Filter</strong></p>
<p>Undesired features, like reverse sequences or contaminant protein
hits, may contain in the result table. Besides, missing values (MVs) is
inevasible label-free MS-based proteomics especially for DDA data.
Filter out reverse, contaminant, and low quality features with many
missing is required for following statistical test.
<code>filter_se</code> can filter features based on either MVs threshold
or other giving criterion.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Filter on MVs.</span></span>
<span><span class="va">filter_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">se_pg</span>,</span>
<span>                       thr <span class="op">=</span> <span class="fl">1</span>,  <span class="co">## the threshold of missing number in at least one condition</span></span>
<span>                       fraction <span class="op">=</span> <span class="fl">0.3</span> <span class="co">## the threshold of missing occupancy in each protein</span></span>
<span>                       </span>
<span>                      <span class="op">)</span></span>
<span><span class="co">#&gt; filter base on missing number is &lt;= 1 in at least one condition.</span></span>
<span><span class="co">#&gt; filter base on missing number fraction &lt; 0.3 in each row</span></span>
<span></span>
<span><span class="co">## Further filter can be performed though filter_formula.</span></span>
<span><span class="co">### Fox example, remove decoys or contaminants.</span></span>
<span><span class="va">filter_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">filter_pg</span>,</span>
<span>                       filter_formula <span class="op">=</span> <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">'+'</span> <span class="op">&amp;</span> <span class="va">Potential.contaminant</span> <span class="op">!=</span><span class="st">"+"</span> <span class="co">## filter upon Reverse and contaminant</span></span>
<span>                      <span class="op">)</span></span>
<span><span class="co">#&gt; filter base on giving formula</span></span>
<span><span class="co">### Or set a unique.peptide threshold.</span></span>
<span><span class="va">filter_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">filter_pg</span>,</span>
<span>                       filter_formula <span class="op">=</span> <span class="op">~</span> <span class="va">Unique.peptides</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; filter base on giving formula</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/get_df_wide.html">get_df_wide</a></span><span class="op">(</span><span class="va">se_pg</span><span class="op">)</span><span class="op">$</span><span class="va">Reverse</span> <span class="op">%&gt;%</span> <span class="va">table</span></span>
<span><span class="co">#&gt; .</span></span>
<span><span class="co">#&gt;           + </span></span>
<span><span class="co">#&gt; 10378   340</span></span>
<span><span class="fu"><a href="../reference/get_df_wide.html">get_df_wide</a></span><span class="op">(</span><span class="va">filter_pg</span><span class="op">)</span><span class="op">$</span><span class="va">Reverse</span> <span class="op">%&gt;%</span> <span class="va">table</span></span>
<span><span class="co">#&gt; .</span></span>
<span><span class="co">#&gt;      </span></span>
<span><span class="co">#&gt; 8982</span></span>
<span></span>
<span><span class="op">(</span><span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span><span class="va">se_pg</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Identification overlap before filter"</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> </span>
<span>  <span class="op">(</span><span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span><span class="va">filter_pg</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Identification overlap after before filter"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/filter%20pg-1.png" width="700"></p>
<p><strong>Normalization</strong></p>
<p>In the DEP and DEP2 workflow, assay is log2-transformed during SE
construction. Next, proteingroup assay is further normalized by variance
stabilizing transformation(vsn).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">norm_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_vsn.html">normalize_vsn</a></span><span class="op">(</span><span class="va">filter_pg</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_normalization.html">plot_normalization</a></span><span class="op">(</span><span class="va">filter_pg</span>, <span class="va">norm_pg</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<p><strong>imputation</strong></p>
<p>A considerable proportion of MVs still remain in the assay after
filtering.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_missval.html">plot_missval</a></span><span class="op">(</span><span class="va">filter_pg</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/missing%20heatmap-1.png" width="700"></p>
<p>MVs can be divided into missing at random (MAR), missing at random
(MAR) and missing not at random (MNAR), based on NA frequency intensity
signal. Function <code><a href="../reference/impute.html">DEP2::impute</a></code> draws kinds of imputation on
SE object. Inheriting from DEP, <code>impute</code> cluster the
imputation functions from <em><a href="https://bioconductor.org/packages/3.16/MSnbase" class="external-link">MSnbase</a></em>
and <em><a href="https://bioconductor.org/packages/3.16/MsCoreUtils" class="external-link">MsCoreUtils</a></em>,
including following methods.</p>
<p>– Left-censored imputation methods, replace: <em>“MinDet”, “MinProb”,
“min”(LOD), “QRILC”, “zero”, “man”(impute by a left-shifted
distribution)</em></p>
<p>– local similarity methods: <em>“knn”, “nbavg”</em></p>
<p>– Global-structure methods: <em>“MLE”, “BPCA”</em></p>
<p>Further detials can view the vignettes in <em><a href="https://bioconductor.org/packages/3.16/DEP" class="external-link">DEP</a></em> and
<em><a href="https://bioconductor.org/packages/3.16/MSnbase" class="external-link">MSnbase</a></em>.
In additional, DEP2 clusters two machine learning based imputation in
<code>impute</code>: <em>“RF”</em> and <em>“GSimp”</em>. “RF” call the
<code><a href="https://rdrr.io/pkg/missForest/man/missForest.html" class="external-link">missForest::missForest</a></code>, which train a random forest based
on observed parts of dataset. And “GSimp” is a Gibbs sampler based
left-censored imputation method proposed by <a href="http:s//doi.org/10.1371/journal.%20pcbi.1005973">Runmin Wei,
Jingye Wang, etc</a>. These two methods take a long time on iteration
but are reported that could draw a better estimation on missing
values.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">35</span><span class="op">)</span></span>
<span><span class="va">sample_rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">norm_pg</span><span class="op">)</span>, <span class="fl">300</span><span class="op">)</span></span>
<span><span class="va">norm_pg_sample</span> <span class="op">=</span> <span class="va">norm_pg</span><span class="op">[</span><span class="va">sample_rows</span>,<span class="op">]</span> <span class="co"># random sample 150 features to reduce runing time</span></span>
<span><span class="fu"><a href="../reference/plot_detect.html">plot_detect</a></span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/impute-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Impute missing data using random draws from a Gaussian distribution centered around a minimal value (for MNAR)</span></span>
<span><span class="va">imp_pg_MinProb</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"MinProb"</span>, q <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: imputeLCMD</span></span>
<span><span class="co">#&gt; [1] 0.31744</span></span>
<span></span>
<span><span class="co">## Impute missing data using k-nearest neighbour approach</span></span>
<span><span class="va">imp_pg_knn</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 11 rows with more than 50 % entries missing;</span></span>
<span><span class="co">#&gt;  mean imputation used for these rows</span></span>
<span></span>
<span><span class="co">## Impute missing data using missForest</span></span>
<span><span class="va">imp_pg_RF</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"RF"</span>, ntree <span class="op">=</span> <span class="fl">50</span>, mtry <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## Impute missing data using Gibbs</span></span>
<span><span class="va">imp_pg_GSimp</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"GSimp"</span>, hi_q <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                       iters_each<span class="op">=</span><span class="fl">40</span>, iters_all<span class="op">=</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; Iteration 1 start...end!</span></span>
<span><span class="co">#&gt; Iteration 2 start...end!</span></span>
<span><span class="co">#&gt; Iteration 3 start...end!</span></span>
<span><span class="co">#&gt; Iteration 4 start...end!</span></span>
<span><span class="co">#&gt; Iteration 5 start...end!</span></span>
<span><span class="co">#&gt; Iteration 6 start...end!</span></span>
<span><span class="co">#&gt; Iteration 7 start...end!</span></span>
<span><span class="co">#&gt; Iteration 8 start...end!</span></span>
<span></span>
<span><span class="co">## Impute missing data using QRILC</span></span>
<span><span class="va">imp_pg_QRILC</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg_sample</span>, fun <span class="op">=</span> <span class="st">"QRILC"</span><span class="op">)</span> </span></code></pre></div>
<p>The distribution of mean intensity of missing or non-missing
proteingroups</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_detect.html">plot_detect</a></span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/unnamed-chunk-2-1.png" width="700">
The difference among imputation methods.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">NAs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co">## the imputed values by different methods.</span></span>
<span><span class="va">imps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"GSimp"</span> <span class="op">=</span> <span class="va">imp_pg_GSimp</span>, <span class="st">"QRILC"</span> <span class="op">=</span> <span class="va">imp_pg_QRILC</span>, <span class="st">"MinProb"</span> <span class="op">=</span> <span class="va">imp_pg_MinProb</span>, <span class="st">"RF"</span> <span class="op">=</span> <span class="va">imp_pg_RF</span>, <span class="st">"knn"</span> <span class="op">=</span> <span class="va">imp_pg_knn</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">x</span> <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">data.frame</span> <span class="op">%&gt;%</span> <span class="fu">gather</span><span class="op">(</span><span class="st">"label"</span>, <span class="st">"value"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu">left_join</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"label"</span>,<span class="st">"condition"</span><span class="op">)</span><span class="op">]</span>,copy <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">magrittr</span><span class="fu">::</span><span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">extract</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">NAs</span><span class="op">)</span>,<span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html" class="external-link">rbindlist</a></span><span class="op">(</span>idcol <span class="op">=</span> <span class="st">"method"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(label)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(label)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(label)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(label)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(label)`</span></span>
<span></span>
<span><span class="co">## the original normalized values without imputation</span></span>
<span><span class="va">nonimps</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">data.frame</span> <span class="op">%&gt;%</span> <span class="fu">gather</span><span class="op">(</span><span class="st">"label"</span>, <span class="st">"value"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">norm_pg_sample</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"label"</span>,<span class="st">"condition"</span><span class="op">)</span><span class="op">]</span>,copy <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">magrittr</span><span class="fu">::</span><span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">extract</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">NAs</span><span class="op">)</span>,<span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"non_impute"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">method</span>,<span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(label)`</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/ggridges/" class="external-link">ggridges</a></span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">imps</span>, <span class="va">nonimps</span><span class="op">)</span>,<span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span>,y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">method</span>,level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">method</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggridges/man/geom_density_ridges.html" class="external-link">geom_density_ridges</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#027ad450"</span>, scale <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>                      jittered_points <span class="op">=</span> <span class="cn">TRUE</span>,position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggridges/man/position_points_jitter.html" class="external-link">position_points_jitter</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>                      point_shape <span class="op">=</span> <span class="st">'|'</span>, point_size <span class="op">=</span> <span class="fl">2</span>, point_alpha <span class="op">=</span> <span class="fl">1</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="st">"Impute method"</span><span class="op">)</span><span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"Log2 value"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlim</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">9</span>,<span class="fl">39</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_DEP1.html">theme_DEP1</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.356</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/compare%20imp-1.png" width="700"></p>
<p>Left-censored approaches, like MinProb and GSimp replace MVs by
conservatively low values; local similarity method “knn” and global data
learning “RF” impute relatively larger values. .These two method cater
to MNAR and MAR respectively. But “knn” will impute a mean value when
the row contains more than 50% missing.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"norm_pg_sample"</span>, <span class="st">"imp_pg_knn"</span>,<span class="st">"imp_pg_MinProb"</span>,<span class="st">"imp_pg_RF"</span>,<span class="st">"imp_pg_GSimp"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Impute missing data using q-th quantile for following analysis</span></span>
<span><span class="va">imp_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">norm_pg</span>, fun <span class="op">=</span> <span class="st">"MinDet"</span> ,q <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p><strong>Hypothesis test</strong></p>
<p>Function <code>test_diff</code> perform a moderated t-test from
limma. In DEP, p values is corrected by the <code>fdrtool</code> to
classify significantly regulated/enriched candidates and stable
proteins. DEP2 additionally provides two alternative frequently-used FDR
control method, <em>“Benjamini-Hochberg fdr”</em> and <em>“Storey’s
qvalue”</em>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Test every sample versus PBS control</span></span>
<span><span class="va">diff_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_pg</span>, type <span class="op">=</span> <span class="st">"control"</span>, control <span class="op">=</span> <span class="st">"PBS"</span>, fdr.type <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Tested contrasts: W10_vs_PBS, W2_vs_PBS, W4_vs_PBS, W6_vs_PBS, W9_vs_PBS</span></span>
<span><span class="co">#&gt; BH</span></span>
<span><span class="co">## Test on manul contrasts</span></span>
<span><span class="va">diff_pg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_pg</span>, type <span class="op">=</span> <span class="st">"manual"</span>, test  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W4_vs_PBS"</span><span class="op">)</span>, fdr.type <span class="op">=</span> <span class="st">"Storey's qvalue"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Tested contrasts: W4_vs_PBS</span></span>
<span><span class="co">#&gt; Storey's qvalue</span></span></code></pre></div>
<p>Function <code>add_rejections</code> can classify significant hits
according L2FC (<em>lfc</em>) and adjusted p value (<em>alpha</em>)
threshold.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Add significant rejections for features, based on </span></span>
<span><span class="va">dep_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_pg</span>, alpha <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get the significant subset</span></span>
<span><span class="va">dep_pg_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dep_pg_sig</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 249</span></span></code></pre></div>
<p>Use the <code>plot_volcano</code> to have a quick look on result.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### volcano plot on contrast "W4_vs_PBS"</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pg</span>, contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, adjusted <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/rejections%20plot%201-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## plot the cutoff line</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pg</span>, contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, adjusted <span class="op">=</span> <span class="cn">F</span>,</span>
<span>             add_threshold_line <span class="op">=</span> <span class="st">"intersect"</span>, pCutoff  <span class="op">=</span> <span class="fl">0.05</span>, fcCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; add intersect threshold line. fcCutoff = 1pCutoff = 0.05</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/rejections%20plot%201-2.png" width="700"></p>
<p>Besides the intersect method, add_rejections can draw a curve cutoff
approach described by <a href="https://www.sciencedirect.com/science/article/pii/S1535947620316674" class="external-link">Eva
C.Keilhauer</a>. Instead of intersect straight lines, the approach used
curve lines with <span class="math inline">\(y &gt; c/(x -
x_0)\)</span>, where <span class="math inline">\(x\)</span> is the log2
FC, <span class="math inline">\(y\)</span> is adjusted P value of
features (<em>c</em> = curvature, <em>x[0]</em> = minimum L2FC). We use
the standard deviation <span class="math inline">\(\sigma\)</span>
determine x0 by <span class="math inline">\(x_0 =
x_0.fold*\sigma\)</span>. <span class="math inline">\(\sigma\)</span> is
the standard deviation of the Gaussian curve distribution of log2 FC in
each contrast.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## thresholdmethod = "curve"</span></span>
<span><span class="va">dep_pg_curve</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_pg</span>, thresholdmethod <span class="op">=</span> <span class="st">"curve"</span>, curvature  <span class="op">=</span> <span class="fl">2</span>, x0_fold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## the cutoff line on volcano</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pg_curve</span>, contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, add_threshold_line <span class="op">=</span> <span class="st">"curve"</span>, curvature <span class="op">=</span> <span class="fl">2</span>, x0_fold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pg_curve</span>, contrast <span class="op">=</span> <span class="st">"W6_vs_PBS"</span>, add_threshold_line <span class="op">=</span> <span class="st">"curve"</span>, curvature <span class="op">=</span> <span class="fl">2</span>, x0_fold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; add curve threshold line. Sigma = 0.41624138608919, x0 = 0.83248277217838, curvature = 2</span></span>
<span><span class="co">#&gt; add curve threshold line. Sigma = 0.478600344109222, x0 = 0.957200688218445, curvature = 2</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/rejections2-1.png" width="700"></p>
<p>You can check the <span class="math inline">\(\sigma\)</span> and
fitted Gaussian curve by <code>plot_diff_hist</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Check the fitted Gaussian curve</span></span>
<span><span class="fu"><a href="../reference/plot_diff_hist.html">plot_diff_hist</a></span><span class="op">(</span><span class="va">dep_pg_curve</span>, contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W4_vs_PBS"</span>, <span class="st">"W6_vs_PBS"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Get the fit result </span></span>
<span><span class="fu"><a href="../reference/plot_diff_hist.html">plot_diff_hist</a></span><span class="op">(</span><span class="va">dep_pg_curve</span>, plot <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="co">## a table of gaussian args \sigma and \mu</span></span>
<span><span class="co">#&gt;          contrast         mu    sigma2</span></span>
<span><span class="co">#&gt; 1 W10_vs_PBS_diff 0.01374325 0.2407074</span></span>
<span><span class="co">#&gt; 2  W2_vs_PBS_diff 0.02327110 0.2266282</span></span>
<span><span class="co">#&gt; 3  W4_vs_PBS_diff 0.02511490 0.1732569</span></span>
<span><span class="co">#&gt; 4  W6_vs_PBS_diff 0.02201209 0.2290583</span></span>
<span><span class="co">#&gt; 5  W9_vs_PBS_diff 0.02354730 0.2595546</span></span></code></pre></div>
<p><code>plot_heatmap</code> can plot a heatmap for significant
candidates.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/heatmap-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Reorder columns by condition</span></span>
<span><span class="va">dep_pg</span> <span class="op">=</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/Order_cols.html">Order_cols</a></span><span class="op">(</span><span class="va">dep_pg</span>,order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBS"</span>,<span class="st">"W2"</span>,<span class="st">"W4"</span>,<span class="st">"W6"</span>,<span class="st">"W9"</span>,<span class="st">"W10"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pg</span>, cluster_columns <span class="op">=</span> <span class="cn">F</span>, kmeans <span class="op">=</span> <span class="cn">T</span>, k <span class="op">=</span> <span class="fl">5</span>, seed <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># cluster features</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/heatmap-2.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Only plot the clusters that are up-regulated in treatment groups.</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pg</span>, cluster_columns <span class="op">=</span> <span class="cn">F</span>, kmeans <span class="op">=</span> <span class="cn">T</span>, k <span class="op">=</span> <span class="fl">5</span>, seed <span class="op">=</span> <span class="fl">1</span>, col_limit <span class="op">=</span> <span class="fl">4</span>,</span>
<span>             split_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span></span>
<span>             <span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/heatmap-3.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## plot on select contrast</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pg</span>, manual_contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/heatmap-4.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="analysis-based-on-protein-re-aggregation">Analysis based on protein re-aggregation<a class="anchor" aria-label="anchor" href="#analysis-based-on-protein-re-aggregation"></a>
</h3>
<p>The second pipeline is aggregation protein abundance from peptide
quantity data. The analysis workflow is different to last one in data
processing steps.</p>
<p><img src="figs/DEP2workflow2.png" title="Fig. 2 the pipeline for protein re-aggregation"></p>
<p><strong>Load data</strong></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Silicosis_peptide"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Convert a QFeatures object</strong></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Here we just select two groups to save computational cost in these vignette</span></span>
<span><span class="op">(</span><span class="va">ecols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Intensity.PBS|Intensity.W4"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Silicosis_peptide</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] "Intensity.PBS_1" "Intensity.PBS_2" "Intensity.PBS_3" "Intensity.PBS_4"</span></span>
<span><span class="co">#&gt; [5] "Intensity.W4_2"  "Intensity.W4_3"  "Intensity.W4_4"  "Intensity.W4_5"</span></span>
<span></span>
<span><span class="va">pe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_pe_parse.html">make_pe_parse</a></span><span class="op">(</span><span class="va">Silicosis_peptide</span>, columns <span class="op">=</span> <span class="va">ecols</span>, mode <span class="op">=</span> <span class="st">"delim"</span><span class="op">)</span></span>
<span><span class="va">pe</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">#&gt;  [1] peptideRaw: SummarizedExperiment with 143528 rows and 8 columns</span></span></code></pre></div>
<p><strong>Filter</strong></p>
<p>QFeatures object can be filtered though by <code>filter_pe</code>.
Its usage is similar to function <code>filter_se</code>, could filter
out unexpected features by missing ratio or based on identification
information in certain rowdata columns.</p>
<!-- ```{r filter pe,include=F,eval=FALSE} -->
<!-- ## Sample 800 protein ID as example, to reduce computational cost in these vignette -->
<!-- set.seed(50) -->
<!-- pe_peptides -->
<!-- sample_protein = sample(unique(Silicosis_peptide$Proteins),1000) -->
<!-- pe_peptides2 <- filter_pe(pe_peptides, filter_formula = ~ Proteins %in% sample_protein) -->
<!-- pe_peptides2 -->
<!-- ## fitlt out low quality peptides -->
<!-- filt_pe <- filter_pe(pe_peptides2, thr = 1,fraction = 0.4, filter_formula = ~ Reverse != '+' & Potential.contaminant !="+" ) -->
<!-- ``` -->
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filt_pe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_pe.html">filter_pe</a></span><span class="op">(</span><span class="va">pe</span>, </span>
<span>                     thr <span class="op">=</span> <span class="fl">1</span>,fraction <span class="op">=</span> <span class="fl">0.4</span>,  <span class="co">## filter based on missing values</span></span>
<span>                     filter_formula <span class="op">=</span> <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">'+'</span> <span class="op">&amp;</span> <span class="va">Potential.contaminant</span> <span class="op">!=</span><span class="st">"+"</span> <span class="op">)</span></span>
<span><span class="co">#&gt; filter base on missing number is &lt;= 1 in at least one condition.</span></span>
<span><span class="co">#&gt; filter base on missing number fraction &lt; 0.4 in each row</span></span>
<span><span class="co">#&gt; filter base on giving formula</span></span></code></pre></div>
<p><strong>imputation</strong></p>
<p>Imputation could perform before normalization, to reduce the
potential effect caused by MVs.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Use QFeatures::addAssay() and DEP2::impute</span></span>
<span><span class="va">pe</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">filt_pe</span>,</span>
<span>                              <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">filt_pe</span><span class="op">[[</span><span class="st">"peptideRaw"</span><span class="op">]</span><span class="op">]</span>, fun <span class="op">=</span> <span class="st">"QRILC"</span><span class="op">)</span>, </span>
<span>                              name <span class="op">=</span> <span class="st">"peptideImp"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_imputation.html">plot_imputation</a></span><span class="op">(</span><span class="va">pe</span><span class="op">[[</span><span class="st">"peptideRaw"</span><span class="op">]</span><span class="op">]</span>, <span class="va">pe</span><span class="op">[[</span><span class="st">"peptideImp"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/imp%20pe-1.png" width="700"></p>
<p><strong>Normalization</strong></p>
<p>Function <code>normalize_pe</code> perform normalize and call
<code>QFeatures:normalize</code>. Four normalization methods are
selectable., including <em>“diff.median”, “quantiles”,
“quantiles.robust” ,“vsn”</em>. Notice that <em>quantiles.robust</em>
can’t handle data table with MVs (ie NA). Therefore It is necessary to
Imputate in advance .</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pe</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">:::</span><span class="fu"><a href="../reference/normalize_pe.html">normalize_pe</a></span><span class="op">(</span><span class="va">pe</span>,method <span class="op">=</span> <span class="st">"quantiles.robust"</span>,</span>
<span>                               i <span class="op">=</span> <span class="st">"peptideImp"</span>,    <span class="co">## nomalize on which assay</span></span>
<span>                               name <span class="op">=</span> <span class="st">"peptideNorm"</span> <span class="co">## output assay name</span></span>
<span>                               <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_normalization.html">plot_normalization</a></span><span class="op">(</span><span class="va">pe</span><span class="op">[[</span><span class="st">"peptideRaw"</span><span class="op">]</span><span class="op">]</span>,<span class="va">pe</span><span class="op">[[</span><span class="st">"peptideImp"</span><span class="op">]</span><span class="op">]</span>,<span class="va">pe</span><span class="op">[[</span><span class="st">"peptideNorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/imp%20before%20norm2-1.png" width="700"></p>
<p><strong>Aggregation</strong></p>
<p>DEP2 packaged function <code>aggregate_pe</code> based on
<code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">QFeatures::aggregateFeatures</a></code>. The summarisation approach
<em>aggrefun</em> can be “RobustSummary” (also called MSqRobSum) <span class="citation">(Sticker et al. 2020)</span>, “medianPolish”,
“totalMean”. Details can see <code>?aggregateFeatures</code>. And
proteingroups can be summed by “Unique + Razor” or “Unique”.
<code>aggregate_pe</code> require several minutes depend on the peptide
assay scale.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">begin_time</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pe</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/aggregate_pe.html">aggregate_pe</a></span><span class="op">(</span><span class="va">pe</span>,</span>
<span>                         aggrefun <span class="op">=</span> <span class="st">"RobustSummary"</span>,</span>
<span>                         aggregate_Peptide_Type <span class="op">=</span> <span class="st">"Unique + Razor"</span>,</span>
<span>                         fcol <span class="op">=</span> <span class="st">"Proteins"</span>,                  <span class="co"># the protein ID column in input table</span></span>
<span>                         i <span class="op">=</span> <span class="st">"peptideNorm"</span>, <span class="co"># the assay to aggregate</span></span>
<span>                         reserve <span class="op">=</span> <span class="st">"Gene.names"</span>              <span class="co"># reserve the Gene.names in row data.</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "aggregate by Unique + Razor peptides"</span></span>
<span><span class="co">#&gt; [1] "peptides distribution finished"</span></span>
<span><span class="co">#&gt; [1] "begin assay aggregate"</span></span>
<span><span class="co">#&gt; [1] "begin reducedataframe"</span></span>
<span><span class="co">#&gt; [1] "reducedataframe finiched"</span></span>
<span><span class="co">#&gt; [1] "assay aggregate finished"</span></span>
<span><span class="co">#&gt; [1] "aggregation finished"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">begin_time</span><span class="op">)</span> <span class="co"># aggregate cost several minutes depent on the assay scale</span></span>
<span><span class="co">#&gt; Time difference of 1.03845 mins</span></span>
<span></span>
<span><span class="va">pe</span> <span class="co"># protein result is in the protein assay</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 4 assays:</span></span>
<span><span class="co">#&gt;  [1] peptideRaw: SummarizedExperiment with 98219 rows and 8 columns </span></span>
<span><span class="co">#&gt;  [2] peptideImp: SummarizedExperiment with 98219 rows and 8 columns </span></span>
<span><span class="co">#&gt;  [3] peptideNorm: SummarizedExperiment with 98219 rows and 8 columns </span></span>
<span><span class="co">#&gt;  [4] protein: SummarizedExperiment with 9768 rows and 8 columns</span></span>
<span><span class="va">se_pep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pe2se.html">pe2se</a></span><span class="op">(</span><span class="va">pe</span>, names <span class="op">=</span> <span class="st">"Gene.names"</span>, ids <span class="op">=</span> <span class="st">"smallestProteingroups"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Differential test</strong> Following differential test is
same to the proteingroups pipeline. Use <code>test_diff</code> and
<code>add_rejections</code> to classify significant protein.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Differential test</span></span>
<span><span class="va">diff_pep</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">se_pep</span>,type <span class="op">=</span> <span class="st">"control"</span>, control <span class="op">=</span> <span class="st">"PBS"</span>, fdr.type <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Tested contrasts: W4_vs_PBS</span></span>
<span><span class="co">#&gt; BH</span></span>
<span><span class="co">## Add rejections</span></span>
<span><span class="va">dep_pep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_pep</span>,alpha <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_pep</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">nrow</span></span>
<span><span class="co">#&gt; [1] 76</span></span></code></pre></div>
<p>Data visualization functions is also worked for DEP results from
re-aggregation.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Volcano</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_pep</span>, contrast <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, add_threshold_line <span class="op">=</span> <span class="st">"intersect"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; add intersect threshold line. fcCutoff = 1pCutoff = 0.05</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/DEP%20visualization-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Heatmap</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap.html">plot_heatmap</a></span><span class="op">(</span><span class="va">dep_pep</span>, kmeans <span class="op">=</span> <span class="cn">T</span>,k <span class="op">=</span> <span class="fl">5</span>,col_limit <span class="op">=</span> <span class="fl">6</span><span class="op">)</span> </span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/DEP%20visualization-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="analysis-for-ptm-specific-proteomics">Analysis for PTM-specific proteomics<a class="anchor" aria-label="anchor" href="#analysis-for-ptm-specific-proteomics"></a>
</h3>
<p>DEP2 designs a new workflow for post-modification proteomics data
analysis begin from modified peptides quantitative result. The
post-translation modification information is necessary to distinguish
modified sites, including position, modified residue (amino acid), gene
name or protein ID that modification is associated with.</p>
<p><code>make_unique_ptm</code> creates(or overwrites) the PTM
information columns, including <em>name</em>, <em>ID</em>,
<em>gene_name</em>, <em>protein_ID</em>, <em>modified_aa</em>,
<em>modified_pos</em>. The <em>name</em>, <em>ID</em> columns is the
identifier of each modified-peptide, following such naming rule: ’(gene
name/protein ID)_(modified amino acid)(position of modification)’
(e.g. “TBCA_K51” and “O75347_K51”).</p>
<p><strong>Load data</strong></p>
<p>The example we used here is the phosphorylated peptide result, table
<strong>Phospho (STY)Sites.txt</strong>, from MaxQuant. In the table</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## phosphorylated peptides table of the silicosis mouse model.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Silicosis_phos</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## Format the modification information and generated modified-peptides identifier.</span></span>
<span><span class="co">## aa and pos is the modified amino acids and modified site in protein.</span></span>
<span><span class="va">unique_pho</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_unique_ptm.html">make_unique_ptm</a></span><span class="op">(</span><span class="va">Silicosis_phos</span>, gene_name <span class="op">=</span> <span class="st">"Gene.names"</span>, </span>
<span>                              protein_ID <span class="op">=</span> <span class="st">"Protein"</span>, aa <span class="op">=</span> <span class="st">"Amino.acid"</span>,</span>
<span>                              pos <span class="op">=</span> <span class="st">"Position"</span><span class="op">)</span> </span></code></pre></div>
<p>The <em>name</em>, <em>ID</em> is generated upon PTM information
columns.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">unique_pho</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span>,<span class="st">"ID"</span>,<span class="st">"Gene.names"</span>,<span class="st">"Protein"</span>,<span class="st">"Amino.acid"</span>,<span class="st">"Position"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>              options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">7</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-17466358a46d1399716e" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-17466358a46d1399716e">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7"],["P0DPB4_S60","P0DPB4_S61","P0DPB4_S62","P0DPB4_S63","P0DPB4_S67","Ccdc173_S151","Tbc1d25_S560"],["P0DPB4_S60","P0DPB4_S61","P0DPB4_S62","P0DPB4_S63","P0DPB4_S67","A0JLY1_S151","A1A5B6_S560"],["","","","","","Ccdc173","Tbc1d25"],["P0DPB4","P0DPB4","P0DPB4","P0DPB4","P0DPB4","A0JLY1","A1A5B6"],["S","S","S","S","S","S","S"],[60,61,62,63,67,151,560]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>ID<\/th>\n      <th>Gene.names<\/th>\n      <th>Protein<\/th>\n      <th>Amino.acid<\/th>\n      <th>Position<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":7,"columnDefs":[{"className":"dt-right","targets":6},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[7,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p><strong>Covert SE</strong></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Take 'Intensity' columns.</span></span>
<span><span class="va">ecols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Intensity."</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">unique_pho</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Construct a SE object mentioned before.</span></span>
<span><span class="va">se_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_se_parse.html">make_se_parse</a></span><span class="op">(</span><span class="va">unique_pho</span>, columns <span class="op">=</span> <span class="va">ecols</span>, </span>
<span>                        mode <span class="op">=</span> <span class="st">"delim"</span>, sep <span class="op">=</span> <span class="st">"_"</span>, remove_prefix <span class="op">=</span> <span class="cn">T</span>, log2transform <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><strong>Filter</strong></p>
<p>The following steps is Basically similar to the protomics analysis
pipeline. <code>filter_se</code> are a universal filter function for SE
object. A additional threshold should be set on ‘Localization.prob’ in
the MaxQuant modification result.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Filter base on both missing occupancy and the localization probability for this site.</span></span>
<span><span class="va">filt_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">se_ptm</span>, </span>
<span>                      thr <span class="op">=</span> <span class="fl">1</span>, fraction <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                      filter_formula <span class="op">=</span> <span class="op">~</span> <span class="va">Reverse</span><span class="op">!=</span><span class="st">"+"</span> <span class="op">&amp;</span> </span>
<span>                        <span class="va">Potential.contaminant</span><span class="op">!=</span><span class="st">"+"</span> <span class="op">&amp;</span> </span>
<span>                        <span class="va">Localization.prob</span><span class="op">&gt;</span><span class="fl">0.7</span></span>
<span>                      <span class="op">)</span></span></code></pre></div>
<p><strong>Normalization</strong></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## VSN normalization</span></span>
<span><span class="va">norm_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_vsn.html">normalize_vsn</a></span><span class="op">(</span><span class="va">filt_ptm</span><span class="op">)</span></span></code></pre></div>
<p><strong>Imputation</strong></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## KNN impute relatively larger values</span></span>
<span><span class="va">imp_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute.html">impute</a></span><span class="op">(</span><span class="va">filt_ptm</span>, fun<span class="op">=</span> <span class="st">"knn"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 2065 rows with more than 50 % entries missing;</span></span>
<span><span class="co">#&gt;  mean imputation used for these rows</span></span>
<span><span class="co">#&gt; Cluster size 11729 broken into 6954 4775 </span></span>
<span><span class="co">#&gt; Cluster size 6954 broken into 2918 4036 </span></span>
<span><span class="co">#&gt; Cluster size 2918 broken into 2328 590 </span></span>
<span><span class="co">#&gt; Cluster size 2328 broken into 1841 487 </span></span>
<span><span class="co">#&gt; Cluster size 1841 broken into 1661 180 </span></span>
<span><span class="co">#&gt; Cluster size 1661 broken into 1404 257 </span></span>
<span><span class="co">#&gt; Done cluster 1404 </span></span>
<span><span class="co">#&gt; Done cluster 257 </span></span>
<span><span class="co">#&gt; Done cluster 1661 </span></span>
<span><span class="co">#&gt; Done cluster 180 </span></span>
<span><span class="co">#&gt; Done cluster 1841 </span></span>
<span><span class="co">#&gt; Done cluster 487 </span></span>
<span><span class="co">#&gt; Done cluster 2328 </span></span>
<span><span class="co">#&gt; Done cluster 590 </span></span>
<span><span class="co">#&gt; Done cluster 2918 </span></span>
<span><span class="co">#&gt; Cluster size 4036 broken into 56 3980 </span></span>
<span><span class="co">#&gt; Done cluster 56 </span></span>
<span><span class="co">#&gt; Cluster size 3980 broken into 2552 1428 </span></span>
<span><span class="co">#&gt; Cluster size 2552 broken into 2523 29 </span></span>
<span><span class="co">#&gt; Cluster size 2523 broken into 468 2055 </span></span>
<span><span class="co">#&gt; Done cluster 468 </span></span>
<span><span class="co">#&gt; Cluster size 2055 broken into 1044 1011 </span></span>
<span><span class="co">#&gt; Done cluster 1044 </span></span>
<span><span class="co">#&gt; Done cluster 1011 </span></span>
<span><span class="co">#&gt; Done cluster 2055 </span></span>
<span><span class="co">#&gt; Done cluster 2523 </span></span>
<span><span class="co">#&gt; Done cluster 29 </span></span>
<span><span class="co">#&gt; Done cluster 2552 </span></span>
<span><span class="co">#&gt; Done cluster 1428 </span></span>
<span><span class="co">#&gt; Done cluster 3980 </span></span>
<span><span class="co">#&gt; Done cluster 4036 </span></span>
<span><span class="co">#&gt; Done cluster 6954 </span></span>
<span><span class="co">#&gt; Cluster size 4775 broken into 2517 2258 </span></span>
<span><span class="co">#&gt; Cluster size 2517 broken into 1379 1138 </span></span>
<span><span class="co">#&gt; Done cluster 1379 </span></span>
<span><span class="co">#&gt; Done cluster 1138 </span></span>
<span><span class="co">#&gt; Done cluster 2517 </span></span>
<span><span class="co">#&gt; Cluster size 2258 broken into 861 1397 </span></span>
<span><span class="co">#&gt; Done cluster 861 </span></span>
<span><span class="co">#&gt; Done cluster 1397 </span></span>
<span><span class="co">#&gt; Done cluster 2258 </span></span>
<span><span class="co">#&gt; Done cluster 4775</span></span></code></pre></div>
<p><strong>Differantail test</strong></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diff_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_ptm</span>, type <span class="op">=</span> <span class="st">"manual"</span>, test <span class="op">=</span> <span class="st">"PBS_vs_W6"</span> , fdr.type <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Tested contrasts: PBS_vs_W6</span></span>
<span><span class="co">#&gt; BH</span></span>
<span></span>
<span><span class="va">dep_ptm</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_ptm</span>, alpha <span class="op">=</span> <span class="fl">0.05</span>, lfc <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">dep_ptm</span>,adjusted <span class="op">=</span> <span class="cn">T</span>, add_threshold_line <span class="op">=</span> <span class="st">"intersect"</span><span class="op">)</span></span>
<span><span class="co">#&gt; add intersect threshold line. fcCutoff = 1pCutoff = 0.05</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 9 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">#&gt; increasing max.overlaps</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="analysis-rna-seq-data">Analysis RNA-seq data<a class="anchor" aria-label="anchor" href="#analysis-rna-seq-data"></a>
</h3>
<p>Multi-omics study has already become a widespread method in research.
In particular, proteomics-transcriptomics combined studies can reveal
regulation details, from multiple perspectives. Thus DEP2 packages a
pipeline for RNA-seq data from DESeq2. here.</p>
<p><strong>Load data</strong></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Silicosis_counts</span><span class="op">)</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Silicosis_counts</span><span class="op">)</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-8228581aa3e4de6c04d3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8228581aa3e4de6c04d3">{"x":{"filter":"none","vertical":false,"data":[["ENSMUSG00000000001","ENSMUSG00000000028","ENSMUSG00000000031","ENSMUSG00000000049","ENSMUSG00000000056","ENSMUSG00000000058"],[133,26,0,88,31,211],[235,28,0,71,55,435],[174,25,38,66,74,215],[418,26,4,58,117,940],[388,16,0,60,96,571],[201,19,1,69,80,252],[153,22,0,68,53,460],[208,16,0,72,78,446],[200,17,0,63,26,405],[350,24,0,57,68,609],[147,22,0,61,45,389],[143,35,1,102,49,153],[318,22,0,40,73,446],[385,27,0,60,72,503],[171,27,1,68,43,155],[319,32,0,75,46,403],[185,5,0,39,49,287],[208,16,0,45,67,324]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>PBS_1<\/th>\n      <th>PBS_2<\/th>\n      <th>PBS_3<\/th>\n      <th>W2_1<\/th>\n      <th>W2_2<\/th>\n      <th>W2_3<\/th>\n      <th>W2_4<\/th>\n      <th>W4_1<\/th>\n      <th>W4_2<\/th>\n      <th>W4_3<\/th>\n      <th>W4_4<\/th>\n      <th>W6_1<\/th>\n      <th>W6_2<\/th>\n      <th>W6_3<\/th>\n      <th>W6_4<\/th>\n      <th>W10_1<\/th>\n      <th>W10_2<\/th>\n      <th>W10_3<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":6,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[6,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p><strong>Construct DESeqDataSet</strong></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_dds_parse.html">make_dds_parse</a></span><span class="op">(</span><span class="va">Silicosis_counts</span>,mode <span class="op">=</span> <span class="st">"delim"</span><span class="op">)</span></span>
<span><span class="va">dds</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 24936 18 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(24936): ENSMUSG00000000001 ENSMUSG00000000028 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000118633 ENSMUSG00000118636</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(18): PBS_1 PBS_2 ... W10_2 W10_3</span></span>
<span><span class="co">#&gt; colData names(4): label ID condition replicate</span></span></code></pre></div>
<p><strong>Filter</strong></p>
<p><em>DESeqDataSet</em> inherits from <em>SummarizedExperiment</em></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">inherits</a></span><span class="op">(</span><span class="va">dds</span>,<span class="st">"SummarizedExperiment"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p><code>Filter_se</code> can also filter <em>DESeqDataSet</em> objects.
Parameter <em>rowsum_threshold</em> set threshold on counts row sum.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">dds</span>, fraction <span class="op">=</span> <span class="fl">0.3</span>, thr <span class="op">=</span> <span class="fl">1</span>, rowsum_threshold <span class="op">=</span> <span class="fl">35</span><span class="op">)</span></span>
<span><span class="co">#&gt; filter base on missing number is &lt;= 1 in at least one condition.</span></span>
<span><span class="co">#&gt; filter base on missing number fraction &lt; 0.3 in each row</span></span>
<span><span class="co">#&gt; filter base on counts sum &gt;=35</span></span></code></pre></div>
<p><strong>Transform ID</strong></p>
<p>In this case, the gene identifier is Ensembl ID. In order to check
the gene conveniently, DEP2 can transform ID to gene symbol through the
<a href="%22http://www.bioconductor.org/packages/release/BiocViews.html#___OrgDb%22" class="external-link">OrgDb</a>
databases. <code>ID_transform</code> can perform the ID convert and
replace the rownames of the object.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Check if the annotation package is already installed.</span></span>
<span><span class="fu"><a href="../reference/check_organismDB_depends.html">check_organismDB_depends</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"Mouse"</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ENSMUSG00000000001" "ENSMUSG00000000028" "ENSMUSG00000000049"</span></span>
<span><span class="co">#&gt; [4] "ENSMUSG00000000056"</span></span>
<span></span>
<span><span class="co">## Transform ID</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/ID_transform.html">ID_transform</a></span><span class="op">(</span><span class="va">dds</span>, species <span class="op">=</span> <span class="st">"Mouse"</span><span class="op">)</span></span>
<span><span class="co">#&gt; org.Mm.eg.db</span></span>
<span><span class="co">#&gt; Loading required package: org.Mm.eg.db</span></span>
<span><span class="co">#&gt; Loading required package: AnnotationDbi</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'AnnotationDbi'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     select</span></span>
<span><span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Gnai3" "Cdc45" "Apoh"  "Narf"</span></span></code></pre></div>
<p>If needed, <code>ID_transform</code> can be used on proteinomics
data, for example:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/ID_transform.html">ID_transform</a></span><span class="op">(</span><span class="va">dep_pg</span>, </span>
<span>                   species <span class="op">=</span> <span class="st">"Mouse"</span>,</span>
<span>                   from_columns <span class="op">=</span> <span class="st">"ID"</span>,</span>
<span>                   fromtype <span class="op">=</span> <span class="st">"UNIPROT"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Differential test</strong></p>
<p>The differential test step is similar to the foregoing pipelines, but
the <code>test_diff</code> is replace by <code>test_diff_deg</code> to
call <em>DESeq2</em>.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff_deg.html">test_diff_deg</a></span><span class="op">(</span><span class="va">dds</span>, type <span class="op">=</span> <span class="st">"control"</span>, control <span class="op">=</span> <span class="st">"PBS"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Performing DESeq analysis...</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span><span class="co">#&gt; Formatting analysis result...</span></span>
<span><span class="co">#&gt; extracted results: W2_vs_PBS, W4_vs_PBS, W6_vs_PBS, W10_vs_PBS</span></span>
<span><span class="co">#&gt; Calulate a rlog &amp; log2 transform assay, stored in rlg or ntd slot...</span></span>
<span></span>
<span><span class="va">deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff</span>, alpha <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><code>test_diff_deg</code></p>
</div>
</div>
<div class="section level2">
<h2 id="export-result-table">Export result table<a class="anchor" aria-label="anchor" href="#export-result-table"></a>
</h2>
<p>Function <code>get_results</code> can get the result table (or
significant subset) from differential test results.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get result table</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_results.html">get_results</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span><span class="op">)</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-a15b670239e029179d2a" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-a15b670239e029179d2a">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["Acp5","Adamts8","Add2","Aif1","Alox12","Amot"],["Q05117","P57110","Q9QYB8","O70200","P39655","Q8VHG2"],[1.78379863582731e-05,0.15178205215544,0.000459654282707716,4.93848799851642e-06,0.00394875934561959,0.0239883081165031],[0.000678614541846775,0.020868134094725,0.00525680247101302,8.96955589615357e-08,0.00440808207017469,0.201858008425827],[0.00529697688646453,0.0114300595527202,0.000105742967668944,1.69833441208805e-07,0.005314785186329,0.163529214279777],[0.00024811571449112,0.000578958161813839,0.00102781721605583,2.99039153135236e-06,0.000177982271283717,0.000461920675940907],[0.000375399803073765,0.00588827634418241,0.00454051674486895,0.000150405954656807,0.00598400802072252,0.00045469080189917],[0.00157,0.309,0.00928,0.000726,0.0302,0.0926],[0.00616,0.0643,0.0246,2.52e-05,0.0217,0.336],[0.0325,0.0536,0.00296,6.97e-05,0.0325,0.314],[0.00318,0.00523,0.00772,0.000236,0.00255,0.00454],[0.00637,0.0329,0.0279,0.00389,0.0332,0.00701],[true,false,true,true,false,false],[false,false,false,true,false,false],[false,false,true,true,false,false],[false,true,true,true,true,true],[false,false,false,true,false,true],[true,true,true,true,true,true],[2.26,1.81,-2.93,3.27,-1.8,-1.56],[1.29,2.52,-1.76,3.64,-1.45,-0.68],[0.993,2.8,-2.78,3.47,-1.41,-0.746],[1.44,4.19,-2.18,2.78,-2.12,-2.23],[1.69,3.81,-2.2,2.41,-1.7,-2.74],[-1.14,-2.46,1.86,-2.55,1.35,1.16],[0.152,0.0511,0.0965,1.09,-0.104,0.482],[-0.148,0.339,-0.922,0.926,-0.0649,0.416],[0.3,1.73,-0.324,0.236,-0.773,-1.07],[0.549,1.35,-0.345,-0.141,-0.352,-1.58],[1.12,-0.654,-1.07,0.725,-0.457,-0.397]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>ID<\/th>\n      <th>W10_vs_PBS_p.val<\/th>\n      <th>W2_vs_PBS_p.val<\/th>\n      <th>W4_vs_PBS_p.val<\/th>\n      <th>W6_vs_PBS_p.val<\/th>\n      <th>W9_vs_PBS_p.val<\/th>\n      <th>W10_vs_PBS_p.adj<\/th>\n      <th>W2_vs_PBS_p.adj<\/th>\n      <th>W4_vs_PBS_p.adj<\/th>\n      <th>W6_vs_PBS_p.adj<\/th>\n      <th>W9_vs_PBS_p.adj<\/th>\n      <th>W10_vs_PBS_significant<\/th>\n      <th>W2_vs_PBS_significant<\/th>\n      <th>W4_vs_PBS_significant<\/th>\n      <th>W6_vs_PBS_significant<\/th>\n      <th>W9_vs_PBS_significant<\/th>\n      <th>significant<\/th>\n      <th>W10_vs_PBS_ratio<\/th>\n      <th>W2_vs_PBS_ratio<\/th>\n      <th>W4_vs_PBS_ratio<\/th>\n      <th>W6_vs_PBS_ratio<\/th>\n      <th>W9_vs_PBS_ratio<\/th>\n      <th>PBS_centered<\/th>\n      <th>W2_centered<\/th>\n      <th>W4_centered<\/th>\n      <th>W6_centered<\/th>\n      <th>W9_centered<\/th>\n      <th>W10_centered<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7,8,9,10,11,12,19,20,21,22,23,24,25,26,27,28,29]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the significant subset</span></span>
<span><span class="va">sig_deg</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_results.html">get_results</a></span><span class="op">(</span><span class="va">deg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sig_deg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "data.frame"</span></span></code></pre></div>
<p>Function <code>get_df_wide</code> can get the full table from a DEP
result.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get full data set with row information.</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_df_wide.html">get_df_wide</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span><span class="op">)</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-8e9539872343d5916c5b" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-8e9539872343d5916c5b">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["1810009A15Rik","4931429I11Rik","4932418E24Rik","9230110C19Rik","A0A0B4J1G0","A2AG58"],[24.9551095274732,28.6996657590783,27.7859761345177,28.1523061125557,24.9551095274732,28.3492118615544],[27.6644631327717,24.9063297331436,27.9504676339097,27.5988420740033,24.9063297331436,27.5143345824403],[28.4669268731458,29.7534556988078,28.1169301358519,27.1884750302214,25.1199563495669,24.9452045681029],[24.8777938207367,29.9601981115361,28.6312448968252,27.6776461047669,24.8777938207367,28.7101047244458],[24.5981571985069,30.348976398091,27.9113100466282,27.8196831036236,23.7951251617475,28.6340486606798],[27.7464912354593,29.5043546017269,27.0592687659758,27.3761877527553,27.1445700721949,28.4366292052833],[24.5985582492306,29.0374565984077,27.2838511222099,27.4422755546221,26.2134741197254,28.8139933181393],[24.6024629366394,29.5550856110306,27.0881235846591,27.5904867230838,26.606362971737,28.309849473514],[24.6091010896248,28.6799601404505,28.3227277883364,27.838337134962,26.0154515768002,27.826023536677],[24.7640138303909,29.3120878961268,28.5648724877273,28.1962793144718,24.7397668442322,28.0219809254425],[24.7237944745986,29.1638385700444,28.0466557799332,28.0554503198361,26.5117924857343,27.611904569082],[24.7662180915475,29.1007203762722,28.6123680523078,27.53592252132,26.0577440635199,24.7662180915475],[24.7422467846833,28.3858257680321,28.126028635263,28.1152587178825,25.8961730301816,27.9855529792708],[27.6602286949651,24.7303401033948,27.5280084767412,27.491925154824,26.7681671759012,27.7840122152021],[24.7059664296668,24.7059664296668,27.6067833126389,27.9020638257491,26.7283748314279,27.8992808373852],[24.7068102544331,24.7068102544331,27.8273154268193,28.0987803976753,26.8436725533042,27.6219725037432],[27.4926503039888,27.1952766003399,28.2511213484431,28.0138484894029,26.5210039667276,28.3843675539803],[24.7259223645165,24.7259223645165,27.5629151279444,27.1006918635893,27.0050816045653,24.7259223645165],[28.4520987666818,24.7927104118464,27.92594222195,27.8057902111971,27.3259553508398,28.0103395366441],[24.9584013596921,28.2298759914394,28.3206878704604,28.0177949429222,26.7376187270605,27.4431798879579],["Q9D937","E9Q793","Q8C5V8","Q8CC70","A0A0B4J1G0","A2AG58"],["Q9D937","E9Q793","Q8C5V8","Q8CC70","A0A0B4J1G0","A2AG58"],["2","7","15","14","3","10"],["2","7","15","14","3","10"],["2","7","15","14","3","10"],["","","","","","Uncharacterized protein CXorf23 homolog"],["1810009A15Rik","4931429I11Rik","4932418E24Rik","9230110C19Rik","",""],["&gt;Q9D937 Uncharacterized protein C11orf98 homolog OS=Mus musculus OX=10090 PE=1 SV=1","&gt;E9Q793 Jhy protein OS=Mus musculus OX=10090 GN=Jhy PE=1 SV=1","&gt;Q8C5V8 Coiled-coil domain-containing protein 187 OS=Mus musculus OX=10090 GN=Ccdc187 PE=2 SV=1","&gt;Q8CC70 Cilia- and flagella-associated protein 300 OS=Mus musculus OX=10090 GN=Cfap300 PE=2 SV=1","&gt;A0A0B4J1G0 Low affinity immunoglobulin gamma Fc region receptor IV OS=Mus musculus OX=10090 GN=Fcgr4 PE=1 SV=1","&gt;A2AG58 BCLAF1 and THRAP3 family member 3 OS=Mus musculus OX=10090 GN=Bclaf3 PE=1 SV=1"],[1,1,1,1,1,1],[2,7,15,14,3,10],[2,7,15,14,3,10],[2,7,15,14,3,10],[12094000,770730,4431600,14583000,1371700,986190],[35112000,31396,5686400,14419000,0,2576800],[88801000,600530,7911100,13432000,2541900,925390],[42952000,73060000,7356500,10295000,214960,1450600],[114290000,830230,10941000,18398000,15586000,70833000],[31462000,44453000,10549000,21712000,7950500,35955000],[125700000,101360000,11432000,27832000,1476700,1985800],[82430000,64003000,5231400,16129000,14318000,4812900],[148790000,52267000,6075000,16620000,8884100,2590500],[145950000,48484000,4720600,22499000,9379500,3013700],[117910000,1762100,19217000,36050000,11780000,1383500],[138720000,924580,19552000,29538000,3277000,1824300],[71657000,1383500,14223000,17628000,11987000,1132000],[66123000,2082500,16121000,18354000,7910500,870670],[7558300,1271700,13346000,19766000,9806000,1906800],[58457000,380510,5071000,18147000,11549000,3721400],[17117000,391260,4442200,19754000,56719000,1907400],[44816000,741790,6194200,27357000,13334000,1398100],[51917000,907870,13517000,30529000,9592600,73530000],[78577000,222640,8740800,23216000,17502000,37570000],["","","","","",""],["","","","","",""],["","","","","",""],["Q9D937","E9Q793","Q8C5V8","Q8CC70","A0A0B4J1G0","A2AG58"],[true,true,false,false,true,true],[14,6,0,0,3,3],[-2.45011220273459,-4.5825907889157,-0.594623897524808,-0.325015096371638,0.675096776148794,-1.83337908241794],[2.87846565204472,0.945352540918628,0.598944589382955,0.839965589717261,3.45888258629129,2.5274706387483],[0.214176724655065,-1.81861912399854,0.00216034592907377,0.257475246672811,2.06698968122004,0.347045778165178],[0.927772389213524,0.347986038203084,0.996044249940369,0.542388912204464,0.0399013721402975,0.847925514524878],[0.866933162176476,0.1823226782213,0.993977157175838,0.362981490899692,0.00618377978393388,0.740400780496394],[0.170232842925207,-1.3933511048956,0.00766581527663136,0.936050225761,3.14474102708114,0.337052877861032],[-3.28003873339007,-0.975217437513374,-1.2727886147395,-0.572760420384824,-0.161390074387161,-0.611393213367703],[1.07072686624443,3.53832939085859,-0.298244026076235,0.378442326653598,2.11156152162931,2.94922567390416],[-1.10465593357282,1.28155597667261,-0.785516320407869,-0.0971590468656132,0.975085723621074,1.16891623026823],[0.441438365853167,0.386169563581204,0.0184764058559671,0.776465442063812,0.182766279577838,0.313966994659411],[0.297967473853185,0.246420318548969,0.00349903878434647,0.671000307425886,0.0877611850166126,0.183199846249162],[-1.07533489362527,1.20254689838281,-3.41378285695199,-0.432606006111521,1.81691665312506,1.39040170739611],[-3.95067426680866,-1.52253399410395,-0.221770967531578,-0.223421381258582,-0.2700844131667,-2.10349159708457],[0.400091332825838,2.99101283426801,0.75277362113169,0.72778136577984,2.00286718284977,1.45712729018729],[-1.77529146699141,0.734239420082034,0.265501326800056,0.252179992260629,0.866391384841535,-0.323182153448641],[0.232282823099388,0.652572688912452,0.43011417229388,0.442265212074974,0.262701979693033,0.811974218734484],[0.102952340097825,0.500581811022801,0.265289747774225,0.27785599997095,0.125735449866438,0.705663632981673],[-1.72816965246058,0.688972899632933,1.1538447443809,1.12284530151887,1.61438209698822,-0.384418494943673],[-3.21264309741203,-4.95445010094571,-0.836393031742178,-0.227911679873308,0.457823741965359,-1.33731874387147],[1.13812250222247,-0.440903272573747,0.13815155692109,0.723291067165114,2.73077533798183,2.22330014340039],[-1.03726029759478,-2.69767668675973,-0.349120737410544,0.247689693645903,1.59429953997359,0.442990699764458],[0.467084247812703,0.0671624933174611,0.266800207305816,0.425946097111223,0.0348560804885355,0.721665905547384],[0.327458195109841,0.0220659672433565,0.148459968393951,0.286192907600337,0.00891387406837743,0.605378687059526],[-1.00972814962232,-2.53136249336947,-1.51724713722016,1.10285199968408,2.97072279295653,0.526928285675155],[-3.04607593166887,-5.13328450813035,-0.810920705536278,-0.67953749693518,0.406352522845096,-3.00499383547056],[2.28250192311044,0.394658821703979,0.382647781371484,0.485443189153718,3.19013833298759,1.35585588569568],[-0.381787004279214,-2.36931284321319,-0.214136462082397,-0.0970471538907312,1.79824542791635,-0.824568974887438],[0.860109495843798,0.204297661979756,0.625275004838571,0.833651862283974,0.0598693684364187,0.604997289201507],[0.765403607245545,0.0880240427367685,0.458253683360846,0.728772480812042,0.0145240874886391,0.434811359767476],[-0.303453548629154,-1.81526996190165,-0.759846161775964,-0.352814538417816,2.73587053932061,-0.800826183364354],[false,false,false,false,false,false],[false,false,false,false,false,false],[false,false,false,false,false,false],[false,false,false,false,false,false],[false,false,false,false,false,false],[false,false,false,false,false,false]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>PBS_1<\/th>\n      <th>PBS_2<\/th>\n      <th>PBS_3<\/th>\n      <th>PBS_4<\/th>\n      <th>W2_1<\/th>\n      <th>W2_3<\/th>\n      <th>W2_4<\/th>\n      <th>W2_5<\/th>\n      <th>W4_2<\/th>\n      <th>W4_3<\/th>\n      <th>W4_4<\/th>\n      <th>W4_5<\/th>\n      <th>W6_2<\/th>\n      <th>W6_3<\/th>\n      <th>W6_4<\/th>\n      <th>W6_6<\/th>\n      <th>W9_2<\/th>\n      <th>W9_4<\/th>\n      <th>W10_2<\/th>\n      <th>W10_4<\/th>\n      <th>Protein.IDs<\/th>\n      <th>Majority.protein.IDs<\/th>\n      <th>Peptide.counts..all.<\/th>\n      <th>Peptide.counts..razor.unique.<\/th>\n      <th>Peptide.counts..unique.<\/th>\n      <th>Protein.names<\/th>\n      <th>Gene.names<\/th>\n      <th>Fasta.headers<\/th>\n      <th>Number.of.proteins<\/th>\n      <th>Peptides<\/th>\n      <th>Razor...unique.peptides<\/th>\n      <th>Unique.peptides<\/th>\n      <th>iBAQ.PBS_1<\/th>\n      <th>iBAQ.PBS_2<\/th>\n      <th>iBAQ.PBS_3<\/th>\n      <th>iBAQ.PBS_4<\/th>\n      <th>iBAQ.W10_2<\/th>\n      <th>iBAQ.W10_4<\/th>\n      <th>iBAQ.W2_1<\/th>\n      <th>iBAQ.W2_3<\/th>\n      <th>iBAQ.W2_4<\/th>\n      <th>iBAQ.W2_5<\/th>\n      <th>iBAQ.W4_2<\/th>\n      <th>iBAQ.W4_3<\/th>\n      <th>iBAQ.W4_4<\/th>\n      <th>iBAQ.W4_5<\/th>\n      <th>iBAQ.W6_2<\/th>\n      <th>iBAQ.W6_3<\/th>\n      <th>iBAQ.W6_4<\/th>\n      <th>iBAQ.W6_6<\/th>\n      <th>iBAQ.W9_2<\/th>\n      <th>iBAQ.W9_4<\/th>\n      <th>Only.identified.by.site<\/th>\n      <th>Reverse<\/th>\n      <th>Potential.contaminant<\/th>\n      <th>ID<\/th>\n      <th>imputed<\/th>\n      <th>num_NAs<\/th>\n      <th>W10_vs_PBS_CI.L<\/th>\n      <th>W10_vs_PBS_CI.R<\/th>\n      <th>W10_vs_PBS_diff<\/th>\n      <th>W10_vs_PBS_p.adj<\/th>\n      <th>W10_vs_PBS_p.val<\/th>\n      <th>W10_vs_PBS_t.stastic<\/th>\n      <th>W2_vs_PBS_CI.L<\/th>\n      <th>W2_vs_PBS_CI.R<\/th>\n      <th>W2_vs_PBS_diff<\/th>\n      <th>W2_vs_PBS_p.adj<\/th>\n      <th>W2_vs_PBS_p.val<\/th>\n      <th>W2_vs_PBS_t.stastic<\/th>\n      <th>W4_vs_PBS_CI.L<\/th>\n      <th>W4_vs_PBS_CI.R<\/th>\n      <th>W4_vs_PBS_diff<\/th>\n      <th>W4_vs_PBS_p.adj<\/th>\n      <th>W4_vs_PBS_p.val<\/th>\n      <th>W4_vs_PBS_t.stastic<\/th>\n      <th>W6_vs_PBS_CI.L<\/th>\n      <th>W6_vs_PBS_CI.R<\/th>\n      <th>W6_vs_PBS_diff<\/th>\n      <th>W6_vs_PBS_p.adj<\/th>\n      <th>W6_vs_PBS_p.val<\/th>\n      <th>W6_vs_PBS_t.stastic<\/th>\n      <th>W9_vs_PBS_CI.L<\/th>\n      <th>W9_vs_PBS_CI.R<\/th>\n      <th>W9_vs_PBS_diff<\/th>\n      <th>W9_vs_PBS_p.adj<\/th>\n      <th>W9_vs_PBS_p.val<\/th>\n      <th>W9_vs_PBS_t.stastic<\/th>\n      <th>W10_vs_PBS_significant<\/th>\n      <th>W2_vs_PBS_significant<\/th>\n      <th>W4_vs_PBS_significant<\/th>\n      <th>W6_vs_PBS_significant<\/th>\n      <th>W9_vs_PBS_significant<\/th>\n      <th>significant<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>Function <code>get_df_wide</code> can get results of significant
candidates.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get full data set of significant features</span></span>
<span><span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_pg</span>,return_type <span class="op">=</span> <span class="st">"table"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">ends_with</span><span class="op">(</span><span class="st">"significant"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">summary</span></span>
<span><span class="co">#&gt;  W10_vs_PBS_significant W2_vs_PBS_significant W4_vs_PBS_significant</span></span>
<span><span class="co">#&gt;  Mode :logical          Mode :logical         Mode :logical        </span></span>
<span><span class="co">#&gt;  FALSE:152              FALSE:140             FALSE:178            </span></span>
<span><span class="co">#&gt;  TRUE :97               TRUE :109             TRUE :71             </span></span>
<span><span class="co">#&gt;  W6_vs_PBS_significant W9_vs_PBS_significant significant   </span></span>
<span><span class="co">#&gt;  Mode :logical         Mode :logical         Mode:logical  </span></span>
<span><span class="co">#&gt;  FALSE:114             FALSE:141             TRUE:249      </span></span>
<span><span class="co">#&gt;  TRUE :135             TRUE :108</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="result-comparation">Result comparation<a class="anchor" aria-label="anchor" href="#result-comparation"></a>
</h2>
<p>After . <code>plot_multi_venn</code> and
<code>plot_multi_heatmap</code> can directly compare the significant
result from two omics experiments via name identifiers.</p>
<p>Use <code>plot_multi_venn</code> to compare the differentce between
two pipelines (analysis from MQ proteingroups or the peptides
re-aggregation pipeline).</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Test on the contrast W4_vs_PBS.</span></span>
<span><span class="va">diff_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_pg</span>, type <span class="op">=</span> <span class="st">"manual"</span>, test  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W4_vs_PBS"</span><span class="op">)</span>, fdr.type <span class="op">=</span> <span class="st">"Storey's qvalue"</span><span class="op">)</span></span>
<span><span class="va">dep_pg</span> <span class="op">&lt;-</span> <span class="va">diff_pg</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.05</span>,lfc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">diff_pep</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">se_pep</span>,type <span class="op">=</span> <span class="st">"control"</span>, control <span class="op">=</span> <span class="st">"PBS"</span>, fdr.type <span class="op">=</span> <span class="st">"Storey's qvalue"</span><span class="op">)</span></span>
<span><span class="va">dep_pep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span><span class="va">diff_pep</span>,alpha <span class="op">=</span> <span class="fl">0.05</span>, lfc <span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Filter out pg quantified which only have one peptides.</span></span>
<span><span class="va">dep_pep</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">dep_pep</span>,filter_formula <span class="op">=</span> <span class="op">~</span><span class="va">.n</span>  <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">dep_pg</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_se.html">filter_se</a></span><span class="op">(</span><span class="va">dep_pg</span>,filter_formula <span class="op">=</span> <span class="op">~</span><span class="va">Razor...unique.peptides</span>  <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot Venn to explore the difference between two ways.</span></span>
<span><span class="fu"><a href="../reference/plot_multi_venn.html">plot_multi_venn</a></span><span class="op">(</span>omics_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>PG<span class="op">=</span> <span class="va">dep_pg</span>, pep<span class="op">=</span><span class="va">dep_pep</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/comparation-1.png" width="700"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## If set the proteins remained in both result as background. </span></span>
<span><span class="co">## dep_pep only contain two conditions of the study. PBS and W4. See the pe_peptides above.</span></span>
<span><span class="va">overlap_protein</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dep_pep</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_multi_venn.html">plot_multi_venn</a></span><span class="op">(</span>omics_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>PG<span class="op">=</span> <span class="va">dep_pg</span>, pep<span class="op">=</span><span class="va">dep_pep</span><span class="op">)</span>, background <span class="op">=</span> <span class="va">overlap_protein</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/comparation2,-1.png" width="700"></p>
<p>Check the expression of interested proteins/genes accorss omics via
<code>plot_multi_heatmap</code>. Here we take some proteins may be
related to IL-1 signaling, the signaling pathway that is reported to be
upregulated in silicosis mouse model in the original paper.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Take the proteins may response to interleukin-1 as an example.</span></span>
<span><span class="va">IL1_relative_genes</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Irg1"</span>, <span class="st">"Il1rn"</span>, <span class="st">"Saa3"</span>, <span class="st">"Zbp1"</span>, <span class="st">"Ccl6"</span>, </span>
<span>                       <span class="st">"Serpine1"</span>, <span class="st">"Ccl21a"</span>, <span class="st">"Pycard"</span>, <span class="st">"Irak2"</span>, <span class="st">"Vrk2"</span>, </span>
<span>                       <span class="st">"Fn1"</span>, <span class="st">"Il1r1"</span>, <span class="st">"Irf1"</span>, <span class="st">"Ccl9"</span>, <span class="st">"Mapk11"</span>, <span class="st">"Tank"</span>, </span>
<span>                       <span class="st">"Mapk13"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Order the samples by time.</span></span>
<span><span class="va">dep_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Order_cols.html">Order_cols</a></span><span class="op">(</span><span class="va">dep_pg</span>, order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBS"</span>,<span class="st">"W2"</span>,<span class="st">"W4"</span>,<span class="st">"W6"</span>,<span class="st">"W9"</span>,<span class="st">"W10"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot the heatmap upon two DEP results.</span></span>
<span><span class="fu"><a href="../reference/plot_multi_heatmap.html">plot_multi_heatmap</a></span><span class="op">(</span>omics_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>PG<span class="op">=</span> <span class="va">dep_pg</span>, pep<span class="op">=</span><span class="va">dep_pep</span><span class="op">)</span>,</span>
<span>                   choose_name <span class="op">=</span> <span class="va">IL1_relative_genes</span>,</span>
<span>                   width <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/mulheatmap-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="post-analysis">Post analysis<a class="anchor" aria-label="anchor" href="#post-analysis"></a>
</h2>
<p>To reduce the barriers in format conversion between deferential test
and downstream analysis, post analysis suites for DEP results is
available in DEP2.</p>
<div class="section level3">
<h3 id="biological-function-enrichment">Biological function enrichment<a class="anchor" aria-label="anchor" href="#biological-function-enrichment"></a>
</h3>
<p>Biological functional enrichment functions from <em><a href="https://bioconductor.org/packages/3.16/enrichR" class="external-link">enrichR</a></em>,
<em><a href="https://bioconductor.org/packages/3.16/clusterProfiler" class="external-link">clusterProfiler</a></em>
and <em><a href="https://bioconductor.org/packages/3.16/ReactomePA" class="external-link">ReactomePA</a></em>
are packaged in <code>test_ORA</code> and <code>test_GSEA</code> to
perform over enrichment analysis upon results output from
<code>test_diff</code> and <code>add_rejection</code>. Objective
gene/protein list is extracted. Then IDs are transformed though organism
annotation data, and further mapped biological knowledge annotated gene
sets.</p>
<p><code>test_ORA</code> can perform enrichment analysis upon Gene
ontology, KEGG or REACTOME data.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1. Extract gene list</span></span>
<span><span class="va">diff_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_diff.html">test_diff</a></span><span class="op">(</span><span class="va">imp_pg</span>, type <span class="op">=</span> <span class="st">"manual"</span>, test  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W4_vs_PBS"</span><span class="op">)</span>, fdr.type <span class="op">=</span> <span class="st">"Storey's qvalue"</span><span class="op">)</span></span>
<span><span class="va">dep_pg</span> <span class="op">&lt;-</span> <span class="va">diff_pg</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/DEP2_add_rejections.html">add_rejections</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.01</span>,lfc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 2. Check install</span></span>
<span><span class="fu"><a href="../reference/check_enrichment_depends.html">check_enrichment_depends</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># check required packages</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="../reference/check_organismDB_depends.html">check_organismDB_depends</a></span><span class="op">(</span><span class="st">"Mouse"</span><span class="op">)</span> <span class="co"># organism annotation for mouse</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co">## Over representation analysis</span></span>
<span><span class="co"># On one contrast</span></span>
<span><span class="va">res_ora</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_ORA.html">test_ORA</a></span><span class="op">(</span><span class="va">dep_pg</span>, contrasts <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>, species <span class="op">=</span> <span class="st">"Mouse"</span>,type <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 86 gene(s) transfrom to ENTREZID from SYMBOL. 2 gene(s) transfrom to ENTREZID from UNIPROT. 4 gene(s) transfrom to ENTREZID from ALIAS. </span></span>
<span><span class="co">#&gt; Star enrich GO terms by 92 ENTREZIDs.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">res_ora</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] "enrichResult"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "DOSE"</span></span>
<span></span>
<span><span class="co"># On each contrasts</span></span>
<span><span class="va">res_ora2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_ORA.html">test_ORA</a></span><span class="op">(</span><span class="va">dep_pg</span>, by_contrast <span class="op">=</span> <span class="cn">T</span>, species <span class="op">=</span> <span class="st">"Mouse"</span>,type <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 86 gene(s) transfrom to ENTREZID from SYMBOL. 2 gene(s) transfrom to ENTREZID from UNIPROT. 4 gene(s) transfrom to ENTREZID from ALIAS. </span></span>
<span><span class="co">#&gt; Star enrich GO terms by 92 ENTREZIDs.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">res_ora2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "enrichResult"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "DOSE"</span></span>
<span></span>
<span><span class="co">## Gene set enrichment analysis</span></span>
<span><span class="va">res_gsea</span> <span class="op">&lt;-</span> <span class="fu">DEP2</span><span class="fu">::</span><span class="fu"><a href="../reference/test_GSEA.html">test_GSEA</a></span><span class="op">(</span><span class="va">dep_pg</span>, by_contrast <span class="op">=</span> <span class="cn">F</span>, species <span class="op">=</span> <span class="st">"Mouse"</span>,type <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 8324 gene(s) transfrom to ENTREZID from SYMBOL. 110 gene(s) transfrom to ENTREZID from UNIPROT. 488 gene(s) transfrom to ENTREZID from ALIAS. </span></span>
<span><span class="co">#&gt; 60 gene(s) do not match in database.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">res_gsea</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "gseaResult"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "DOSE"</span></span></code></pre></div>
<p>The output of <code>test_ORA</code> and <code>test_GSEA</code> is the
same object as from source functions, which is seamlessly compatible for
plot functions and manipulate function provided for <em>enrichR</em>
result. The visualization methods from <em><a href="https://bioconductor.org/packages/3.16/enrichplot" class="external-link">enrichplot</a></em>
package can seamlessly draw on output of these two output.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Dotplot function from DOSE package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">enrichplot</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: package 'enrichplot' was built under R version 4.2.2</span></span>
<span><span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">res_ora</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/dotplot-1.png" width="700"></p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## GSEA plot on one category</span></span>
<span><span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/gseaplot.html" class="external-link">gseaplot</a></span><span class="op">(</span><span class="va">res_gsea</span>,geneSetID <span class="op">=</span> <span class="st">"GO:0071347"</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/gesaplot-1.png" width="700"></p>
<p><strong>cnet</strong> on interested biological process which enriched
in GSEA.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Cnetplot</span></span>
<span><span class="va">res_gsea2</span> <span class="op">&lt;-</span> <span class="va">res_gsea</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:0071347"</span>,<span class="st">"GO:0035329"</span>,<span class="st">"GO:0032660"</span>,<span class="st">"GO:0090263"</span>,<span class="st">"GO:0035456"</span>,<span class="st">"GO:0038061"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/cnetplot.html" class="external-link">cnetplot</a></span><span class="op">(</span><span class="va">res_gsea2</span>,foldChange <span class="op">=</span> <span class="va">res_gsea</span><span class="op">@</span><span class="va">geneList</span>,node_label <span class="op">=</span> <span class="st">"category"</span>,showCategory <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in cnetplot.enrichResult(x, ...): Use 'color.params = list(foldChange = your_value)' instead of 'foldChange'.</span></span>
<span><span class="co">#&gt;  The foldChange parameter will be removed in the next version.</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">size</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">size</span>, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Warning: Removed 101 rows containing missing values</span></span>
<span><span class="co">#&gt; (`geom_text_repel()`).</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/cnet-1.png" width="700"></p>
<p>More usages and details are available the Yu’s introduction <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">biomedical-knowledge-mining-book</a>.</p>
</div>
<div class="section level3">
<h3 id="expression-pattern-cluster">Expression pattern cluster<a class="anchor" aria-label="anchor" href="#expression-pattern-cluster"></a>
</h3>
<p>In the time-course omics studies or the experiments with complex
group design, it’s not easy to find out key candidates across entire
experiment from cumbersome contrasts. DEP2 provides a
<code>get_tc_cluster</code> function to perform timecourse cluster for
omics data with multiple groups. It cluster features by “cmeans”, and
return a list contains heatmap plot and cluster result. In the omics
experiement with multiple experimental groups,
<code>get_tc_cluster</code> can classify genes/proteins according to
their expression patterns.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Perform cluster on signicant protein in all contrast</span></span>
<span><span class="va">tc_pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_tc_cluster.html">get_tc_cluster</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_pg</span><span class="op">)</span>,</span>
<span>               group_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBS"</span>,<span class="st">"W2"</span>,<span class="st">"W4"</span>,<span class="st">"W6"</span>,<span class="st">"W9"</span>,<span class="st">"W10"</span><span class="op">)</span> <span class="co"># Set group order</span></span>
<span>              <span class="op">)</span></span>
<span><span class="co">#&gt; aabb[1] "start timeclust"</span></span>
<span><span class="co">#&gt; beging 'cm' cluster by dist = 'euclidean'</span></span>
<span><span class="co">#&gt; [1] "finish heatmap"</span></span>
<span><span class="co">## Time course heatmap</span></span>
<span><span class="va">tc_pg</span><span class="op">$</span><span class="va">ht</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/tcheatmap-1.png" width="700"></p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Expression cluster on phospholated sites</span></span>
<span><span class="va">tc_ptm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_tc_cluster.html">get_tc_cluster</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_signicant.html">get_signicant</a></span><span class="op">(</span><span class="va">dep_ptm</span><span class="op">)</span>,</span>
<span>                         group_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBS"</span>,<span class="st">"W2"</span>,<span class="st">"W4"</span>,<span class="st">"W6"</span>,<span class="st">"W9"</span>,<span class="st">"W10"</span><span class="op">)</span>, <span class="co"># Set group order</span></span>
<span>                         row_font_size <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                         k <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                         heatmap_height <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; aabb[1] "start timeclust"</span></span>
<span><span class="co">#&gt; beging 'cm' cluster by dist = 'euclidean'</span></span>
<span><span class="co">#&gt; `use_raster` is automatically set to TRUE for a matrix with more than</span></span>
<span><span class="co">#&gt; 2000 rows. You can control `use_raster` argument by explicitly setting</span></span>
<span><span class="co">#&gt; TRUE/FALSE to it.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Set `ht_opt$message = FALSE` to turn off this message.</span></span>
<span><span class="co">#&gt; [1] "finish heatmap"</span></span>
<span><span class="va">tc_ptm</span><span class="op">$</span><span class="va">ht</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/tcheatmap2-1.png" width="700"></p>
<p>The output of <code>get_tc_cluster</code> contains a heatmap combined
with expression line chart and a cluster result table in store in the
‘res’ slot. The cluster table is a data.frame recording the closest
cluster, membership values and expression values.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The clustering result table</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tc_pg</span><span class="op">$</span><span class="va">res</span><span class="op">)</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">T</span>,pageLength <span class="op">=</span> <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-c403e1f7b162d7d6b0f4" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-c403e1f7b162d7d6b0f4">{"x":{"filter":"none","vertical":false,"data":[["Add2","Aif1","Ank1","Apob","Apobec1","Arpc3"],["Add2","Aif1","Ank1","Apob","Apobec1","Arpc3"],["1","3","1","1","2","6"],["0.850399601367122","0.61203757665229","0.96110505471313","0.347414328160518","0.721838476682026","0.421373835738485"],["1.40316513576088","-2.95383674846645","1.36561856239686","0.770023191760686","-2.91925275953333","-2.77728077765252"],["0.955760049771861","-3.00261654279608","1.60504582100851","0.810079354822982","0.49115726902183","-2.18424394899138"],["3.3467448142682","-2.96374170783676","2.51084415357863","1.97673886603006","-2.92915771890364","-3.03568763902479"],["1.72185719929216","-1.26382193700546","1.52969869041465","1.02185198065149","-2.99656846626982","-0.242083514046044"],["-1.17976238487415","0.660651522946313","-1.3539065899566","-0.422297893974516","0.490888885916331","-0.310026982206892"],["0.738692770143125","2.20399710532227","0.969056724811935","0.809063948714488","0.990362202836156","0.0859763791939443"],["0.454436133223691","0.939742783460872","0.0606094953455383","0.346875690433734","0.752896223811412","0.543920263841137"],["0.372676510006997","0.565227804927591","0.318148140626498","-0.00945952340494927","0.778161667235405","0.192892037788923"],["-1.16881849375616","1.50287186837278","-0.978807836434228","-1.79624891613904","0.414429204987453","0.336148318751327"],["-1.01390575299013","0.181303703494162","-1.09898432359997","-1.56109300486699","0.0572895705329302","0.437131617765203"],["-1.60913166393512","1.0735227411656","-0.258315345156468","-0.624076793958487","1.02823299315419","1.14464041625407"],["0.103275922578305","0.944632902827333","-0.350548727881858","-1.14106316271157","0.434183238146808","0.287487933472185"],["0.284319854715843","0.129000828585493","-0.328133110348382","-0.302432571850858","0.433874661970428","0.673625826011815"],["0.201384419502492","0.218937617140245","-0.358624221733798","0.0523036330228948","0.563265541721456","1.12260556842367"],["-0.709619846943678","0.545253592049178","-0.268285741513004","0.0763236995886061","0.721167115440821","0.925178593859634"],["-1.07110932894792","0.0512807120004197","-0.522983609461228","-0.19539481745262","0.761114051144901","0.87875747272075"],["0.361086838227603","-0.0534141775450649","-0.154670536404176","1.12588259814945","-0.139427477690685","0.120048466599869"],["-1.05199721886448","-0.228191835240935","-1.49875588067605","0.62325858790873","0.95939313468849","0.605097494531854"],["-1.31953673349061","0.934238476935956","-0.904369130554194","-0.489895702146786","0.120520896471991","0.512180334685659"],["-0.81951822368892","0.514961289662548","-0.282636534462647","-1.07043916457733","-0.0125302346831546","0.683632138021537"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>Timecourse_cluster<\/th>\n      <th>mebership<\/th>\n      <th>PBS_1<\/th>\n      <th>PBS_2<\/th>\n      <th>PBS_3<\/th>\n      <th>PBS_4<\/th>\n      <th>W2_1<\/th>\n      <th>W2_3<\/th>\n      <th>W2_4<\/th>\n      <th>W2_5<\/th>\n      <th>W4_2<\/th>\n      <th>W4_3<\/th>\n      <th>W4_4<\/th>\n      <th>W4_5<\/th>\n      <th>W6_2<\/th>\n      <th>W6_3<\/th>\n      <th>W6_4<\/th>\n      <th>W6_6<\/th>\n      <th>W9_2<\/th>\n      <th>W9_4<\/th>\n      <th>W10_2<\/th>\n      <th>W10_4<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"pageLength":6,"columnDefs":[{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[6,10,25,50,100]}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="protein-protein-interaction">Protein-protein interaction<a class="anchor" aria-label="anchor" href="#protein-protein-interaction"></a>
</h3>
<p>DEP2 provides a function <code>test_PPI</code> to construct
protein-protein interaction network though STRING data.
<code>test_PPI</code> perform PPI analysis locally and will download the
species data from STRING. Notably, <code>test_PPI</code> use a local
STRING data. Function <code>load_PPIdata</code> checks whether local
database is exist. If STRING data is absent, <code>load_PPIdata</code>
and <code>test_PPI</code> will try to download the PPI data of the
species from website and save in .</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Check required packages for network construction.</span></span>
<span><span class="fu"><a href="../reference/check_PPI_depends.html">check_PPI_depends</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co"># 2. Load STRING data. </span></span>
<span><span class="co">## If local STRING is missing, load_PPIdata will download to system.file("PPIdata", "Mouse",package = "DEP2")</span></span>
<span><span class="fu"><a href="../reference/load_PPIdata.html">load_PPIdata</a></span><span class="op">(</span>speciesname <span class="op">=</span> <span class="st">"Mouse"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Input species is Mouse   The species ID is 10090</span></span>
<span><span class="co">#&gt; fristDownload</span></span>
<span><span class="co">#&gt; fristDownload</span></span>
<span><span class="co">#&gt; fristDownload</span></span>
<span><span class="co">#&gt; load succeed</span></span>
<span><span class="co">#&gt; [1] "Loading finished!"</span></span>
<span></span>
<span><span class="co"># 3.Load STRING data and perform PPI analysis. </span></span>
<span><span class="co">## It also automatically download data if loacl files is missing. </span></span>
<span><span class="va">PPI_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_PPI.html">test_PPI</a></span><span class="op">(</span></span>
<span>  <span class="va">dep_pg</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="st">"W4_vs_PBS"</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Mouse"</span>,</span>
<span>  STRING.version <span class="op">=</span> <span class="st">"11.5"</span>,</span>
<span>  score_cutoff <span class="op">=</span> <span class="fl">400</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 86 gene(s) transfrom to ENTREZID from SYMBOL. 2 gene(s) transfrom to ENTREZID from UNIPROT. 4 gene(s) transfrom to ENTREZID from ALIAS.</span></span>
<span></span>
<span><span class="co"># igraph network</span></span>
<span><span class="va">PPI_ig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PPInetwork.html">PPInetwork</a></span><span class="op">(</span><span class="va">PPI_res</span>,returntype <span class="op">=</span> <span class="st">"igraph"</span><span class="op">)</span></span>
<span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/plot.igraph.html" class="external-link">plot.igraph</a></span><span class="op">(</span><span class="va">PPI_ig</span><span class="op">)</span></span></code></pre></div>
<p><img src="DEP2_analysis2_files/figure-html/PPI1-1.png" width="700"></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## visNetwork network</span></span>
<span><span class="fu"><a href="../reference/PPInetwork.html">PPInetwork</a></span><span class="op">(</span><span class="va">PPI_res</span>,returntype <span class="op">=</span> <span class="st">"visNetwork"</span><span class="op">)</span> </span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.2.1 (2022-06-23 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Windows 10 x64 (build 19043)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] enrichplot_1.18.3           org.Mm.eg.db_3.16.0        </span></span>
<span><span class="co">#&gt;  [3] AnnotationDbi_1.60.0        ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [5] patchwork_1.1.2             missForest_1.5             </span></span>
<span><span class="co">#&gt;  [7] ggplot2_3.4.1               dplyr_1.1.0                </span></span>
<span><span class="co">#&gt;  [9] tibble_3.1.8                DEP2_0.3.06.12             </span></span>
<span><span class="co">#&gt; [11] R6_2.5.1                    limma_3.54.1               </span></span>
<span><span class="co">#&gt; [13] MSnbase_2.24.2              ProtGenerics_1.30.0        </span></span>
<span><span class="co">#&gt; [15] mzR_2.32.0                  Rcpp_1.0.10                </span></span>
<span><span class="co">#&gt; [17] MsCoreUtils_1.10.0          SummarizedExperiment_1.28.0</span></span>
<span><span class="co">#&gt; [19] Biobase_2.58.0              GenomicRanges_1.50.2       </span></span>
<span><span class="co">#&gt; [21] GenomeInfoDb_1.34.9         IRanges_2.32.0             </span></span>
<span><span class="co">#&gt; [23] S4Vectors_0.36.1            BiocGenerics_0.44.0        </span></span>
<span><span class="co">#&gt; [25] MatrixGenerics_1.10.0       matrixStats_0.63.0         </span></span>
<span><span class="co">#&gt; [27] tidyr_1.3.0                 BiocStyle_2.26.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] rappdirs_0.3.3              R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt;   [3] ragg_1.2.5                  bit64_4.0.5                </span></span>
<span><span class="co">#&gt;   [5] knitr_1.42                  R.utils_2.12.2             </span></span>
<span><span class="co">#&gt;   [7] DelayedArray_0.23.2         data.table_1.14.8          </span></span>
<span><span class="co">#&gt;   [9] KEGGREST_1.38.0             RCurl_1.98-1.10            </span></span>
<span><span class="co">#&gt;  [11] AnnotationFilter_1.22.0     doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [13] generics_0.1.3              snow_0.4-4                 </span></span>
<span><span class="co">#&gt;  [15] preprocessCore_1.60.2       cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [17] RSQLite_2.3.0               shadowtext_0.1.2           </span></span>
<span><span class="co">#&gt;  [19] proxy_0.4-27                bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [21] assertthat_0.2.1            TCseq_1.22.6               </span></span>
<span><span class="co">#&gt;  [23] viridis_0.6.2               xfun_0.39                  </span></span>
<span><span class="co">#&gt;  [25] jquerylib_0.1.4             babelgene_22.9             </span></span>
<span><span class="co">#&gt;  [27] evaluate_0.20               fansi_1.0.4                </span></span>
<span><span class="co">#&gt;  [29] dbplyr_2.3.0                igraph_1.4.0               </span></span>
<span><span class="co">#&gt;  [31] DBI_1.1.3                   geneplotter_1.76.0         </span></span>
<span><span class="co">#&gt;  [33] htmlwidgets_1.6.1           purrr_1.0.1                </span></span>
<span><span class="co">#&gt;  [35] ellipsis_0.3.2              RSpectra_0.16-1            </span></span>
<span><span class="co">#&gt;  [37] crosstalk_1.2.0             ggnewscale_0.4.8           </span></span>
<span><span class="co">#&gt;  [39] QFeatures_1.8.0             bookdown_0.34              </span></span>
<span><span class="co">#&gt;  [41] annotate_1.76.0             vctrs_0.5.2                </span></span>
<span><span class="co">#&gt;  [43] imputeLCMD_2.1              abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [45] cachem_1.0.6                withr_2.5.0                </span></span>
<span><span class="co">#&gt;  [47] ggforce_0.4.1               RVenn_1.1.0                </span></span>
<span><span class="co">#&gt;  [49] HDO.db_0.99.1               itertools_0.1-3            </span></span>
<span><span class="co">#&gt;  [51] treeio_1.22.0               GenomicAlignments_1.34.0   </span></span>
<span><span class="co">#&gt;  [53] fdrtool_1.2.17              MultiAssayExperiment_1.24.0</span></span>
<span><span class="co">#&gt;  [55] cluster_2.1.4               DOSE_3.24.2                </span></span>
<span><span class="co">#&gt;  [57] BiocBaseUtils_1.0.0         ape_5.7                    </span></span>
<span><span class="co">#&gt;  [59] lazyeval_0.2.2              crayon_1.5.2               </span></span>
<span><span class="co">#&gt;  [61] glmnet_4.1-6                edgeR_3.40.2               </span></span>
<span><span class="co">#&gt;  [63] pkgconfig_2.0.3             labeling_0.4.2             </span></span>
<span><span class="co">#&gt;  [65] units_0.8-1                 tweenr_2.0.2               </span></span>
<span><span class="co">#&gt;  [67] nlme_3.1-162                rlang_1.0.6                </span></span>
<span><span class="co">#&gt;  [69] lifecycle_1.0.3             sandwich_3.0-2             </span></span>
<span><span class="co">#&gt;  [71] downloader_0.4              affyio_1.68.0              </span></span>
<span><span class="co">#&gt;  [73] randomForest_4.7-1.1        rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt;  [75] polyclip_1.10-4             graph_1.76.0               </span></span>
<span><span class="co">#&gt;  [77] rngtools_1.5.2              Matrix_1.5-3               </span></span>
<span><span class="co">#&gt;  [79] aplot_0.1.9                 zoo_1.8-11                 </span></span>
<span><span class="co">#&gt;  [81] GlobalOptions_0.1.2         png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [83] viridisLite_0.4.1           rjson_0.2.21               </span></span>
<span><span class="co">#&gt;  [85] bitops_1.0-7                R.oo_1.25.0                </span></span>
<span><span class="co">#&gt;  [87] gson_0.0.9                  visNetwork_2.1.2           </span></span>
<span><span class="co">#&gt;  [89] KernSmooth_2.23-20          Biostrings_2.66.0          </span></span>
<span><span class="co">#&gt;  [91] blob_1.2.3                  doRNG_1.8.6                </span></span>
<span><span class="co">#&gt;  [93] shape_1.4.6                 classInt_0.4-8             </span></span>
<span><span class="co">#&gt;  [95] stringr_1.5.0               qvalue_2.30.0              </span></span>
<span><span class="co">#&gt;  [97] gridGraphics_0.5-1          tmvtnorm_1.5               </span></span>
<span><span class="co">#&gt;  [99] reactome.db_1.82.0          scales_1.2.1               </span></span>
<span><span class="co">#&gt; [101] graphite_1.44.0             memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [103] magrittr_2.0.3              plyr_1.8.8                 </span></span>
<span><span class="co">#&gt; [105] zlibbioc_1.44.0             scatterpie_0.1.8           </span></span>
<span><span class="co">#&gt; [107] compiler_4.2.1              RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt; [109] pcaMethods_1.90.0           clue_0.3-64                </span></span>
<span><span class="co">#&gt; [111] DESeq2_1.38.3               Rsamtools_2.14.0           </span></span>
<span><span class="co">#&gt; [113] cli_3.6.0                   affy_1.76.0                </span></span>
<span><span class="co">#&gt; [115] XVector_0.38.0              MASS_7.3-58.2              </span></span>
<span><span class="co">#&gt; [117] tidyselect_1.2.0            vsn_3.66.0                 </span></span>
<span><span class="co">#&gt; [119] stringi_1.7.12              textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [121] highr_0.10                  yaml_2.3.7                 </span></span>
<span><span class="co">#&gt; [123] GOSemSim_2.24.0             askpass_1.1                </span></span>
<span><span class="co">#&gt; [125] locfit_1.5-9.7              norm_1.0-10.0              </span></span>
<span><span class="co">#&gt; [127] MALDIquant_1.22             ggrepel_0.9.3              </span></span>
<span><span class="co">#&gt; [129] sass_0.4.5                  fastmatch_1.1-3            </span></span>
<span><span class="co">#&gt; [131] tools_4.2.1                 parallel_4.2.1             </span></span>
<span><span class="co">#&gt; [133] circlize_0.4.15             rstudioapi_0.14            </span></span>
<span><span class="co">#&gt; [135] foreach_1.5.2               gridExtra_2.3              </span></span>
<span><span class="co">#&gt; [137] farver_2.1.1                mzID_1.36.0                </span></span>
<span><span class="co">#&gt; [139] Rtsne_0.16                  ggraph_2.1.0               </span></span>
<span><span class="co">#&gt; [141] digest_0.6.31               BiocManager_1.30.19        </span></span>
<span><span class="co">#&gt; [143] ncdf4_1.21                  org.Hs.eg.db_3.16.0        </span></span>
<span><span class="co">#&gt; [145] httr_1.4.4                  sf_1.0-9                   </span></span>
<span><span class="co">#&gt; [147] ComplexHeatmap_2.14.0       colorspace_2.1-0           </span></span>
<span><span class="co">#&gt; [149] XML_3.99-0.13               fs_1.6.1                   </span></span>
<span><span class="co">#&gt; [151] reticulate_1.28             umap_0.2.10.0              </span></span>
<span><span class="co">#&gt; [153] splines_4.2.1               yulab.utils_0.0.6          </span></span>
<span><span class="co">#&gt; [155] ggVennDiagram_1.2.2         tidytree_0.4.2             </span></span>
<span><span class="co">#&gt; [157] pkgdown_2.0.7               graphlayouts_0.8.4         </span></span>
<span><span class="co">#&gt; [159] gmm_1.7                     ggplotify_0.1.0            </span></span>
<span><span class="co">#&gt; [161] systemfonts_1.0.4           xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [163] ggtree_3.6.2                jsonlite_1.8.4             </span></span>
<span><span class="co">#&gt; [165] tidygraph_1.2.3             ggfun_0.0.9                </span></span>
<span><span class="co">#&gt; [167] pillar_1.8.1                htmltools_0.5.4            </span></span>
<span><span class="co">#&gt; [169] glue_1.6.2                  fastmap_1.1.0              </span></span>
<span><span class="co">#&gt; [171] clusterProfiler_4.6.2       DT_0.27                    </span></span>
<span><span class="co">#&gt; [173] BiocParallel_1.32.5         class_7.3-21               </span></span>
<span><span class="co">#&gt; [175] codetools_0.2-19            fgsea_1.24.0               </span></span>
<span><span class="co">#&gt; [177] mvtnorm_1.1-3               utf8_1.2.3                 </span></span>
<span><span class="co">#&gt; [179] lattice_0.20-45             bslib_0.4.2                </span></span>
<span><span class="co">#&gt; [181] clusterProfiler.dplyr_0.0.2 ReactomePA_1.42.0          </span></span>
<span><span class="co">#&gt; [183] magick_2.7.4                GO.db_3.16.0               </span></span>
<span><span class="co">#&gt; [185] openssl_2.0.5               survival_3.5-3             </span></span>
<span><span class="co">#&gt; [187] rmarkdown_2.20              desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [189] munsell_0.5.0               e1071_1.7-13               </span></span>
<span><span class="co">#&gt; [191] GetoptLong_1.0.5            GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">#&gt; [193] iterators_1.0.14            impute_1.72.3              </span></span>
<span><span class="co">#&gt; [195] msigdbr_7.5.1               reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [197] gtable_0.3.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Sticker:2020" class="csl-entry">
Sticker, A., L. Goeminne, L. Martens, and L. Clement. 2020.
<span>“Robust Summarization and Inference in Proteome-Wide Label-Free
Quantification.”</span> Journal Article. <em>Mol Cell Proteomics</em> 19
(7): 1209–19. <a href="https://doi.org/10.1074/mcp.RA119.001624" class="external-link">https://doi.org/10.1074/mcp.RA119.001624</a>.
</div>
<div id="ref-Wu:2021" class="csl-entry">
Wu, T., E. Hu, S. Xu, M. Chen, P. Guo, Z. Dai, T. Feng, et al. 2021.
<span>“clusterProfiler 4.0: A Universal Enrichment Tool for Interpreting
Omics Data.”</span> Journal Article. <em>Innovation (Camb)</em> 2 (3):
100141. <a href="https://doi.org/10.1016/j.xinn.2021.100141" class="external-link">https://doi.org/10.1016/j.xinn.2021.100141</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zhenhuan Feng, Xiaofei Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
