<h1 id="introduction">Introduction</h1>
<p>Welcome to the shiny app of <code>DEP2</code>. Here we provide an interactive platform for biologists with little programming experience. This shiny application provides a differentially analyzing pipeline for various types of mass spectrometry-based quantitative proteomics data, including protein-level quantitative data and peptide-level quantitative data. Differential analysis steps including data filter, normalization, imputation, statistical test and FDR correction. To extend compatibility of inputs generated by various quantitative software, nearly all customization options in DEP2 are included in this application. Default parameters and detailed tips are provided to guide unfamiliar users. Generally, the order of the analytical steps is from the top to the bottom in the sidebar. To build an integrative toolkit, several additional analysis modules based on other packages are available. Analyze functions are separated into individual modules. Further details can be available in the DEP2 website: mildpiggy.github.io/DEP2/articles/04_app_demo.html.</p>
<!-- [简图](简图) -->
<h1 id="create-analysis-modules">Create analysis modules</h1>
<p>Analysis modules can be imported by clicking Add module and selecting the desired analysis module. The modules are categorized into two types: Omics analysis and Post analysis. The former focuses on differential tests, data exploration, and visualization, while the latter focuses on functional analysis and multi-omics comparison. After adding a selected module, a new tab will be added to the application page. The tab’s name will be a combination of the analysis type and a serial number. For omics analysis, a unique remark name is required as a title or label.</p>
<h1 id="omics-analysis-modules">Omics analysis modules</h1>
<p>This application offers four types of <strong>omics differential analysis pipeline</strong> to choose from.</p>
<ul>
<li>
<p><strong>Quantificated proteome(based on proteingroup quantification)</strong></p>
<p>This module performs differential analysis of protein-level quantitative data, such as <em>intensity</em>, <em>LFQ intensity</em>, or <em>quantity.</em></p>
</li>
<li>
<p><strong>Quantificated proteome(based on peptide quantification)</strong></p>
<p>This module performs differential analysis of proteins based on peptide-level quantitative data through peptide-to-protein aggregation.</p>
</li>
<li>
<p><strong>Modification-specific proteome(modified peptide quantification)</strong></p>
<p>This module performs differential analysis of modification-specific peptides by using peptide quantity obtained from modification-specific enrichment.</p>
</li>
<li>
<p><strong>Additional transcriptome</strong></p>
<p>This module performs differential analysis of parallel transcriptome using the <a href="https://bioconductor.org/packages/DESeq2/">DEseq2</a> package.</p>
</li>
</ul>
<h2 id="analysis-from-proteingroups-quantitative-data">Analysis from proteingroups quantitative data.</h2>
<p>This module performs a differential expression/enrichment test for protein-level (proteingroup) quantity data in tabular input (.txt file) format, which is generated by various quantitative proteomics software such as MaxQuant. The test is conducted using the generalized approach <code>limma</code>. Below is the graphical user interface illustration of this module.</p>
<p><img src="DEP_illu1.png" alt="Illustration of DEP based on proteingroup" /></p>
<p>Options for analyzing DEP2 are located in the sidebar area. In the DEP2 pipeline, the expression/enrichment data is first uploaded in the <strong>Files</strong> panel. This includes the <code>proteingroup.txt</code> (or other quantitative table) and <code>ExperimentalDesign.txt</code> (optional, only needed when grouping experiments by colnames). The application has a few requirements for the input table: 1. The tabular quantitative table should have columns separated by tabs (.txt file). 2. The quantitative table should include protein names, gene symbols, and quantitative values of groups.</p>
<p>Next, the columns are assigned in the <strong>Columns</strong> panel, such as the <code>Name column</code>, <code>ID column</code>, and <code>Expression columns</code>.</p>
<ul>
<li>
<p><code>Name column</code>: This refers to the column containing the <strong>gene symbol</strong> that corresponds to the “name” column in the result. If a proteingroup has multiple gene symbols, only the first one will be chosen as the <em>name</em>.</p>
</li>
<li>
<p><code>ID column</code>, the column containing the protein ID, such as the <strong>Uniprot ID</strong>, which corresponds to the “ID” column in the result. If a proteingroup contains multiple IDs, only the first one will be chosen as the <em>ID</em>.</p>
</li>
<li>
<p><code>Expression columns</code>: These are the columns of expression/enrichment assays, corresponding to the samples.</p>
</li>
</ul>
<p>The application automatically fills in these columns if handling a table from MaxQuant. Finally, the user selects the <code>imputation</code>, <code>FDR correction</code>, <code>Threshold method</code>, and clicks the Analyze button. The result table and visual figures are located on the right side of the page, providing fine-tuning options and a download button with detailed tips.</p>
<h2 id="analysis-from-peptide-quantitative-data">Analysis from peptide quantitative data.</h2>
<p>DEP2 also offers another workflow for aggregating proteingroup intensities from peptide quantitative data using the <code>aggregateFeatures</code> function from the <code>Qfeatures</code> package. This function aggregates quantitative features from the peptide assay into a proteingroup assay using a highly customizable approach. In this workflow, a <code>limma</code> differential test is performed on the re-aggregated assay without the <code>proteingroup</code> table from software.</p>
<p>The analysis settings in this module are similar to the previous one, with an additional step: aggregation. The “Aggregation” panel provides options for peptide quantitative/intensity data filtering, normalization, and aggregation. The expression columns must contain the column names of the quantitative assay. After selecting the appropriate options, the user clicks the <code>Aggregate</code> button to begin the aggregation process, which may take a minute or two. Once the aggregation is complete, the <code>Columns</code> panel provides the option for control selection.</p>
<h2 id="analysis-ptm-peptide-quantitative-data">Analysis PTM-peptide quantitative data.</h2>
<p>In DEP2, we have expanded the differentially enriched test to include modification-specific proteomics (such as phosphoproteomics or ubiquitomics) directly on modified-peptide intensity. This requires the input of an enriched modified peptide quantitative table. The pipeline and options are similar to the protein-level analysis. The modified-peptide intensity assay is extracted from the uploaded table, filtered based on missing values and a localization probability (selectable in the <code>Cutoff based on</code> input). The enriched peptide intensity can be corrected by the parallel proteome data (an existing proteome analysis result). Imputation and the differentially enriched test are performed similar to those in the protein-level analysis.</p>
<h2 id="analysis-rnaseq-data-by-deseq2">Analysis RNAseq data by DEseq2.</h2>
<p>Due to the availability of multiple omics experiments, such as RNA-sequencing and quantitative proteomics performed in parallel, we also provide an RNA-sequencing analysis module based on the workflow from the <code>DESeq2</code> package. Both the <strong>gene count</strong> table and <strong>experimental design</strong> table are required. In the gene count table, the first column should contain the ENSEMBL id of genes. ID transformation can be used to transform the <strong>ENSEMBL</strong> to <strong>gene symbol</strong> for the convenience of result interpretation and subsequent post-analysis.</p>
<h1 id="post-analysis-modules">Post analysis modules</h1>
<p>There are four types of post-analysis modules that can be utilized on omics analysis results:</p>
<ul>
<li>
<p><strong>Over-representation analysis</strong></p>
<p>The classical geneontology(GO) enrichment analysis</p>
</li>
<li>
<p><strong>GSEA</strong></p>
<p>The gene set enrichment analysis(GSEA), an alternative method for analyzing functional enrichment.</p>
</li>
<li>
<p><strong>Protein-protein interaction</strong></p>
<p>Building a protein-protein interaction (PPI) network of input genes/proteins based on the string database.</p>
</li>
<li>
<p><strong>Timecourse cluster</strong></p>
<p>An expression pattern clustering method for timecourse or multiple groups omics experiments.</p>
</li>
<li>
<p><strong>Integrated analysis</strong></p>
<p>A comparison tool for mulit-omics analysis.</p>
</li>
</ul>
<p>Each imported modules could run independently, the results from omics analysis will be transmitted to post analysis modules.</p>
<h2 id="over-representation-analysis">Over-representation analysis</h2>
<p>This module relies on the <code>enrichGO</code> functions from the <code>clusterProfiler</code> and <code>ReactomePA</code> packages. The GO annotation and gene ID conversion data are from <a href="https://www.bioconductor.org/packages/release/BiocViews.html#___OrgDb">orgDB</a>, an organism annotation database. This module accepts either the <strong>symbol</strong> of the result genes from the omics analysis or text input by pasting the candidate gene symbols.</p>
<!-- To optimize the speed of `enrichGO`, we have modified the function to avoid extracting data from orgDB in each run by saving the intermediate data. -->
<p>The differentially expressed or enriched results from existing omics analysis modules can be transmitted for candidate gene list extraction. We have designed a drag-and-drop input feature using the <code>dndselectr</code> package (<a href="https://github.com/serenity-r/dndselectr">https://github.com/serenity-r/dndselectr</a>), which provides an intuitive way to set extraction options by using mouse drag and drop functionality.</p>
<p><img src="DEP_illu2_DDN.png" alt="Illustration of ORA input" /></p>
<h2 id="gsea">GSEA</h2>
<p>This module is similar to the over-representation analysis module, is also based on the clusterProfiler package. The main difference is that the GSEA analysis accepts the entire gene list along with log2 fold change values, rather than just candidate symbols.</p>
<h2 id="protein-protein-interaction">Protein-protein interaction</h2>
<p>To study interactors, we have designed a PPI analysis module that utilizes a local string database of different organisms. This database can be downloaded from <code>STRING</code> (<a href="https://string-db.org/cgi/download">https://string-db.org/cgi/download</a>) and includes gzip files such as <strong>protein.links.detailed</strong>, <strong>protein.aliases</strong>, and <strong>protein.info</strong>. These files should be stored in the <strong>PPIdata</strong> folder, which is located within the DEP2 package path.</p>
<h2 id="timecourse-cluster">Timecourse cluster</h2>
<p>The timecourse cluster module is a clustering method specifically designed for expression pattern analysis in timecourse studies or experiments with multiple groups. The clustering of expression/enrichment assays is performed using the <code>timeclust</code> function in the <code>TCseq</code> package.</p>
<h2 id="integrated-analysis">Integrated analysis</h2>
<p>Multi-omics studies have become increasingly important in biological pathway research and drug target discovery. DEP2 provides an easy-to-use module for comparing and visualizing results from different omics datasets in the shiny interface.</p>
<h2 id="more-details">More details</h2>
<p>Step to step demonstration can be found in the DEP2 website at <a href="https://mildpiggy.github.io/DEP2/articles/04_app_demo.html">https://mildpiggy.github.io/DEP2/articles/04_app_demo.html</a>. It is also worth paying attention to the hover tips for each option.</p>

